// semiotic/lib/annotationRules/networkframeRules.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.svgHighlightRule = exports.svgHullEncloseRule = exports.svgRectEncloseRule = exports.svgEncloseRule = exports.svgReactAnnotationRule = exports.svgNodeRule = exports.htmlFrameHoverRule = void 0;

var React = _interopRequireWildcard(require("react"));

var _Annotation = _interopRequireDefault(require("../Annotation"));

var _AnnotationCalloutCircle = _interopRequireDefault(require("react-annotation/lib/Types/AnnotationCalloutCircle"));

var _d3Hierarchy = require("d3-hierarchy");

var _baseRules = require("./baseRules");

var _SpanOrDiv = _interopRequireDefault(require("../SpanOrDiv"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var htmlFrameHoverRule = function htmlFrameHoverRule(_ref) {
  var baseD = _ref.d,
      i = _ref.i,
      tooltipContent = _ref.tooltipContent,
      useSpans = _ref.useSpans,
      nodes = _ref.nodes,
      edges = _ref.edges,
      nodeIDAccessor = _ref.nodeIDAccessor;
  var d = baseD.x && baseD.y ? baseD : baseD.edge ? _objectSpread({}, edges.find(function (p) {
    return nodeIDAccessor(p.source) === nodeIDAccessor(baseD.source) && nodeIDAccessor(p.target) === nodeIDAccessor(baseD.target);
  }) || {}, baseD) : nodes.find(function (p) {
    return nodeIDAccessor(p) === baseD.id;
  });
  if (!d) return null;
  var content = d.edge ? React.createElement(_SpanOrDiv.default, {
    span: useSpans,
    className: "tooltip-content"
  }, React.createElement("p", {
    key: "html-annotation-content-1"
  }, (d.source || d.edge.source).id, " to ", (d.target || d.edge.target).id)) : React.createElement(_SpanOrDiv.default, {
    span: useSpans,
    className: "tooltip-content"
  }, React.createElement("p", {
    key: "html-annotation-content-1"
  }, d.id), React.createElement("p", {
    key: "html-annotation-content-2"
  }, "Degree: ", d.degree));

  if (d.type === "frame-hover" && tooltipContent) {
    content = tooltipContent(d);
  }

  return React.createElement(_SpanOrDiv.default, {
    span: useSpans,
    key: "network-annotation-label-".concat(i),
    className: "annotation annotation-network-label ".concat(d.className || ""),
    style: {
      position: "absolute",
      top: "".concat(d.y, "px"),
      left: "".concat(d.x, "px")
    }
  }, content);
};

exports.htmlFrameHoverRule = htmlFrameHoverRule;

var svgNodeRule = function svgNodeRule(_ref2) {
  var d = _ref2.d,
      i = _ref2.i,
      nodeSizeAccessor = _ref2.nodeSizeAccessor;

  if (!d) {
    return null;
  }

  var noteData = _extends({
    dx: d.dx || -25,
    dy: d.dy || -25,
    x: d.x,
    y: d.y,
    note: {
      label: d.label,
      orientation: d.orientation,
      align: d.align
    },
    connector: {
      end: "arrow"
    }
  }, d, {
    type: _AnnotationCalloutCircle.default,
    subject: {
      radius: d.radius || d.radius || nodeSizeAccessor(d)
    }
  });

  return React.createElement(_Annotation.default, {
    key: d.key || "annotation-".concat(i),
    noteData: noteData
  });
};

exports.svgNodeRule = svgNodeRule;

var svgReactAnnotationRule = function svgReactAnnotationRule(_ref3) {
  var d = _ref3.d,
      i = _ref3.i,
      projectedNodes = _ref3.projectedNodes,
      nodeIDAccessor = _ref3.nodeIDAccessor;
  var selectedNode = d.x && d.y ? d : projectedNodes.find(function (p) {
    return nodeIDAccessor(p) === d.id;
  });

  if (!selectedNode) {
    return null;
  }

  var noteData = _extends({
    dx: 0,
    dy: 0,
    x: selectedNode.x,
    y: selectedNode.y,
    note: {
      label: d.label
    },
    connector: {
      end: "arrow"
    }
  }, d, {
    type: typeof d.type === "function" ? d.type : undefined
  });

  return React.createElement(_Annotation.default, {
    key: d.key || "annotation-".concat(i),
    noteData: noteData
  });
};

exports.svgReactAnnotationRule = svgReactAnnotationRule;

var svgEncloseRule = function svgEncloseRule(_ref4) {
  var d = _ref4.d,
      i = _ref4.i,
      projectedNodes = _ref4.projectedNodes,
      nodeIDAccessor = _ref4.nodeIDAccessor,
      nodeSizeAccessor = _ref4.nodeSizeAccessor;
  var selectedNodes = projectedNodes.filter(function (p) {
    return d.ids.indexOf(nodeIDAccessor(p)) !== -1;
  });

  if (selectedNodes.length === 0) {
    return null;
  }

  var circle = (0, _d3Hierarchy.packEnclose)(selectedNodes.map(function (p) {
    return {
      x: p.x,
      y: p.y,
      r: nodeSizeAccessor(p)
    };
  }));
  return (0, _baseRules.circleEnclosure)({
    circle: circle,
    d: d,
    i: i
  });
};

exports.svgEncloseRule = svgEncloseRule;

var svgRectEncloseRule = function svgRectEncloseRule(_ref5) {
  var d = _ref5.d,
      i = _ref5.i,
      projectedNodes = _ref5.projectedNodes,
      nodeIDAccessor = _ref5.nodeIDAccessor,
      nodeSizeAccessor = _ref5.nodeSizeAccessor;
  var selectedNodes = projectedNodes.filter(function (p) {
    return d.ids.indexOf(nodeIDAccessor(p)) !== -1;
  });

  if (selectedNodes.length === 0) {
    return null;
  }

  var bboxNodes = selectedNodes.map(function (p) {
    if (p.shapeNode) {
      return {
        x0: p.x0,
        x1: p.x1,
        y0: p.y0,
        y1: p.y1
      };
    }

    var nodeSize = nodeSizeAccessor(p);
    return {
      x0: p.x - nodeSize,
      x1: p.x + nodeSize,
      y0: p.y - nodeSize,
      y1: p.y + nodeSize
    };
  });
  return (0, _baseRules.rectangleEnclosure)({
    bboxNodes: bboxNodes,
    d: d,
    i: i
  });
};

exports.svgRectEncloseRule = svgRectEncloseRule;

var svgHullEncloseRule = function svgHullEncloseRule(_ref6) {
  var d = _ref6.d,
      i = _ref6.i,
      projectedNodes = _ref6.projectedNodes,
      nodeIDAccessor = _ref6.nodeIDAccessor,
      nodeSizeAccessor = _ref6.nodeSizeAccessor;
  var selectedNodes = projectedNodes.filter(function (p) {
    return d.ids.indexOf(nodeIDAccessor(p)) !== -1;
  });

  if (selectedNodes.length === 0) {
    return null;
  }

  var projectedPoints = [];
  selectedNodes.forEach(function (p) {
    if (p.shapeNode) {
      projectedPoints.push({
        x: p.x0,
        y: p.y0
      });
      projectedPoints.push({
        x: p.x0,
        y: p.y1
      });
      projectedPoints.push({
        x: p.x1,
        y: p.y0
      });
      projectedPoints.push({
        x: p.x1,
        y: p.y1
      });
    } else {
      var nodeSize = nodeSizeAccessor(p);
      projectedPoints.push({
        x: p.x - nodeSize,
        y: p.y - nodeSize
      });
      projectedPoints.push({
        x: p.x + nodeSize,
        y: p.y - nodeSize
      });
      projectedPoints.push({
        x: p.x - nodeSize,
        y: p.y + nodeSize
      });
      projectedPoints.push({
        x: p.x + nodeSize,
        y: p.y + nodeSize
      });
    }
  });
  return (0, _baseRules.hullEnclosure)({
    points: projectedPoints.map(function (d) {
      return [d.x, d.y];
    }),
    d: d,
    i: i
  });
};

exports.svgHullEncloseRule = svgHullEncloseRule;

var svgHighlightRule = function svgHighlightRule(_ref7) {
  var d = _ref7.d,
      networkFrameRender = _ref7.networkFrameRender;
  var nodes = networkFrameRender.nodes;
  var customMark = nodes.customMark,
      baseStyle = nodes.styleFn;
  var styleFn = baseStyle;

  if (d.style && typeof d.style === "function") {
    styleFn = function styleFn(d) {
      return _objectSpread({}, baseStyle(d), d.style(d));
    };
  } else if (d.style) {
    styleFn = function styleFn(d) {
      return _objectSpread({}, baseStyle(d), d.style);
    };
  }

  var transform = "translate(".concat(d.x, ",").concat(d.y, ")");
  var baseMarkProps = {
    forceUpdate: true
  };
  var HighlightMark = customMark({
    d: d,
    styleFn: styleFn,
    transform: transform,
    baseMarkProps: baseMarkProps
  });
  return HighlightMark;
};

exports.svgHighlightRule = svgHighlightRule;
