// semiotic/lib/annotationRules/orframeRules.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.svgOrdinalLine = exports.svgRectEncloseRule = exports.htmlColumnHoverRule = exports.htmlFrameHoverRule = exports.svgCategoryRule = exports.svgRRule = exports.svgEncloseRule = exports.basicReactAnnotationRule = exports.svgORRule = exports.screenProject = exports.findIDPiece = exports.svgHighlightRule = exports.getColumnScreenCoordinates = void 0;

var React = _interopRequireWildcard(require("react"));

var _semioticMark = require("semiotic-mark");

var _Annotation = _interopRequireDefault(require("../Annotation"));

var _AnnotationCalloutCircle = _interopRequireDefault(require("react-annotation/lib/Types/AnnotationCalloutCircle"));

var _AnnotationBracket = _interopRequireDefault(require("react-annotation/lib/Types/AnnotationBracket"));

var _AnnotationXYThreshold = _interopRequireDefault(require("react-annotation/lib/Types/AnnotationXYThreshold"));

var _d3Hierarchy = require("d3-hierarchy");

var _d3Array = require("d3-array");

var _pieceDrawing = require("../svg/pieceDrawing");

var _baseRules = require("./baseRules");

var _SpanOrDiv = _interopRequireDefault(require("../SpanOrDiv"));

var _multiAccessorUtils = require("../data/multiAccessorUtils");

var _d3Shape = require("d3-shape");

var _general = require("../visualizationLayerBehavior/general");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function polarToCartesian(centerX, centerY, radius, angleInDegrees) {
  var angleInRadians = (angleInDegrees - 90) * Math.PI / 180.0;
  return {
    x: centerX + radius * Math.cos(angleInRadians),
    y: centerY + radius * Math.sin(angleInRadians)
  };
}

function pieContentGenerator(_ref) {
  var column = _ref.column,
      useSpans = _ref.useSpans;
  return React.createElement(_SpanOrDiv.default, {
    span: useSpans,
    className: "tooltip-content"
  }, React.createElement("p", {
    key: "or-annotation-1"
  }, column.name), React.createElement("p", {
    key: "or-annotation-2"
  }, "".concat((column.pct * 100).toFixed(0), "%")));
}

function arcBracket(_ref2) {
  var x = _ref2.x,
      y = _ref2.y,
      radius = _ref2.radius,
      startAngle = _ref2.startAngle,
      endAngle = _ref2.endAngle,
      inset = _ref2.inset,
      outset = _ref2.outset,
      _ref2$curly = _ref2.curly,
      curly = _ref2$curly === void 0 ? true : _ref2$curly;
  var start = polarToCartesian(x, y, radius + outset, endAngle);
  var end = polarToCartesian(x, y, radius + outset, startAngle);
  var innerStart = polarToCartesian(x, y, radius + outset - inset, endAngle);
  var innerEnd = polarToCartesian(x, y, radius + outset - inset, startAngle);
  var angleSize = endAngle - startAngle;
  var largeArcFlag = angleSize <= 180 ? "0" : "1";
  var d;

  if (curly) {
    var curlyOffset = Math.min(10, angleSize / 4);
    var middleLeft = polarToCartesian(x, y, radius + outset, (startAngle + endAngle) / 2 + curlyOffset);
    var middle = polarToCartesian(x, y, radius + outset + 10, (startAngle + endAngle) / 2);
    var middleRight = polarToCartesian(x, y, radius + outset, (startAngle + endAngle) / 2 - curlyOffset);
    d = ["M", innerStart.x, innerStart.y, "L", start.x, start.y, "A", radius + outset, radius + outset, 0, 0, 0, middleLeft.x, middleLeft.y, "A", radius + outset, radius + outset, 1, 0, 1, middle.x, middle.y, "A", radius + outset, radius + outset, 1, 0, 1, middleRight.x, middleRight.y, "A", radius + outset, radius + outset, 0, 0, 0, end.x, end.y, "L", innerEnd.x, innerEnd.y].join(" ");
  } else {
    d = ["M", innerStart.x, innerStart.y, "L", start.x, start.y, "A", radius + outset, radius + outset, 0, largeArcFlag, 0, end.x, end.y, "L", innerEnd.x, innerEnd.y].join(" ");
  }

  var midAngle = (startAngle + endAngle) / 2;
  var textOffset, largeTextArcFlag, finalTextEnd, finalTextStart, arcFlip;
  var lowerArc = midAngle > 90 && midAngle < 270;

  if (lowerArc) {
    textOffset = 12;
    largeTextArcFlag = endAngle - startAngle <= 180 ? "0" : "1";
    arcFlip = 0;
  } else {
    largeTextArcFlag = endAngle - startAngle <= 180 ? "0" : "1";
    textOffset = 5;
    arcFlip = 1;
  }

  textOffset += curly ? 10 : 0;
  var textStart = polarToCartesian(x, y, radius + outset + textOffset, endAngle);
  var textEnd = polarToCartesian(x, y, radius + outset + textOffset, startAngle);

  if (lowerArc) {
    finalTextStart = textStart;
    finalTextEnd = textEnd;
  } else {
    finalTextStart = textEnd;
    finalTextEnd = textStart;
  }

  var textD = ["M", finalTextStart.x, finalTextStart.y, "A", radius + outset + textOffset, radius + outset + textOffset, arcFlip, largeTextArcFlag, arcFlip, finalTextEnd.x, finalTextEnd.y].join(" ");
  return {
    arcPath: d,
    textArcPath: textD
  };
}

var getColumnScreenCoordinates = function getColumnScreenCoordinates(_ref3) {
  var d = _ref3.d,
      projectedColumns = _ref3.projectedColumns,
      oAccessor = _ref3.oAccessor,
      summaryType = _ref3.summaryType,
      type = _ref3.type,
      projection = _ref3.projection,
      adjustedPosition = _ref3.adjustedPosition,
      adjustedSize = _ref3.adjustedSize;
  var column = d.column || projectedColumns[d.facetColumn] || projectedColumns[(0, _multiAccessorUtils.findFirstAccessorValue)(oAccessor, d)];
  var pieces = column.pieceData || column.pieces;
  var positionValue = summaryType.type && summaryType.type !== "none" || ["swarm", "point", "clusterbar"].find(function (p) {
    return p === type.type;
  }) ? (0, _d3Array.max)(pieces.map(function (p) {
    return p.scaledValue;
  })) : projection === "horizontal" ? (0, _d3Array.max)(pieces.map(function (p) {
    return p.scaledValue + p.bottom;
  })) : (0, _d3Array.min)(pieces.map(function (p) {
    return p.bottom - p.scaledValue;
  }));
  var xPosition = column.middle + adjustedPosition[0];
  var yPosition = projection === "horizontal" ? adjustedSize[0] - positionValue : summaryType.type && summaryType.type !== "none" || ["swarm", "point", "clusterbar"].find(function (p) {
    return p === type.type;
  }) ? adjustedSize[1] - positionValue : positionValue;
  yPosition += 10;

  if (projection === "horizontal") {
    yPosition = column.middle;
    xPosition = positionValue + adjustedPosition[0];
  } else if (projection === "radial") {
    ;

    var _pointOnArcAtAngle = (0, _pieceDrawing.pointOnArcAtAngle)([d.arcAngles.translate[0], d.arcAngles.translate[1]], d.arcAngles.midAngle, d.arcAngles.length);

    var _pointOnArcAtAngle2 = _slicedToArray(_pointOnArcAtAngle, 2);

    xPosition = _pointOnArcAtAngle2[0];
    yPosition = _pointOnArcAtAngle2[1];
    yPosition += 10;
  }

  return {
    coordinates: [xPosition, yPosition],
    pieces: pieces,
    column: column
  };
};

exports.getColumnScreenCoordinates = getColumnScreenCoordinates;

var svgHighlightRule = function svgHighlightRule(_ref4) {
  var d = _ref4.d,
      pieceIDAccessor = _ref4.pieceIDAccessor,
      orFrameRender = _ref4.orFrameRender,
      oAccessor = _ref4.oAccessor;
  var thisID = pieceIDAccessor(d);
  var thisO = (0, _multiAccessorUtils.findFirstAccessorValue)(oAccessor, d);
  var pieces = orFrameRender.pieces;
  var styleFn = pieces.styleFn;
  var foundPieces = pieces && pieces.data.filter(function (p) {
    return (thisID === undefined || pieceIDAccessor(_objectSpread({}, p.piece, p.piece.data)) === thisID) && (thisO === undefined || (0, _multiAccessorUtils.findFirstAccessorValue)(oAccessor, p.piece.data) === thisO);
  }).map(function (p, q) {
    var styleObject = {
      style: styleFn(_objectSpread({}, p.piece, p.piece.data))
    };

    if (d.style && typeof d.style === "function") {
      styleObject = {
        style: _objectSpread({}, styleObject, d.style(_objectSpread({}, p.piece, p.piece.data)))
      };
    } else if (d.style) {
      styleObject = {
        style: _objectSpread({}, styleObject, d.style)
      };
    }

    var styledD = _objectSpread({}, p.renderElement, styleObject);

    var className = "highlight-annotation ".concat(d.class && typeof d.class === "function" && d.class(p.piece.data, q) || d.class && d.class || "");

    if (React.isValidElement(p.renderElement)) {
      return React.cloneElement(p.renderElement, _objectSpread({}, styleObject, {
        className: className
      }));
    }

    return React.createElement(_semioticMark.Mark, _extends({
      fill: "none",
      stroke: "black",
      strokeWidth: "2px",
      key: "highlight-piece-".concat(q)
    }, styledD, {
      className: className
    }));
  }) || [];
  /*
  const foundSummaries =
    (orFrameRender.summaries &&
      orFrameRender.summaries.data
        .filter(p => {
          return oAccessor(p.piece.data) === thisID
        })
        .map(p => {
          if (React.isValidElement(p.renderElement)) {
            console.error(
              "OrdinalFrame highlighting currently only works with built-in pieces and not custom pieces"
            )
            return null
          }
          let styleObject = {
            style: { fill: "none", stroke: "black", strokeWidth: "2px" }
          }
          if (d.style && typeof d.style === "function") {
            styleObject = { style: d.style(p.piece.data) }
          } else if (d.style) {
            styleObject = { style: d.style }
          }
          const styledD = { ...p.renderElement, ...styleObject }
           return <Mark {...styledD} />
        })) ||
    []
  */

  return _toConsumableArray(foundPieces);
};

exports.svgHighlightRule = svgHighlightRule;

var findIDPiece = function findIDPiece(pieceIDAccessor, oColumn, d) {
  var pieceID = pieceIDAccessor(d);
  if (pieceID === "") return d;
  var basePiece = pieceIDAccessor(d) && oColumn && oColumn.pieceData.find(function (r) {
    return pieceIDAccessor(r.data) === pieceIDAccessor(d);
  });
  var reactAnnotationProps = ["type", "label", "note", "connector", "disabled", "color", "subject"];

  if (basePiece) {
    reactAnnotationProps.forEach(function (prop) {
      if (d[prop]) basePiece[prop] = d[prop];
    });
  }

  return basePiece;
};

exports.findIDPiece = findIDPiece;

var screenProject = function screenProject(_ref5) {
  var p = _ref5.p,
      adjustedSize = _ref5.adjustedSize,
      rScale = _ref5.rScale,
      oColumn = _ref5.oColumn,
      rAccessor = _ref5.rAccessor,
      idPiece = _ref5.idPiece,
      projection = _ref5.projection,
      rScaleType = _ref5.rScaleType;
  var pValue = (0, _multiAccessorUtils.findFirstAccessorValue)(rAccessor, p) || p.value;
  var o;

  if (oColumn) {
    o = oColumn.middle;
  } else {
    o = 0;
  }

  if (oColumn && projection === "radial") {
    return (0, _pieceDrawing.pointOnArcAtAngle)([adjustedSize[0] / 2, adjustedSize[1] / 2], oColumn.pct_middle, idPiece && (idPiece.x || idPiece.scaledValue) ? idPiece.x / 2 || (idPiece.bottom + idPiece.scaledValue / 2) / 2 : pValue / 2);
  }

  if (projection === "horizontal") {
    return [idPiece && (idPiece.x || idPiece.scaledValue) ? idPiece.x === undefined ? idPiece.bottom + idPiece.scaledValue / 2 : idPiece.x : rScale(pValue), o];
  }

  var newScale = rScaleType.copy().domain(rScale.domain()).range(rScale.range().reverse());
  return [o, idPiece && (idPiece.x || idPiece.scaledValue) ? idPiece.y === undefined ? idPiece.bottom - idPiece.scaledValue / 2 : idPiece.y : newScale(pValue)];
};

exports.screenProject = screenProject;

var svgORRule = function svgORRule(_ref6) {
  var d = _ref6.d,
      i = _ref6.i,
      screenCoordinates = _ref6.screenCoordinates,
      projection = _ref6.projection;
  return React.createElement(_semioticMark.Mark, {
    markType: "text",
    key: "".concat(d.label, "annotationtext").concat(i),
    forceUpdate: true,
    x: screenCoordinates[0] + (projection === "horizontal" ? 10 : 0),
    y: screenCoordinates[1] + (projection === "vertical" ? 10 : 0),
    className: "annotation annotation-or-label ".concat(d.className || ""),
    textAnchor: "middle"
  }, d.label);
};

exports.svgORRule = svgORRule;

var basicReactAnnotationRule = function basicReactAnnotationRule(_ref7) {
  var d = _ref7.d,
      i = _ref7.i,
      screenCoordinates = _ref7.screenCoordinates;

  var noteData = _extends({
    dx: 0,
    dy: 0,
    note: {
      label: d.label,
      orientation: d.orientation,
      align: d.align
    },
    connector: {
      end: "arrow"
    }
  }, d, {
    x: screenCoordinates[0],
    y: screenCoordinates[1],
    type: typeof d.type === "function" ? d.type : undefined,
    screenCoordinates: screenCoordinates
  });

  if (d.fixedX) noteData.x = d.fixedX;
  if (d.fixedY) noteData.y = d.fixedY;
  return React.createElement(_Annotation.default, {
    key: d.key || "annotation-".concat(i),
    noteData: noteData
  });
};

exports.basicReactAnnotationRule = basicReactAnnotationRule;

var svgEncloseRule = function svgEncloseRule(_ref8) {
  var d = _ref8.d,
      i = _ref8.i,
      screenCoordinates = _ref8.screenCoordinates;
  var circle = (0, _d3Hierarchy.packEnclose)(screenCoordinates.map(function (p) {
    return {
      x: p[0],
      y: p[1],
      r: 2
    };
  }));
  return (0, _baseRules.circleEnclosure)({
    d: d,
    i: i,
    circle: circle
  });
};

exports.svgEncloseRule = svgEncloseRule;

var svgRRule = function svgRRule(_ref9) {
  var d = _ref9.d,
      i = _ref9.i,
      screenCoordinates = _ref9.screenCoordinates,
      rScale = _ref9.rScale,
      rAccessor = _ref9.rAccessor,
      adjustedSize = _ref9.adjustedSize,
      adjustedPosition = _ref9.adjustedPosition,
      projection = _ref9.projection;
  var x, y, xPosition, yPosition, subject, dx, dy;

  if (projection === "radial") {
    return React.createElement(_Annotation.default, {
      key: d.key || "annotation-".concat(i),
      noteData: _extends({
        dx: 50,
        dy: 50,
        note: {
          label: d.label
        },
        connector: {
          end: "arrow"
        }
      }, d, {
        type: _AnnotationCalloutCircle.default,
        subject: {
          radius: rScale((0, _multiAccessorUtils.findFirstAccessorValue)(rAccessor, d)) / 2,
          radiusPadding: 0
        },
        x: adjustedSize[0] / 2,
        y: adjustedSize[1] / 2
      })
    });
  } else if (projection === "horizontal") {
    dx = 50;
    dy = 50;
    yPosition = d.offset || i * 25;
    x = screenCoordinates[0];
    y = yPosition;
    subject = {
      x: x,
      y1: 0,
      y2: adjustedSize[1] + adjustedPosition[1]
    };
  } else {
    dx = 50;
    dy = -20;
    xPosition = d.offset || i * 25;
    y = screenCoordinates[1];
    x = xPosition;
    subject = {
      y: y,
      x1: 0,
      x2: adjustedSize[0] + adjustedPosition[0]
    };
  }

  var noteData = _extends({
    dx: dx,
    dy: dy,
    note: {
      label: d.label
    },
    connector: {
      end: "arrow"
    }
  }, d, {
    type: _AnnotationXYThreshold.default,
    x: x,
    y: y,
    subject: subject
  });

  return React.createElement(_Annotation.default, {
    key: d.key || "annotation-".concat(i),
    noteData: noteData
  });
};

exports.svgRRule = svgRRule;

var svgCategoryRule = function svgCategoryRule(_ref10) {
  var projection = _ref10.projection,
      d = _ref10.d,
      i = _ref10.i,
      categories = _ref10.categories,
      adjustedSize = _ref10.adjustedSize;
  var _d$bracketType = d.bracketType,
      bracketType = _d$bracketType === void 0 ? "curly" : _d$bracketType,
      _d$position = d.position,
      position = _d$position === void 0 ? projection === "vertical" ? "top" : "left" : _d$position,
      _d$depth = d.depth,
      depth = _d$depth === void 0 ? 30 : _d$depth,
      _d$offset = d.offset,
      offset = _d$offset === void 0 ? 0 : _d$offset,
      _d$padding = d.padding,
      padding = _d$padding === void 0 ? 0 : _d$padding;
  var actualCategories = Array.isArray(d.categories) ? d.categories : [d.categories];
  var cats = actualCategories.map(function (c) {
    return categories[c];
  });

  if (projection === "radial") {
    var arcPadding = padding / adjustedSize[1];
    var leftX = (0, _d3Array.min)(cats.map(function (p) {
      return p.pct_start + p.pct_padding / 2 + arcPadding / 2;
    }));
    var rightX = (0, _d3Array.max)(cats.map(function (p) {
      return p.pct_start + p.pct - p.pct_padding / 2 - arcPadding / 2;
    }));
    var chartSize = Math.min(adjustedSize[0], adjustedSize[1]) / 2;
    var centerX = adjustedSize[0] / 2;
    var centerY = adjustedSize[1] / 2;

    var _arcBracket = arcBracket({
      x: 0,
      y: 0,
      radius: chartSize,
      startAngle: leftX * 360,
      endAngle: rightX * 360,
      inset: depth,
      outset: offset,
      curly: bracketType === "curly"
    }),
        arcPath = _arcBracket.arcPath,
        textArcPath = _arcBracket.textArcPath;

    var textPathID = "text-path-".concat(i, "-").concat(Math.random());
    return React.createElement("g", {
      className: "category-annotation annotation",
      transform: "translate(".concat(centerX, ",").concat(centerY, ")")
    }, React.createElement("path", {
      d: arcPath,
      fill: "none",
      stroke: "black"
    }), React.createElement("path", {
      id: textPathID,
      d: textArcPath,
      style: {
        display: "none"
      }
    }), React.createElement("text", {
      "font-size": "12.5"
    }, React.createElement("textPath", {
      startOffset: "50%",
      textAnchor: "middle",
      xlinkHref: "#".concat(textPathID)
    }, d.label)));
  } else {
    var _leftX = (0, _d3Array.min)(cats.map(function (p) {
      return p.x;
    }));

    var _rightX = (0, _d3Array.max)(cats.map(function (p) {
      return p.x + p.width;
    }));

    if (projection === "vertical") {
      var yPosition = position === "top" ? 0 : adjustedSize[1];
      yPosition += position === "top" ? -offset : offset;
      var noteData = {
        type: _AnnotationBracket.default,
        y: yPosition,
        x: _leftX - padding,
        note: {
          title: d.title || d.label,
          label: d.title ? d.label : undefined
        },
        subject: {
          type: bracketType,
          width: _rightX - _leftX + padding * 2,
          depth: position === "top" ? -depth : depth
        }
      };
      return React.createElement(_Annotation.default, {
        key: d.key || "annotation-".concat(i),
        noteData: noteData
      });
    } else if (projection === "horizontal") {
      var _yPosition = position === "left" ? 0 : adjustedSize[0];

      _yPosition += position === "left" ? -offset : offset;
      var _noteData = {
        type: _AnnotationBracket.default,
        x: _yPosition,
        y: _leftX - padding,
        note: {
          title: d.title || d.label,
          label: d.title ? d.label : undefined
        },
        subject: {
          type: bracketType,
          height: _rightX - _leftX + padding * 2,
          depth: position === "left" ? -depth : depth
        }
      };
      return React.createElement(_Annotation.default, {
        key: d.key || "annotation-".concat(i),
        noteData: _noteData
      });
    }
  }
};

exports.svgCategoryRule = svgCategoryRule;

var htmlFrameHoverRule = function htmlFrameHoverRule(_ref11) {
  var d = _ref11.d,
      i = _ref11.i,
      rAccessor = _ref11.rAccessor,
      oAccessor = _ref11.oAccessor,
      projection = _ref11.projection,
      tooltipContent = _ref11.tooltipContent,
      useSpans = _ref11.useSpans,
      pieceIDAccessor = _ref11.pieceIDAccessor,
      projectedColumns = _ref11.projectedColumns,
      adjustedSize = _ref11.adjustedSize,
      rScale = _ref11.rScale,
      type = _ref11.type,
      rScaleType = _ref11.rScaleType;
  tooltipContent = tooltipContent === "pie" ? function () {
    return pieContentGenerator({
      column: d.column,
      useSpans: useSpans
    });
  } : tooltipContent; //To string because React gives a DOM error if it gets a date

  var contentFill;
  var pO = (0, _multiAccessorUtils.findFirstAccessorValue)(oAccessor, d) || d.column;
  var oColumn = projectedColumns[pO];
  var idPiece = findIDPiece(pieceIDAccessor, oColumn, d);

  if (!idPiece) {
    return null;
  }

  var screenCoordinates = (type.type === "clusterbar" || type.type === "point" || type.type === "swarm") && d.x !== undefined && d.y !== undefined || d.isSummaryData ? [d.x, d.y] : screenProject({
    p: d,
    adjustedSize: adjustedSize,
    rScale: rScale,
    oColumn: oColumn,
    rAccessor: rAccessor,
    idPiece: idPiece,
    projection: projection,
    rScaleType: rScaleType
  });

  if (d.isSummaryData) {
    var summaryContent = d.label;

    if (d.pieces && d.pieces.length !== 0) {
      if (d.pieces.length === 1) {
        summaryContent = [];
        rAccessor.forEach(function (actualRAccessor) {
          summaryContent.push(actualRAccessor(d.pieces[0].data));
        });
      } else {
        summaryContent = [];
        rAccessor.forEach(function (actualRAccessor) {
          var pieceData = (0, _d3Array.extent)(d.pieces.map(function (p) {
            return p.data;
          }).map(actualRAccessor));
          summaryContent.push("From ".concat(pieceData[0], " to ").concat(pieceData[1]));
        });
      }
    }

    var summaryLabel = React.createElement("p", {
      key: "html-annotation-content-2"
    }, summaryContent);
    contentFill = [React.createElement("p", {
      key: "html-annotation-content-1"
    }, d.key), summaryLabel, React.createElement("p", {
      key: "html-annotation-content-3"
    }, d.value)];
  } else if (d.data) {
    contentFill = [];
    oAccessor.forEach(function (actualOAccessor, i) {
      if (actualOAccessor(idPiece.data)) contentFill.push(React.createElement("p", {
        key: "html-annotation-content-o-".concat(i)
      }, actualOAccessor(idPiece.data).toString()));
    });
    rAccessor.forEach(function (actualRAccessor, i) {
      if (actualRAccessor(idPiece.data)) contentFill.push(React.createElement("p", {
        key: "html-annotation-content-r-".concat(i)
      }, actualRAccessor(idPiece.data).toString()));
    });
  } else if (d.label) {
    contentFill = d.label;
  }

  var content = React.createElement(_SpanOrDiv.default, {
    span: useSpans,
    className: "tooltip-content"
  }, contentFill);

  if (d.type === "frame-hover" && tooltipContent) {
    content = tooltipContent(idPiece);
  }

  return React.createElement(_SpanOrDiv.default, {
    span: useSpans,
    key: "xylabel-".concat(i),
    className: "annotation annotation-or-label ".concat(projection, " ").concat(d.className || ""),
    style: {
      position: "absolute",
      top: "".concat(screenCoordinates[1], "px"),
      left: "".concat(screenCoordinates[0], "px")
    }
  }, content);
};

exports.htmlFrameHoverRule = htmlFrameHoverRule;

var htmlColumnHoverRule = function htmlColumnHoverRule(_ref12) {
  var d = _ref12.d,
      i = _ref12.i,
      summaryType = _ref12.summaryType,
      oAccessor = _ref12.oAccessor,
      type = _ref12.type,
      adjustedPosition = _ref12.adjustedPosition,
      adjustedSize = _ref12.adjustedSize,
      projection = _ref12.projection,
      tooltipContent = _ref12.tooltipContent,
      useSpans = _ref12.useSpans,
      projectedColumns = _ref12.projectedColumns;

  //we need to ignore negative pieces to make sure the hover behavior populates on top of the positive bar
  var _getColumnScreenCoord = getColumnScreenCoordinates({
    d: d,
    projectedColumns: projectedColumns,
    oAccessor: oAccessor,
    summaryType: summaryType,
    type: type,
    projection: projection,
    adjustedPosition: adjustedPosition,
    adjustedSize: adjustedSize
  }),
      _getColumnScreenCoord2 = _slicedToArray(_getColumnScreenCoord.coordinates, 2),
      xPosition = _getColumnScreenCoord2[0],
      yPosition = _getColumnScreenCoord2[1],
      pieces = _getColumnScreenCoord.pieces,
      column = _getColumnScreenCoord.column; //To string because React gives a DOM error if it gets a date


  var oContent = [];
  oAccessor.forEach(function (actualOAccessor, i) {
    if (pieces[0].data) oContent.push(React.createElement("p", {
      key: "or-annotation-o-".concat(i)
    }, actualOAccessor(pieces[0].data).toString()));
  });
  var content = React.createElement(_SpanOrDiv.default, {
    span: useSpans,
    className: "tooltip-content"
  }, oContent, React.createElement("p", {
    key: "or-annotation-2"
  }, (0, _d3Array.sum)(pieces.map(function (p) {
    return p.value;
  }).filter(function (p) {
    return p > 0;
  }))));

  if (d.type === "column-hover" && tooltipContent) {
    if (tooltipContent === "pie") {
      tooltipContent = pieContentGenerator;
    }

    content = tooltipContent(_objectSpread({}, d, {
      pieces: pieces.map(function (p) {
        return p.data;
      }),
      column: column,
      oAccessor: oAccessor
    }));
  } else if (d.label) {
    content = React.createElement(_SpanOrDiv.default, {
      span: useSpans,
      className: "tooltip-content"
    }, d.label);
  }

  return React.createElement(_SpanOrDiv.default, {
    span: useSpans,
    key: "orlabel-".concat(i),
    className: "annotation annotation-or-label ".concat(projection, " ").concat(d.className || ""),
    style: {
      position: "absolute",
      top: "".concat(yPosition, "px"),
      left: "".concat(xPosition, "px")
    }
  }, content);
};

exports.htmlColumnHoverRule = htmlColumnHoverRule;

var svgRectEncloseRule = function svgRectEncloseRule(_ref13) {
  var d = _ref13.d,
      i = _ref13.i,
      screenCoordinates = _ref13.screenCoordinates;
  var bboxNodes = screenCoordinates.map(function (p) {
    return {
      x0: p.x0 = p[0],
      x1: p.x1 = p[0],
      y0: p.y0 = p[1],
      y1: p.y1 = p[1]
    };
  });
  return (0, _baseRules.rectangleEnclosure)({
    bboxNodes: bboxNodes,
    d: d,
    i: i
  });
};

exports.svgRectEncloseRule = svgRectEncloseRule;

var svgOrdinalLine = function svgOrdinalLine(_ref14) {
  var screenCoordinates = _ref14.screenCoordinates,
      d = _ref14.d,
      voronoiHover = _ref14.voronoiHover;
  var lineGenerator = (0, _d3Shape.line)().x(function (d) {
    return d[0];
  }).y(function (d) {
    return d[1];
  });

  if (d.curve) {
    var interpolator = _general.curveHash[d.curve] || d.curve;
    lineGenerator.curve(interpolator);
  }

  var lineStyle = typeof d.lineStyle === "function" ? d.lineStyle(d) : d.lineStyle || {};
  return React.createElement("g", null, React.createElement("path", {
    stroke: "black",
    fill: "none",
    style: lineStyle,
    d: lineGenerator(screenCoordinates)
  }), (d.points || d.interactive) && screenCoordinates.map(function (p, q) {
    var pointStyle = typeof d.pointStyle === "function" ? d.pointStyle(d.coordinates[q], q) : d.pointStyle || {};
    return React.createElement("g", {
      transform: "translate(".concat(p[0], ",").concat(p[1], ")"),
      key: "ordinal-line-point-".concat(q)
    }, d.points && React.createElement("circle", {
      style: pointStyle,
      r: d.radius || 5,
      fill: "black"
    }), d.interactive && React.createElement("circle", {
      style: {
        pointerEvents: "all"
      },
      r: d.hoverRadius || 15,
      fill: "pink",
      opacity: 0,
      onMouseEnter: function onMouseEnter() {
        return voronoiHover(_objectSpread({
          type: "frame-hover"
        }, d.coordinates[q], {
          data: d.coordinates[q]
        }));
      },
      onMouseOut: function onMouseOut() {
        return voronoiHover();
      }
    }), "}");
  }));
};

exports.svgOrdinalLine = svgOrdinalLine;
