// semiotic/lib/annotationRules/xyframeRules.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.svgHullEncloseAnnotation = exports.svgEncloseAnnotation = exports.svgRectEncloseAnnotation = exports.htmlTooltipAnnotation = exports.svgAreaAnnotation = exports.svgLineAnnotation = exports.svgBoundsAnnotation = exports.svgYAnnotation = exports.svgXAnnotation = exports.basicReactAnnotation = exports.svgXYAnnotation = exports.svgHighlight = exports.svgVerticalPointsAnnotation = exports.svgHorizontalPointsAnnotation = void 0;

var React = _interopRequireWildcard(require("react"));

var _semioticMark = require("semiotic-mark");

var _Annotation = _interopRequireDefault(require("../Annotation"));

var _AnnotationCalloutRect = _interopRequireDefault(require("react-annotation/lib/Types/AnnotationCalloutRect"));

var _AnnotationXYThreshold = _interopRequireDefault(require("react-annotation/lib/Types/AnnotationXYThreshold"));

var _d3Shape = require("d3-shape");

var _d3Hierarchy = require("d3-hierarchy");

var _d3Array = require("d3-array");

var _baseRules = require("./baseRules");

var _SpanOrDiv = _interopRequireDefault(require("../SpanOrDiv"));

var _multiAccessorUtils = require("../data/multiAccessorUtils");

var _general = require("../visualizationLayerBehavior/general");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

var pointsAlong = function pointsAlong(along) {
  return function (_ref) {
    var d = _ref.d,
        lines = _ref.lines,
        points = _ref.points,
        xScale = _ref.xScale,
        yScale = _ref.yScale,
        pointStyle = _ref.pointStyle;
    var alongScale = along === "x" ? xScale : yScale;
    along = along === "yTop" && d.yMiddle ? "yMiddle" : along;

    if (d && d[along]) {
      var _d$threshold = d.threshold,
          threshold = _d$threshold === void 0 ? 1 : _d$threshold,
          _d$r = d.r,
          r = _d$r === void 0 ? function () {
        return 4;
      } : _d$r,
          _d$styleFn = d.styleFn,
          styleFn = _d$styleFn === void 0 ? pointStyle : _d$styleFn;
      var foundPoints = [];
      var halfThreshold = threshold / 2;

      if (lines && lines.length > 0) {
        lines.forEach(function (linedata) {
          var linePoints = linedata.data.filter(function (p) {
            var pAlong = alongScale(p[along]);
            var dAlong = alongScale(d[along]);
            return pAlong <= dAlong + halfThreshold && pAlong >= dAlong - halfThreshold;
          });
          foundPoints.push.apply(foundPoints, _toConsumableArray(linePoints));
        });
      }

      if (points && points.length > 0) {
        var pointPoints = points.filter(function (p) {
          var pAlong = alongScale(p[along]);
          var dAlong = alongScale(d[along]);
          return pAlong <= dAlong + halfThreshold && pAlong >= dAlong - halfThreshold;
        });
        foundPoints.push.apply(foundPoints, _toConsumableArray(pointPoints));
      }

      return foundPoints.map(function (p, i) {
        return React.createElement("circle", {
          key: "found-circle-".concat(i),
          r: r(p, i),
          style: styleFn(p, i),
          cx: xScale(p.xMiddle || p.x),
          cy: yScale(p.yMiddle || p.yTop)
        });
      });
    }

    return null;
  };
};

var svgHorizontalPointsAnnotation = pointsAlong("yTop");
exports.svgHorizontalPointsAnnotation = svgHorizontalPointsAnnotation;
var svgVerticalPointsAnnotation = pointsAlong("x");
exports.svgVerticalPointsAnnotation = svgVerticalPointsAnnotation;

var svgHighlight = function svgHighlight(_ref2) {
  var d = _ref2.d,
      i = _ref2.i,
      _ref2$points = _ref2.points,
      points = _ref2$points === void 0 ? {
    data: []
  } : _ref2$points,
      _ref2$lines = _ref2.lines,
      lines = _ref2$lines === void 0 ? {
    data: [],
    type: {}
  } : _ref2$lines,
      _ref2$summaries = _ref2.summaries,
      summaries = _ref2$summaries === void 0 ? {
    data: []
  } : _ref2$summaries,
      idAccessor = _ref2.idAccessor,
      xScale = _ref2.xScale,
      yScale = _ref2.yScale,
      xyFrameRender = _ref2.xyFrameRender;
  var dID;
  var baseID = idAccessor(_objectSpread({}, d, d.data), i);

  if (baseID !== undefined) {
    dID = baseID;
  } else if (d.parentLine && idAccessor(d.parentLine, i) !== undefined) {
    dID = idAccessor(d.parentLine, i);
  } else if (d.parentSummary && idAccessor(d.parentSummary, i) !== undefined) {
    dID = idAccessor(d.parentSummary, i);
  }

  var foundPoints = points.data.filter(function (p, q) {
    return idAccessor(_objectSpread({}, p, p.data), q) === dID;
  }).map(function (p, q) {
    var baseStyle = xyFrameRender.points.styleFn(_objectSpread({}, p, p.data));
    var highlightStyle = typeof d.style === "function" ? d.style(_objectSpread({}, p, p.data), q) : d.style || {};
    return React.createElement("circle", {
      key: "highlight-point-".concat(q),
      cx: xScale(p.x),
      cy: yScale(p.y),
      r: 5,
      fill: "none",
      stroke: "black",
      strokeWidth: 2,
      style: _objectSpread({}, baseStyle, highlightStyle),
      className: "highlight-annotation ".concat(d.class && typeof d.class === "function" && d.class(_objectSpread({}, p, p.data), q) || d.class && d.class || "")
    });
  });
  var lineGenerator = (0, _d3Shape.area)().x(function (p) {
    return xScale(p.x);
  }).y0(function (p) {
    return yScale(p.yBottom);
  }).y1(function (p) {
    return yScale(p.yTop);
  });
  var interpolatorSetting = lines.type.interpolator || lines.type.curve;
  var actualInterpolator = typeof interpolatorSetting === "string" ? _general.curveHash[interpolatorSetting] : interpolatorSetting;

  if (actualInterpolator) {
    lineGenerator.curve(actualInterpolator);
  }

  var foundLines = lines.data.filter(function (p, q) {
    return idAccessor(p, q) === dID;
  }).map(function (p, q) {
    var baseStyle = xyFrameRender.lines.styleFn(p, q);
    var highlightStyle = typeof d.style === "function" ? d.style(p, q) : d.style || {};
    return React.createElement("path", {
      className: "highlight-annotation ".concat(d.class && typeof d.class === "function" && d.class(p, q) || d.class && d.class || ""),
      key: "highlight-summary-".concat(q),
      d: lineGenerator(p.data),
      fill: "none",
      stroke: "black",
      strokeWidth: 1,
      style: _objectSpread({}, baseStyle, highlightStyle)
    });
  });
  var foundSummaries = summaries.data.filter(function (p, q) {
    return idAccessor(p, q) === dID;
  }).map(function (p, q) {
    var baseStyle = xyFrameRender.summaries.styleFn(p, q);
    var highlightStyle = typeof d.style === "function" ? d.style(p, q) : d.style || {};
    return React.createElement("path", {
      className: "highlight-annotation ".concat(d.class && typeof d.class === "function" && d.class(p, q) || d.class && d.class || ""),
      key: "highlight-summary-".concat(q),
      d: "M".concat(p.coordinates.join("L")),
      fill: "none",
      stroke: "black",
      strokeWidth: 1,
      style: _objectSpread({}, baseStyle, highlightStyle)
    });
  });
  return [].concat(_toConsumableArray(foundSummaries), _toConsumableArray(foundLines), _toConsumableArray(foundPoints));
};

exports.svgHighlight = svgHighlight;

var svgXYAnnotation = function svgXYAnnotation(_ref3) {
  var screenCoordinates = _ref3.screenCoordinates,
      i = _ref3.i,
      d = _ref3.d;
  var inlineStyle;
  if (d.color) inlineStyle = {
    fill: d.color
  };
  var laLine = React.createElement(_semioticMark.Mark, {
    className: "annotation ".concat(d.type, " ").concat(d.className || "", " "),
    key: "annotationpoint".concat(i),
    markType: "circle",
    cx: screenCoordinates[0],
    cy: screenCoordinates[1],
    forceUpdate: true,
    style: inlineStyle,
    fill: "none",
    stroke: "black",
    r: 5
  });
  var laLabel;

  if (d.type === "xy") {
    laLabel = React.createElement(_semioticMark.Mark, {
      markType: "text",
      key: "".concat(d.label, "annotationtext").concat(i),
      forceUpdate: true,
      x: screenCoordinates[0],
      y: 10 + screenCoordinates[1],
      className: "annotation annotation-xy-label ".concat(d.className || "", " ")
    }, d.label);
  }

  return [laLine, laLabel];
};

exports.svgXYAnnotation = svgXYAnnotation;

var basicReactAnnotation = function basicReactAnnotation(_ref4) {
  var screenCoordinates = _ref4.screenCoordinates,
      d = _ref4.d,
      i = _ref4.i;

  var noteData = _extends({
    dx: 0,
    dy: 0,
    note: {
      label: d.label,
      orientation: d.orientation,
      align: d.align
    },
    connector: {
      end: "arrow"
    }
  }, d, {
    type: d.type,
    screenCoordinates: screenCoordinates,
    i: i
  });

  noteData.x = noteData.fixedX ? noteData.fixedX : screenCoordinates[0];
  noteData.y = noteData.fixedY ? noteData.fixedY : screenCoordinates[1];
  return React.createElement(_Annotation.default, {
    key: d.key || "annotation-".concat(i),
    noteData: noteData
  });
};

exports.basicReactAnnotation = basicReactAnnotation;

var svgXAnnotation = function svgXAnnotation(_ref5) {
  var screenCoordinates = _ref5.screenCoordinates,
      d = _ref5.d,
      i = _ref5.i,
      adjustedSize = _ref5.adjustedSize;

  var noteData = _extends({
    dx: 50,
    dy: 20,
    y: 0,
    note: {
      label: d.label
    },
    connector: {
      end: "arrow"
    }
  }, d, {
    type: _AnnotationXYThreshold.default,
    x: screenCoordinates[0],
    subject: {
      x: screenCoordinates[0],
      y1: 0,
      y2: adjustedSize[1]
    },
    i: i
  });

  return React.createElement(_Annotation.default, {
    key: d.key || "annotation-".concat(i),
    noteData: noteData
  });
};

exports.svgXAnnotation = svgXAnnotation;

var svgYAnnotation = function svgYAnnotation(_ref6) {
  var screenCoordinates = _ref6.screenCoordinates,
      d = _ref6.d,
      i = _ref6.i,
      adjustedSize = _ref6.adjustedSize,
      adjustedPosition = _ref6.adjustedPosition;
  var xPosition = i * 25;

  var noteData = _extends({
    dx: 50,
    dy: -20,
    x: xPosition,
    note: {
      label: d.label
    },
    connector: {
      end: "arrow"
    }
  }, d, {
    type: _AnnotationXYThreshold.default,
    y: screenCoordinates[1],
    subject: {
      y: screenCoordinates[1],
      x1: 0,
      x2: adjustedSize[0] + adjustedPosition[0]
    },
    i: i
  });

  return React.createElement(_Annotation.default, {
    key: d.key || "annotation-".concat(i),
    noteData: noteData
  });
};

exports.svgYAnnotation = svgYAnnotation;

var svgBoundsAnnotation = function svgBoundsAnnotation(_ref7) {
  var d = _ref7.d,
      i = _ref7.i,
      adjustedSize = _ref7.adjustedSize,
      xAccessor = _ref7.xAccessor,
      yAccessor = _ref7.yAccessor,
      xScale = _ref7.xScale,
      yScale = _ref7.yScale;
  var startXValue = (0, _multiAccessorUtils.findFirstAccessorValue)(xAccessor, d.bounds[0]);
  var startYValue = (0, _multiAccessorUtils.findFirstAccessorValue)(yAccessor, d.bounds[0]);
  var endXValue = (0, _multiAccessorUtils.findFirstAccessorValue)(xAccessor, d.bounds[1]);
  var endYValue = (0, _multiAccessorUtils.findFirstAccessorValue)(yAccessor, d.bounds[1]);
  var x0Position = startXValue ? xScale(startXValue) : 0;
  var y0Position = startYValue ? yScale(startYValue) : adjustedSize[1];
  var x1Position = endXValue ? xScale(endXValue) : adjustedSize[0];
  var y1Position = endYValue ? yScale(endYValue) : 0;

  var noteData = _extends({
    dx: 250,
    dy: -20,
    note: {
      label: d.label
    },
    connector: {
      end: "arrow"
    }
  }, d, {
    type: _AnnotationCalloutRect.default,
    x: Math.min(x0Position, x1Position),
    y: Math.min(y0Position, y1Position),
    subject: {
      width: Math.abs(x1Position - x0Position),
      height: Math.abs(y0Position - y1Position)
    },
    i: i
  });

  return React.createElement(_Annotation.default, {
    key: d.key || "annotation-".concat(i),
    noteData: noteData
  });
};

exports.svgBoundsAnnotation = svgBoundsAnnotation;

var svgLineAnnotation = function svgLineAnnotation(_ref8) {
  var d = _ref8.d,
      i = _ref8.i,
      screenCoordinates = _ref8.screenCoordinates;
  var lineGenerator = (0, _d3Shape.line)().x(function (p) {
    return p[0];
  }).y(function (p) {
    return p[1];
  });
  var lineD = lineGenerator(screenCoordinates);
  var laLine = React.createElement(_semioticMark.Mark, {
    key: "".concat(d.label, "annotationline").concat(i),
    markType: "path",
    d: lineD,
    className: "annotation annotation-line ".concat(d.className || "", " ")
  });
  var laLabel = React.createElement(_semioticMark.Mark, {
    markType: "text",
    key: "".concat(d.label, "annotationlinetext").concat(i),
    x: (screenCoordinates[0][0] + screenCoordinates[1][0]) / 2,
    y: (screenCoordinates[0][1] + screenCoordinates[1][1]) / 2,
    className: "annotation annotation-line-label ".concat(d.className || "", " ")
  }, d.label);
  return [laLine, laLabel];
};

exports.svgLineAnnotation = svgLineAnnotation;

var svgAreaAnnotation = function svgAreaAnnotation(_ref9) {
  var d = _ref9.d,
      i = _ref9.i,
      xScale = _ref9.xScale,
      xAccessor = _ref9.xAccessor,
      yScale = _ref9.yScale,
      yAccessor = _ref9.yAccessor,
      annotationLayer = _ref9.annotationLayer;
  var mappedCoordinates = "M".concat(d.coordinates.map(function (p) {
    return [xScale((0, _multiAccessorUtils.findFirstAccessorValue)(xAccessor, p)), yScale((0, _multiAccessorUtils.findFirstAccessorValue)(yAccessor, p))];
  }).join("L"), "Z");
  var xBounds = (0, _d3Array.extent)(d.coordinates.map(function (p) {
    return xScale((0, _multiAccessorUtils.findFirstAccessorValue)(xAccessor, p));
  }));
  var yBounds = (0, _d3Array.extent)(d.coordinates.map(function (p) {
    return yScale((0, _multiAccessorUtils.findFirstAccessorValue)(yAccessor, p));
  }));
  var xCenter = (xBounds[0] + xBounds[1]) / 2;
  var yCenter = (yBounds[0] + yBounds[1]) / 2;
  var laLine = React.createElement(_semioticMark.Mark, {
    key: "".concat(d.label, "-annotation-area-").concat(i),
    markType: "path",
    transform: "translate(".concat(annotationLayer.position, ")"),
    d: mappedCoordinates,
    className: "annotation annotation-area ".concat(d.className || "", " ")
  });
  var laLabel = React.createElement(_semioticMark.Mark, {
    markType: "text",
    key: "".concat(d.label, "-annotationtext-").concat(i),
    forceUpdate: true,
    x: xCenter,
    y: yCenter,
    transform: "translate(".concat(annotationLayer.position, ")"),
    className: "annotation annotation-area-label ".concat(d.className || "", " "),
    style: {
      textAnchor: "middle"
    }
  }, d.label);
  return [laLine, laLabel];
};

exports.svgAreaAnnotation = svgAreaAnnotation;

var htmlTooltipAnnotation = function htmlTooltipAnnotation(_ref10) {
  var content = _ref10.content,
      screenCoordinates = _ref10.screenCoordinates,
      i = _ref10.i,
      d = _ref10.d,
      useSpans = _ref10.useSpans;
  //To string because React gives a DOM error if it gets a date
  return React.createElement(_SpanOrDiv.default, {
    span: useSpans,
    key: "xylabel-".concat(i),
    className: "annotation annotation-xy-label ".concat(d.className || "", " "),
    style: {
      position: "absolute",
      top: "".concat(screenCoordinates[1], "px"),
      left: "".concat(screenCoordinates[0], "px")
    }
  }, content);
};

exports.htmlTooltipAnnotation = htmlTooltipAnnotation;

var svgRectEncloseAnnotation = function svgRectEncloseAnnotation(_ref11) {
  var d = _ref11.d,
      i = _ref11.i,
      screenCoordinates = _ref11.screenCoordinates;
  var bboxNodes = screenCoordinates.map(function (p) {
    return {
      x0: p.x0 = p[0],
      x1: p.x1 = p[0],
      y0: p.y0 = p[1],
      y1: p.y1 = p[1]
    };
  });
  return (0, _baseRules.rectangleEnclosure)({
    bboxNodes: bboxNodes,
    d: d,
    i: i
  });
};

exports.svgRectEncloseAnnotation = svgRectEncloseAnnotation;

var svgEncloseAnnotation = function svgEncloseAnnotation(_ref12) {
  var screenCoordinates = _ref12.screenCoordinates,
      d = _ref12.d,
      i = _ref12.i;
  var circle = (0, _d3Hierarchy.packEnclose)(screenCoordinates.map(function (p) {
    return {
      x: p[0],
      y: p[1],
      r: 2
    };
  }));
  return (0, _baseRules.circleEnclosure)({
    d: d,
    circle: circle,
    i: i
  });
};

exports.svgEncloseAnnotation = svgEncloseAnnotation;

var svgHullEncloseAnnotation = function svgHullEncloseAnnotation(_ref13) {
  var screenCoordinates = _ref13.screenCoordinates,
      d = _ref13.d,
      i = _ref13.i;
  return (0, _baseRules.hullEnclosure)({
    points: screenCoordinates,
    d: d,
    i: i
  });
};

exports.svgHullEncloseAnnotation = svgHullEncloseAnnotation;
