// semiotic/lib/constants/coordinateNames.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.projectedXBottom = exports.projectedXTop = exports.projectedXMiddle = exports.projectedYBottom = exports.projectedYTop = exports.projectedYMiddle = exports.projectedY = exports.projectedX = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var projectedX = "x";
exports.projectedX = projectedX;
var projectedY = "y";
exports.projectedY = projectedY;
var projectedYMiddle = "yMiddle";
exports.projectedYMiddle = projectedYMiddle;
var projectedYTop = "yTop";
exports.projectedYTop = projectedYTop;
var projectedYBottom = "yBottom";
exports.projectedYBottom = projectedYBottom;
var projectedXMiddle = "xMiddle";
exports.projectedXMiddle = projectedXMiddle;
var projectedXTop = "xTop";
exports.projectedXTop = projectedXTop;
var projectedXBottom = "xBottom";
exports.projectedXBottom = projectedXBottom;
