// semiotic/lib/visualizationLayerBehavior/general.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.lineGeneratorDecorator = lineGeneratorDecorator;
exports.createPoints = createPoints;
exports.createLines = createLines;
exports.createSummaries = createSummaries;
exports.clonedAppliedElement = clonedAppliedElement;
exports.curveHash = void 0;

var React = _interopRequireWildcard(require("react"));

var _semioticMark = require("semiotic-mark");

var _d3Shape = require("d3-shape");

var _areaDrawing = require("../svg/areaDrawing");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var curveHash = {
  step: _d3Shape.curveStep,
  stepbefore: _d3Shape.curveStepBefore,
  stepafter: _d3Shape.curveStepAfter,
  cardinal: _d3Shape.curveCardinal,
  basis: _d3Shape.curveBasis,
  linear: _d3Shape.curveLinear,
  catmullrom: _d3Shape.curveCatmullRom,
  monotone: _d3Shape.curveMonotoneY,
  monotonex: _d3Shape.curveMonotoneX,
  monotoney: _d3Shape.curveMonotoneY,
  natural: _d3Shape.curveNatural
};
exports.curveHash = curveHash;

function lineGeneratorDecorator(_ref) {
  var generator = _ref.generator,
      projectedCoordinateNames = _ref.projectedCoordinateNames,
      defined = _ref.defined,
      xScale = _ref.xScale,
      yScale = _ref.yScale,
      interpolator = _ref.interpolator,
      simpleLine = _ref.simpleLine;
  var x = projectedCoordinateNames.x,
      y = projectedCoordinateNames.y,
      yTop = projectedCoordinateNames.yTop,
      yBottom = projectedCoordinateNames.yBottom;
  generator.x(function (d) {
    return xScale(d[x]);
  }).curve(interpolator);

  if (simpleLine) {
    generator.y(function (d) {
      return yScale(d[y]);
    });
  } else {
    generator.y0(function (d) {
      return yScale(d[yBottom]);
    }).y1(function (d) {
      return yScale(d[yTop]);
    });
  }

  if (defined) {
    generator.defined(function (p, q) {
      return defined(p, q);
    });
  } else {
    generator.defined(function (p) {
      return !p._xyFrameUndefined;
    });
  }
}

function createPoints(_ref2) {
  var xScale = _ref2.xScale,
      yScale = _ref2.yScale,
      canvasDrawing = _ref2.canvasDrawing,
      data = _ref2.data,
      projectedCoordinateNames = _ref2.projectedCoordinateNames,
      customMark = _ref2.customMark,
      canvasRender = _ref2.canvasRender,
      styleFn = _ref2.styleFn,
      classFn = _ref2.classFn,
      renderKeyFn = _ref2.renderKeyFn,
      renderMode = _ref2.renderMode,
      baseMarkProps = _ref2.baseMarkProps,
      baseShowLinePoints = _ref2.showLinePoints;
  var y = projectedCoordinateNames.y,
      x = projectedCoordinateNames.x,
      xMiddle = projectedCoordinateNames.xMiddle,
      yMiddle = projectedCoordinateNames.yMiddle,
      yTop = projectedCoordinateNames.yTop,
      yBottom = projectedCoordinateNames.yBottom;
  var showLinePoints = baseShowLinePoints === true ? undefined : baseShowLinePoints;
  var whichPoints = {
    top: yTop,
    bottom: yBottom
  };
  var whichWay = whichPoints[showLinePoints];
  var mappedPoints = [];
  data.forEach(function (d, i) {
    var dX = xScale(d[xMiddle] !== undefined ? d[xMiddle] : d[x]);
    var dY = yScale(d[whichWay] !== undefined ? d[whichWay] : d[yMiddle] !== undefined ? d[yMiddle] : d[y]);
    var pointAriaLabel = "Point at x ".concat(d.x, " and y ").concat(d.y); // CUSTOM MARK IMPLEMENTATION

    var renderedCustomMark = !customMark ? undefined : React.isValidElement(customMark) ? customMark : customMark({
      d: d.data,
      xy: d,
      i: i,
      xScale: xScale,
      yScale: yScale
    });
    var markProps = customMark ? _extends(baseMarkProps, renderedCustomMark.props, {
      "aria-label": pointAriaLabel
    }) : _objectSpread({}, baseMarkProps, {
      key: "piece-".concat(i),
      markType: "circle",
      r: 2,
      "aria-label": pointAriaLabel
    });

    if (renderedCustomMark && renderedCustomMark.props && !renderedCustomMark.props.markType && (!canvasRender || canvasRender(d.data, i) !== true)) {
      mappedPoints.push(React.createElement("g", {
        transform: "translate(".concat(dX, ",").concat(dY, ")"),
        key: renderKeyFn ? renderKeyFn(d.data, i) : "custom-point-mark-".concat(i),
        style: styleFn ? styleFn(d.data, i) : {},
        className: classFn ? classFn(d.data, i) : ""
      }, renderedCustomMark));
    } else {
      if (canvasRender && canvasRender(d.data, i) === true) {
        var canvasPoint = {
          type: "point",
          baseClass: "frame-piece",
          tx: dX,
          ty: dY,
          d: d,
          i: i,
          markProps: markProps,
          styleFn: styleFn,
          renderFn: renderMode,
          classFn: classFn
        };
        canvasDrawing.push(canvasPoint);
      } else {
        var yCoordinates = Array.isArray(d[y]) ? d[y].map(function (p) {
          return yScale(p);
        }) : [dY];
        yCoordinates.forEach(function (yc, yi) {
          var xCoordinates = Array.isArray(d[x]) ? d[x].map(function (p) {
            return xScale(p);
          }) : [dX];
          xCoordinates.forEach(function (xc, xi) {
            mappedPoints.push(clonedAppliedElement({
              baseClass: "frame-piece",
              tx: xc,
              ty: yc,
              d: d.data && _objectSpread({}, d, d.data) || d,
              i: yi === 0 && xi === 0 ? i : "".concat(i, "-").concat(yi, "-").concat(xi),
              markProps: markProps,
              styleFn: styleFn,
              renderFn: renderMode,
              renderKeyFn: renderKeyFn,
              classFn: classFn,
              yi: yi
            }));
          });
        });
      }
    }
  });
  return mappedPoints;
}

function createLines(_ref3) {
  var xScale = _ref3.xScale,
      yScale = _ref3.yScale,
      canvasDrawing = _ref3.canvasDrawing,
      data = _ref3.data,
      projectedCoordinateNames = _ref3.projectedCoordinateNames,
      customMark = _ref3.customMark,
      canvasRender = _ref3.canvasRender,
      styleFn = _ref3.styleFn,
      classFn = _ref3.classFn,
      renderMode = _ref3.renderMode,
      renderKeyFn = _ref3.renderKeyFn,
      type = _ref3.type,
      defined = _ref3.defined,
      baseMarkProps = _ref3.baseMarkProps,
      ariaLabel = _ref3.ariaLabel,
      _ref3$axesData = _ref3.axesData,
      axesData = _ref3$axesData === void 0 ? [] : _ref3$axesData;
  var xAxis = axesData.find(function (d) {
    return d.orient === "bottom" || d.orient === "top";
  });
  var yAxis = axesData.find(function (d) {
    return d.orient === "left" || d.orient === "right";
  });

  var xAxisFormatter = xAxis && xAxis.tickFormat || function (d) {
    return d;
  };

  var yAxisFormatter = yAxis && yAxis.tickFormat || function (d) {
    return d;
  };

  var customLine = _typeof(type) === "object" ? type : {
    type: type
  };
  var interpolator = typeof customLine.interpolator === "string" ? curveHash[customLine.interpolator] : customLine.interpolator || _d3Shape.curveLinear;
  var lineGenerator = customLine.simpleLine ? (0, _d3Shape.line)() : (0, _d3Shape.area)();
  lineGeneratorDecorator({
    projectedCoordinateNames: projectedCoordinateNames,
    defined: defined,
    interpolator: interpolator,
    generator: lineGenerator,
    xScale: xScale,
    yScale: yScale,
    simpleLine: customLine.simpleLine
  });

  var dynamicLineGenerator = interpolator.dynamicInterpolator && function (d, i) {
    var dynLineGenerator = (0, _d3Shape.area)();
    lineGeneratorDecorator({
      projectedCoordinateNames: projectedCoordinateNames,
      defined: defined,
      interpolator: interpolator.dynamicInterpolator(d, i),
      generator: dynLineGenerator,
      xScale: xScale,
      yScale: yScale,
      simpleLine: customLine.simpleLine
    });
    return dynLineGenerator;
  } || function () {
    return lineGenerator;
  };

  var mappedLines = [];
  data.forEach(function (d, i) {
    if (customMark && typeof customMark === "function") {
      //shim to make customLineMark work until Semiotic 2
      var compatibleData = _objectSpread({}, d, {
        data: d.data.map(function (p) {
          return _objectSpread({}, p.data, p);
        })
      });

      mappedLines.push(customMark({
        d: compatibleData,
        i: i,
        xScale: xScale,
        yScale: yScale,
        canvasDrawing: canvasDrawing
      }));
    } else {
      var builtInDisplayProps = {};

      if (customLine.simpleLine) {
        builtInDisplayProps.fill = "none";
        builtInDisplayProps.stroke = "black";
      }

      var pathString = dynamicLineGenerator(d, i)(d.data.map(function (p) {
        return _extends({}, p.data, p);
      }));

      var markProps = _objectSpread({}, builtInDisplayProps, baseMarkProps, {
        markType: "path",
        d: pathString,
        "aria-label": d.data && d.data.length > 0 && "".concat(d.data.length, " point ").concat(ariaLabel.items, " starting value ").concat(yAxisFormatter(d.data[0].y), " at ").concat(xAxisFormatter(d.data[0].x), " ending value ").concat(yAxisFormatter(d.data[d.data.length - 1].y), " at ").concat(xAxisFormatter(d.data[d.data.length - 1].x))
      });

      if (canvasRender && canvasRender(d, i) === true) {
        var canvasLine = {
          type: "line",
          baseClass: "xyframe-line",
          tx: 0,
          ty: 0,
          d: d,
          i: i,
          markProps: markProps,
          styleFn: styleFn,
          renderFn: renderMode,
          classFn: classFn
        };
        canvasDrawing.push(canvasLine);
      } else {
        mappedLines.push(clonedAppliedElement({
          baseClass: "xyframe-line",
          d: d,
          i: i,
          markProps: markProps,
          styleFn: styleFn,
          renderFn: renderMode,
          renderKeyFn: renderKeyFn,
          classFn: classFn
        }));
      }
    }
  });

  if (customLine.type === "difference" && data.length === 2) {
    //Create the overlay line for the difference chart
    var diffdataA = data[0].data.map(function (basedata, baseI) {
      var linePoint = basedata.yTop > data[1].data[baseI].yTop ? basedata.yTop : basedata.yBottom;
      return {
        x: basedata.x,
        y: linePoint,
        yBottom: linePoint,
        yTop: linePoint
      };
    });
    var diffdataB = data[0].data.map(function (basedata, baseI) {
      var linePoint = data[1].data[baseI].yTop > basedata.yTop ? data[1].data[baseI].yTop : data[1].data[baseI].yBottom;
      return {
        x: basedata.x,
        y: linePoint,
        yBottom: linePoint,
        yTop: linePoint
      };
    });
    var doClassname = classFn ? "xyframe-line ".concat(classFn(diffdataA)) : "xyframe-line";
    var overLine = (0, _d3Shape.line)();
    lineGeneratorDecorator({
      projectedCoordinateNames: projectedCoordinateNames,
      defined: defined,
      interpolator: interpolator,
      generator: overLine,
      xScale: xScale,
      yScale: yScale,
      simpleLine: true
    }); //      let baseStyle = props.lineStyle ? props.lineStyle(diffdata, 0) : {}

    var diffOverlayA = React.createElement(_semioticMark.Mark, {
      key: "xyline-diff-a",
      className: "".concat(doClassname, " difference-overlay-a"),
      markType: "path",
      d: overLine(diffdataA),
      style: {
        fill: "none",
        pointerEvents: "none"
      }
    });
    mappedLines.push(diffOverlayA);
    var diffOverlayB = React.createElement(_semioticMark.Mark, {
      key: "xyline-diff-b",
      className: "".concat(doClassname, " difference-overlay-b"),
      markType: "path",
      d: overLine(diffdataB),
      style: {
        fill: "none",
        pointerEvents: "none"
      }
    });
    mappedLines.push(diffOverlayB);
  }

  return mappedLines;
}

function createSummaries(_ref4) {
  var xScale = _ref4.xScale,
      yScale = _ref4.yScale,
      canvasDrawing = _ref4.canvasDrawing,
      data = _ref4.data,
      canvasRender = _ref4.canvasRender,
      styleFn = _ref4.styleFn,
      classFn = _ref4.classFn,
      renderKeyFn = _ref4.renderKeyFn,
      renderMode = _ref4.renderMode,
      baseMarkProps = _ref4.baseMarkProps,
      customMark = _ref4.customMark;

  var summaryClass = classFn || function () {
    return "";
  };

  var summaryStyle = styleFn || function () {
    return {};
  };

  var renderFn = renderMode;

  if (!Array.isArray(data)) {
    data = [data];
  }

  var renderedSummaries = [];
  data.forEach(function (d, i) {
    var className = "xyframe-summary";

    if (summaryClass) {
      className = "xyframe-summary ".concat(summaryClass(d));
    }

    var drawD = "";
    var shouldBeValid = false;

    if (typeof d.customMark === "string" || React.isValidElement(d.customMark)) {
      drawD = d.customMark;
      shouldBeValid = true;
    } else if (d.type === "MultiPolygon") {
      var polycoords = d.coordinates;
      polycoords.forEach(function (coord) {
        coord.forEach(function (c) {
          drawD += "M".concat(c.map(function (p) {
            return "".concat(xScale(p[0]), ",").concat(yScale(p[1]));
          }).join("L"), "Z ");
        });
      });
    } else if (customMark) {
      var xyfCoords = d._xyfCoordinates;
      var projectedCoordinates = xyfCoords.map(function (p) {
        return [xScale(p[0]), yScale(p[1])];
      }); // CUSTOM MARK IMPLEMENTATION

      drawD = customMark({
        d: d,
        i: i,
        classFn: summaryClass,
        styleFn: summaryStyle,
        renderFn: renderFn,
        projectedCoordinates: projectedCoordinates,
        xScale: xScale,
        yScale: yScale,
        bounds: (0, _areaDrawing.shapeBounds)(projectedCoordinates)
      });
      shouldBeValid = true;
    } else {
      var _xyfCoords = d._xyfCoordinates;

      if (d.curve) {
        var lineDrawing = (0, _d3Shape.line)().x(function (d) {
          return xScale(d[0]);
        }).y(function (d) {
          return yScale(d[1]);
        }).curve(d.curve);
        drawD = lineDrawing(_xyfCoords);
      } else {
        drawD = "M".concat(_xyfCoords.map(function (p) {
          return "".concat(xScale(p[0]), ",").concat(yScale(p[1]));
        }).join("L"), "Z");
      }
    }

    var renderKey = renderKeyFn ? renderKeyFn(d, i) : "summary-".concat(i);

    if (shouldBeValid && React.isValidElement(drawD)) {
      renderedSummaries.push(drawD);
    } else if (canvasRender && canvasRender(d, i) === true) {
      var canvasSummary = {
        type: "summary",
        baseClass: "xyframe-summary",
        tx: 0,
        ty: 0,
        d: d,
        i: i,
        markProps: {
          markType: "path",
          d: drawD
        },
        styleFn: summaryStyle,
        renderFn: renderFn,
        classFn: function classFn() {
          return className;
        }
      };
      canvasDrawing.push(canvasSummary);
    } else {
      renderedSummaries.push(React.createElement(_semioticMark.Mark, _extends({}, baseMarkProps, {
        key: renderKey,
        forceUpdate: true,
        renderMode: renderFn ? renderFn(d, i) : undefined,
        className: className,
        markType: "path",
        d: drawD,
        style: summaryStyle(d, i)
      })));
    }
  });
  return renderedSummaries;
}

function clonedAppliedElement(_ref5) {
  var tx = _ref5.tx,
      ty = _ref5.ty,
      d = _ref5.d,
      i = _ref5.i,
      markProps = _ref5.markProps,
      styleFn = _ref5.styleFn,
      renderFn = _ref5.renderFn,
      classFn = _ref5.classFn,
      renderKeyFn = _ref5.renderKeyFn,
      baseClass = _ref5.baseClass,
      yi = _ref5.yi;
  markProps.style = styleFn ? styleFn(d, i, yi) : {};
  markProps.className = baseClass;
  markProps.key = renderKeyFn ? renderKeyFn(d, i, yi) : "".concat(baseClass, "-").concat(d.key === undefined ? i : d.key);

  if (tx || ty) {
    markProps.transform = "translate(".concat(tx || 0, ",").concat(ty || 0, ")");
  }

  if (classFn) {
    markProps.className = "".concat(baseClass, " ").concat(classFn(d, i, yi));
  }

  if (!markProps.markType) {
    var RenderableMark = markProps;
    return React.createElement(RenderableMark);
  }

  markProps.renderMode = renderFn ? renderFn(d, i, yi) : undefined;
  return React.createElement(_semioticMark.Mark, markProps);
}
