// semiotic/lib/AnnotationLayer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _annotationHandling = require("./annotationLayerBehavior/annotationHandling");

var _Legend = _interopRequireDefault(require("./Legend"));

var _Annotation = _interopRequireDefault(require("./Annotation"));

var labella = _interopRequireWildcard(require("labella"));

var _SpanOrDiv = _interopRequireDefault(require("./SpanOrDiv"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function marginOffsetFn(orient, axisSettings, marginOffset) {
  if (typeof marginOffset === "number") {
    return marginOffset;
  }

  if (axisSettings && axisSettings.find(function (d) {
    return d.props.orient === orient;
  })) {
    return 50;
  }

  return 10;
}

function adjustedAnnotationKeyMapper(d) {
  return d.props.noteData.id || "".concat(d.props.noteData.x, "-").concat(d.props.noteData.y);
}

function noteDataWidth(noteData) {
  var charWidth = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 8;
  var wrap = noteData.note && noteData.note.wrap || 120;
  var noteText = noteData.note.label || noteData.note.label || "";
  return Math.min(wrap, noteText.length * charWidth);
}

function noteDataHeight(noteData) {
  var charWidth = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 8;
  var lineHeight = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 20;
  var wrap = noteData.note && noteData.note.wrap || 120;
  var text = noteData.note.label || noteData.note.title || "";
  return Math.ceil(text.length * charWidth / wrap) * lineHeight + (noteData.note.label && noteData.note.title ? lineHeight : 0);
}

var AnnotationLayer =
/*#__PURE__*/
function (_React$Component) {
  _inherits(AnnotationLayer, _React$Component);

  function AnnotationLayer(_props) {
    var _this;

    _classCallCheck(this, AnnotationLayer);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(AnnotationLayer).call(this, _props));

    _defineProperty(_assertThisInitialized(_this), "generateSVGAnnotations", function (props, annotations) {
      var renderedAnnotations = annotations.map(function (d, i) {
        return props.svgAnnotationRule(d, i, props);
      }).filter(function (d) {
        return d !== null && d !== undefined;
      });
      return renderedAnnotations;
    });

    _defineProperty(_assertThisInitialized(_this), "generateHTMLAnnotations", function (props, annotations) {
      var renderedAnnotations = annotations.map(function (d, i) {
        return props.htmlAnnotationRule(d, i, props);
      }).filter(function (d) {
        return d !== null && d !== undefined;
      });
      return renderedAnnotations;
    });

    _defineProperty(_assertThisInitialized(_this), "processAnnotations", function (adjustableAnnotations, annotationProcessor, props) {
      var _annotationProcessor$ = annotationProcessor.layout,
          layout = _annotationProcessor$ === void 0 ? {
        type: false
      } : _annotationProcessor$;

      if (layout.type === false) {
        return adjustableAnnotations;
      }

      var _props$margin = props.margin,
          margin = _props$margin === void 0 ? {
        top: 0,
        bottom: 0,
        left: 0,
        right: 0
      } : _props$margin;
      var size = props.size,
          _props$axes = props.axes,
          axes = _props$axes === void 0 ? [] : _props$axes;
      margin = typeof margin === "number" ? {
        top: margin,
        left: margin,
        right: margin,
        bottom: margin
      } : margin;

      if (layout.type === "bump") {
        var _adjustedAnnotations = (0, _annotationHandling.bumpAnnotations)(adjustableAnnotations, layout, size, props.pointSizeFunction, props.labelSizeFunction);

        return _adjustedAnnotations;
      } else if (layout.type === "marginalia") {
        var marginOffset = layout.marginOffset,
            _layout$orient = layout.orient,
            orient = _layout$orient === void 0 ? "nearest" : _layout$orient,
            _layout$characterWidt = layout.characterWidth,
            characterWidth = _layout$characterWidt === void 0 ? 8 : _layout$characterWidt,
            _layout$lineHeight = layout.lineHeight,
            lineHeight = _layout$lineHeight === void 0 ? 20 : _layout$lineHeight,
            _layout$padding = layout.padding,
            padding = _layout$padding === void 0 ? 2 : _layout$padding,
            _layout$axisMarginOve = layout.axisMarginOverride,
            axisMarginOverride = _layout$axisMarginOve === void 0 ? null : _layout$axisMarginOve;
        var finalOrientation = orient === "nearest" ? ["left", "right", "top", "bottom"] : Array.isArray(orient) ? orient : [orient];

        var finalAxisMarginOverride = _objectSpread({
          top: null,
          right: null,
          bottom: null,
          left: null
        }, axisMarginOverride);

        var leftOn = finalOrientation.find(function (d) {
          return d === "left";
        });
        var rightOn = finalOrientation.find(function (d) {
          return d === "right";
        });
        var topOn = finalOrientation.find(function (d) {
          return d === "top";
        });
        var bottomOn = finalOrientation.find(function (d) {
          return d === "bottom";
        });
        var leftNodes = [];
        var rightNodes = [];
        var topNodes = [];
        var bottomNodes = [];
        adjustableAnnotations.forEach(function (aNote) {
          var noteData = aNote.props.noteData;
          var noteX = noteData.x[0] || noteData.x;
          var noteY = noteData.y[0] || noteData.y;
          var leftDist = leftOn ? noteX : Infinity;
          var rightDist = rightOn ? size[0] - noteX : Infinity;
          var topDist = topOn ? noteY : Infinity;
          var bottomDist = bottomOn ? size[1] - noteY : Infinity;
          var minDist = Math.min(leftDist, rightDist, topDist, bottomDist);

          if (leftDist === minDist) {
            leftNodes.push(aNote);
          } else if (rightDist === minDist) {
            rightNodes.push(aNote);
          } else if (topDist === minDist) {
            topNodes.push(aNote);
          } else {
            bottomNodes.push(aNote);
          }
        }); //Adjust the margins based on which regions are active

        var leftForce = new labella.Force({
          minPos: finalAxisMarginOverride.top ? 0 + finalAxisMarginOverride.top : 0 - margin.top,
          maxPos: finalAxisMarginOverride.bottom ? size[1] - finalAxisMarginOverride.bottom : bottomOn ? size[1] : size[1] + margin.bottom
        }).nodes(leftNodes.map(function (d) {
          var noteY = d.props.noteData.y[0] || d.props.noteData.y;
          return new labella.Node(noteY, noteDataHeight(d.props.noteData, characterWidth, lineHeight) + padding);
        })).compute();
        var rightForce = new labella.Force({
          minPos: finalAxisMarginOverride.top ? 0 + finalAxisMarginOverride.top : topOn ? 0 : 0 - margin.top,
          maxPos: finalAxisMarginOverride.bottom ? size[1] - finalAxisMarginOverride.bottom : size[1] + margin.bottom
        }).nodes(rightNodes.map(function (d) {
          var noteY = d.props.noteData.y[0] || d.props.noteData.y;
          return new labella.Node(noteY, noteDataHeight(d.props.noteData, characterWidth, lineHeight) + padding);
        })).compute();
        var topForce = new labella.Force({
          minPos: finalAxisMarginOverride.left ? 0 + finalAxisMarginOverride.left : leftOn ? 0 : 0 - margin.left,
          maxPos: finalAxisMarginOverride.right ? size[0] - finalAxisMarginOverride.right : size[0] + margin.right
        }).nodes(topNodes.map(function (d) {
          var noteX = d.props.noteData.x[0] || d.props.noteData.x;
          return new labella.Node(noteX, noteDataWidth(d.props.noteData, characterWidth) + padding);
        })).compute();
        var bottomForce = new labella.Force({
          minPos: finalAxisMarginOverride.left ? 0 + finalAxisMarginOverride.left : 0 - margin.left,
          maxPos: finalAxisMarginOverride.right ? size[0] - finalAxisMarginOverride.right : rightOn ? size[0] : size[0] + margin.right
        }).nodes(bottomNodes.map(function (d) {
          var noteX = d.props.noteData.x[0] || d.props.noteData.x;
          return new labella.Node(noteX, noteDataWidth(d.props.noteData, characterWidth) + padding);
        })).compute();
        var bottomOffset = Math.max.apply(Math, _toConsumableArray(bottomNodes.map(function (d) {
          return noteDataHeight(d.props.noteData, characterWidth, lineHeight) + padding;
        })));
        var topOffset = Math.max.apply(Math, _toConsumableArray(topNodes.map(function (d) {
          return noteDataHeight(d.props.noteData, characterWidth, lineHeight) + padding;
        })));
        var leftOffset = Math.max.apply(Math, _toConsumableArray(leftNodes.map(function (d) {
          return noteDataWidth(d.props.noteData, characterWidth) + padding;
        })));
        var rightOffset = Math.max.apply(Math, _toConsumableArray(rightNodes.map(function (d) {
          return noteDataWidth(d.props.noteData, characterWidth) + padding;
        }))); //      const nodeOffsetHeight = Math.max()

        var leftSortedNodes = leftForce.nodes();
        var rightSortedNodes = rightForce.nodes();
        var topSortedNodes = topForce.nodes();
        var bottomSortedNodes = bottomForce.nodes();
        leftNodes.forEach(function (note, i) {
          note.props.noteData.ny = leftSortedNodes[i].currentPos;
          note.props.noteData.nx = 0 - leftSortedNodes[i].layerIndex * leftOffset - marginOffsetFn("left", axes, marginOffset);

          if (note.props.noteData.note) {
            note.props.noteData.note.orientation = note.props.noteData.note.orientation || "leftRight";
            note.props.noteData.note.align = note.props.noteData.note.align || "right";
          }
        });
        rightNodes.forEach(function (note, i) {
          note.props.noteData.ny = rightSortedNodes[i].currentPos;
          note.props.noteData.nx = size[0] + rightSortedNodes[i].layerIndex * rightOffset + marginOffsetFn("right", axes, marginOffset);

          if (note.props.noteData.note) {
            note.props.noteData.note.orientation = note.props.noteData.note.orientation || "leftRight";
            note.props.noteData.note.align = note.props.noteData.note.align || "left";
          }
        });
        topNodes.forEach(function (note, i) {
          note.props.noteData.nx = topSortedNodes[i].currentPos;
          note.props.noteData.ny = 0 - topSortedNodes[i].layerIndex * topOffset - marginOffsetFn("top", axes, marginOffset);
        });
        bottomNodes.forEach(function (note, i) {
          note.props.noteData.nx = bottomSortedNodes[i].currentPos;
          note.props.noteData.ny = size[1] + bottomSortedNodes[i].layerIndex * bottomOffset + marginOffsetFn("bottom", axes, marginOffset);
        });
        return adjustableAnnotations;
      }

      return adjustableAnnotations;
    });

    _defineProperty(_assertThisInitialized(_this), "createAnnotations", function (props) {
      var renderedSVGAnnotations = _this.state.svgAnnotations,
          renderedHTMLAnnotations = [],
          adjustedAnnotations = _this.state.adjustedAnnotations,
          adjustableAnnotationsKey = _this.state.adjustedAnnotationsKey;
      var adjustedAnnotationsKey = _this.state.adjustedAnnotationsKey,
          adjustedAnnotationsDataVersion = _this.state.adjustedAnnotationsDataVersion;
      var annotations = props.annotations,
          _props$annotationHand = props.annotationHandling,
          annotationHandling = _props$annotationHand === void 0 ? false : _props$annotationHand,
          size = props.size,
          svgAnnotationRule = props.svgAnnotationRule,
          htmlAnnotationRule = props.htmlAnnotationRule;
      var annotationProcessor = _typeof(annotationHandling) === "object" ? annotationHandling : {
        layout: {
          type: annotationHandling
        },
        dataVersion: ""
      };
      var _annotationProcessor$2 = annotationProcessor.dataVersion,
          dataVersion = _annotationProcessor$2 === void 0 ? "" : _annotationProcessor$2;

      if (svgAnnotationRule) {
        var initialSVGAnnotations = _this.generateSVGAnnotations(props, annotations);

        var adjustableAnnotations = initialSVGAnnotations.filter(function (d) {
          return d.props && d.props.noteData && !d.props.noteData.fixedPosition;
        });
        var fixedAnnotations = initialSVGAnnotations.filter(function (d) {
          return !d.props || !d.props.noteData || d.props.noteData.fixedPosition;
        });
        adjustableAnnotationsKey = "".concat(adjustableAnnotations.map(adjustedAnnotationKeyMapper).join(",")).concat(JSON.stringify(annotationProcessor)).concat(size.join(","));

        if (annotationHandling === false) {
          adjustedAnnotations = adjustableAnnotations;
        }

        if (adjustedAnnotations.length !== adjustableAnnotations.length || adjustedAnnotationsKey !== adjustableAnnotationsKey || adjustedAnnotationsDataVersion !== dataVersion) {
          adjustedAnnotations = _this.processAnnotations(adjustableAnnotations, annotationProcessor, props);
        } else {
          //Handle when style or other attributes change
          adjustedAnnotations = adjustedAnnotations.map(function (d, i) {
            var newNoteData = _extends(adjustableAnnotations[i].props.noteData, {
              nx: d.props.noteData.nx,
              ny: d.props.noteData.ny,
              note: d.props.noteData.note
            });

            return React.createElement(_Annotation.default, {
              key: d.key,
              noteData: newNoteData
            });
          });
        }

        renderedSVGAnnotations = [].concat(_toConsumableArray(adjustedAnnotations), _toConsumableArray(fixedAnnotations));
      }

      if (htmlAnnotationRule) {
        renderedHTMLAnnotations = _this.generateHTMLAnnotations(props, annotations);
      }

      _this.setState({
        svgAnnotations: renderedSVGAnnotations,
        htmlAnnotations: renderedHTMLAnnotations,
        adjustedAnnotations: adjustedAnnotations,
        adjustedAnnotationsKey: adjustableAnnotationsKey,
        adjustedAnnotationsDataVersion: dataVersion
      });
    });

    _this.state = {
      svgAnnotations: [],
      htmlAnnotations: [],
      adjustedAnnotations: [],
      adjustedAnnotationsKey: "",
      adjustedAnnotationsDataVersion: ""
    };
    return _this;
  }
  /*    componentWillMount() {
      const fontLocation = this.props.fontLocation
       if (fontLocation) {
        load(fontLocation, function(err, font) {
            if (err) {
                return null
            } else {
                this.setState({ font });
            }
        });
      }
    } */


  _createClass(AnnotationLayer, [{
    key: "componentWillMount",
    value: function componentWillMount() {
      this.createAnnotations(this.props);
    }
  }, {
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      this.createAnnotations(nextProps);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state = this.state,
          svgAnnotations = _this$state.svgAnnotations,
          htmlAnnotations = _this$state.htmlAnnotations;
      var _this$props = this.props,
          useSpans = _this$props.useSpans,
          legendSettings = _this$props.legendSettings,
          margin = _this$props.margin,
          size = _this$props.size;
      var renderedLegend;

      if (legendSettings) {
        var positionHash = {
          left: [15, 15],
          right: [size[0] + 15, 15]
        };

        var _legendSettings$posit = legendSettings.position,
            _position = _legendSettings$posit === void 0 ? "right" : _legendSettings$posit,
            _legendSettings$title = legendSettings.title,
            title = _legendSettings$title === void 0 ? "Legend" : _legendSettings$title;

        var legendPosition = positionHash[_position];
        renderedLegend = React.createElement("g", {
          transform: "translate(".concat(legendPosition.join(","), ")")
        }, React.createElement(_Legend.default, _extends({}, legendSettings, {
          title: title,
          position: _position
        })));
      }

      return React.createElement(_SpanOrDiv.default, {
        span: useSpans,
        className: "annotation-layer",
        style: {
          position: "absolute",
          pointerEvents: "none",
          background: "none"
        }
      }, React.createElement("svg", {
        className: "annotation-layer-svg",
        height: size[1],
        width: size[0],
        style: {
          background: "none",
          pointerEvents: "none",
          position: "absolute",
          left: "".concat(margin.left, "px"),
          top: "".concat(margin.top, "px"),
          overflow: "visible"
        }
      }, React.createElement("g", null, renderedLegend, svgAnnotations)), React.createElement(_SpanOrDiv.default, {
        span: useSpans,
        className: "annotation-layer-html",
        style: {
          background: "none",
          pointerEvents: "none",
          position: "absolute",
          height: "".concat(size[1], "px"),
          width: "".concat(size[0], "px"),
          left: "".concat(margin.left, "px"),
          top: "".concat(margin.top, "px")
        }
      }, htmlAnnotations));
    }
  }]);

  return AnnotationLayer;
}(React.Component);

var _default = AnnotationLayer;
exports.default = _default;
module.exports = exports.default;
