// semiotic/lib/Axis.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _axis = require("./visualizationLayerBehavior/axis");

var _summaryLayouts = require("./svg/summaryLayouts");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// components
var marginalPointMapper = function marginalPointMapper(orient, width, data) {
  var xMod = orient === "left" || orient === "right" ? width / 2 : 0;
  var yMod = orient === "bottom" || orient === "top" ? width / 2 : 0;
  return data.map(function (p) {
    return [p.xy.x + xMod, p.xy.y + yMod];
  });
};

function formatValue(value, props) {
  if (props.tickFormat) {
    return props.tickFormat(value);
  }

  if (value.toString) {
    return value.toString();
  }

  return value;
}

var Axis =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Axis, _React$Component);

  function Axis(props) {
    var _this;

    _classCallCheck(this, Axis);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Axis).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "axisRef", null);

    _defineProperty(_assertThisInitialized(_this), "boundingBoxMax", function () {
      // && this.props.dynamicLabel ???
      if (!_this.axisRef) return 30;
      var _this$props$orient = _this.props.orient,
          orient = _this$props$orient === void 0 ? "left" : _this$props$orient;
      var positionType = orient === "left" || orient === "right" ? "width" : "height";
      var axisLabelMax = Math.max.apply(Math, _toConsumableArray(_toConsumableArray(_this.axisRef.querySelectorAll(".axis-label")).map(function (l) {
        return l.getBBox && l.getBBox() || {
          height: 30,
          width: 30
        };
      }).map(function (d) {
        return d[positionType];
      }))) + 25;
      return axisLabelMax;
    });

    _this.state = {
      hoverAnnotation: 0
    };
    return _this;
  }

  _createClass(Axis, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      var _this$props$label = this.props.label,
          label = _this$props$label === void 0 ? {
        position: false
      } : _this$props$label;

      if (!label.position && this.props.dynamicLabelPosition) {
        var newBBMax = this.boundingBoxMax();

        if (newBBMax !== this.state.calculatedLabelPosition) {
          this.setState({
            calculatedLabelPosition: newBBMax
          });
        }
      }
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this$props$label2 = this.props.label,
          label = _this$props$label2 === void 0 ? {
        position: false
      } : _this$props$label2;

      if (!label.position && this.props.dynamicLabelPosition) {
        var newBBMax = this.boundingBoxMax();
        this.setState({
          calculatedLabelPosition: newBBMax
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          rotate = _this$props.rotate,
          label = _this$props.label,
          _this$props$orient2 = _this$props.orient,
          orient = _this$props$orient2 === void 0 ? "left" : _this$props$orient2,
          marginalSummaryType = _this$props.marginalSummaryType,
          _this$props$tickForma = _this$props.tickFormat,
          tickFormat = _this$props$tickForma === void 0 ? marginalSummaryType ? function () {
        return "";
      } : function (d) {
        return d;
      } : _this$props$tickForma,
          size = _this$props.size,
          _this$props$width = _this$props.width,
          width = _this$props$width === void 0 ? size && size[0] || 0 : _this$props$width,
          _this$props$height = _this$props.height,
          height = _this$props$height === void 0 ? size && size[1] || 0 : _this$props$height,
          className = _this$props.className,
          padding = _this$props.padding,
          tickValues = _this$props.tickValues,
          scale = _this$props.scale,
          ticks = _this$props.ticks,
          footer = _this$props.footer,
          tickSize = _this$props.tickSize,
          tickLineGenerator = _this$props.tickLineGenerator,
          _this$props$baseline = _this$props.baseline,
          baseline = _this$props$baseline === void 0 ? true : _this$props$baseline,
          _this$props$margin = _this$props.margin,
          margin = _this$props$margin === void 0 ? {
        top: 0,
        bottom: 0,
        left: 0,
        right: 0
      } : _this$props$margin,
          _this$props$center = _this$props.center,
          center = _this$props$center === void 0 ? false : _this$props$center;
      var _this$props2 = this.props,
          axisParts = _this$props2.axisParts,
          _this$props2$position = _this$props2.position,
          position = _this$props2$position === void 0 ? [0, 0] : _this$props2$position;
      var axisTickLines;

      if (!axisParts) {
        axisParts = (0, _axis.axisPieces)({
          padding: padding,
          tickValues: tickValues,
          scale: scale,
          ticks: ticks,
          orient: orient,
          size: [width, height],
          footer: footer,
          tickSize: tickSize
        });
        axisTickLines = React.createElement("g", {
          className: "axis ".concat(className)
        }, (0, _axis.axisLines)({
          axisParts: axisParts,
          orient: orient,
          tickLineGenerator: tickLineGenerator,
          className: className
        }));
      }

      if (axisParts.length === 0) {
        return null;
      }

      var hoverWidth = 50;
      var hoverHeight = height;
      var hoverX = -50;
      var hoverY = 0;
      var baselineX = 0;
      var baselineY = 0;
      var baselineX2 = 0;
      var baselineY2 = height;

      var hoverFunction = function hoverFunction(e) {
        return _this2.setState({
          hoverAnnotation: e.nativeEvent.offsetY
        });
      };

      var circleX = 25;
      var textX = -25;
      var textY = 18;
      var lineWidth = width + 25;
      var lineHeight = 0;
      var circleY = this.state.hoverAnnotation;
      var annotationOffset = 0;
      var annotationType = "y";

      switch (orient) {
        case "right":
          position = [position[0], position[1]];
          hoverX = width;
          baselineX2 = baselineX = width;
          annotationOffset = margin.top;
          lineWidth = -width - 25;
          textX = 5;

          hoverFunction = function hoverFunction(e) {
            return _this2.setState({
              hoverAnnotation: e.nativeEvent.offsetY - annotationOffset
            });
          };

          if (center === true) {
            baselineX2 = baselineX = width / 2;
          }

          break;

        case "top":
          position = [position[0], 0];
          hoverWidth = width;
          hoverHeight = 50;
          hoverY = -50;
          hoverX = 0;
          annotationOffset = margin.left;
          annotationType = "x";
          baselineX2 = width;
          baselineY2 = 0;

          if (center === true) {
            baselineY2 = baselineY = height / 2;
          }

          hoverFunction = function hoverFunction(e) {
            return _this2.setState({
              hoverAnnotation: e.nativeEvent.offsetX - annotationOffset
            });
          };

          circleX = this.state.hoverAnnotation;
          circleY = 25;
          textX = 0;
          textY = -10;
          lineWidth = 0;
          lineHeight = height + 25;
          break;

        case "bottom":
          position = [position[0], 0];
          hoverWidth = width;
          hoverHeight = 50;
          baselineY = baselineY2 = hoverY = height;
          baselineX = hoverX = 0;
          baselineX2 = width;
          annotationOffset = margin.left;

          hoverFunction = function hoverFunction(e) {
            return _this2.setState({
              hoverAnnotation: e.nativeEvent.offsetX - annotationOffset
            });
          };

          circleX = this.state.hoverAnnotation;
          circleY = 25;
          textX = 0;
          textY = 15;
          lineWidth = 0;
          lineHeight = -height - 25;
          annotationType = "x";

          if (center === true) {
            baselineY2 = baselineY = height / 2;
          }

          break;

        default:
          position = [position[0], position[1]];
          annotationOffset = margin.top;

          if (center === true) {
            baselineX2 = baselineX = width / 2;
          }

          hoverFunction = function hoverFunction(e) {
            return _this2.setState({
              hoverAnnotation: e.nativeEvent.offsetY - annotationOffset
            });
          };

      }

      var annotationBrush;

      if (this.props.annotationFunction) {
        var formattedValue = formatValue(this.props.scale.invert(this.state.hoverAnnotation), this.props);
        var hoverGlyph = this.props.glyphFunction ? this.props.glyphFunction({
          lineHeight: lineHeight,
          lineWidth: lineWidth,
          value: this.props.scale.invert(this.state.hoverAnnotation)
        }) : React.createElement("g", null, React.isValidElement(formattedValue) ? React.createElement("g", {
          transform: "translate(".concat(textX, ",").concat(textY, ")")
        }, formattedValue) : React.createElement("text", {
          x: textX,
          y: textY
        }, formattedValue), React.createElement("circle", {
          r: 5
        }), React.createElement("line", {
          x1: lineWidth,
          y1: lineHeight,
          style: {
            stroke: "black"
          }
        }));
        var annotationSymbol = this.state.hoverAnnotation ? React.createElement("g", {
          style: {
            pointerEvents: "none"
          },
          transform: "translate(".concat(circleX, ",").concat(circleY, ")")
        }, hoverGlyph) : null;
        annotationBrush = React.createElement("g", {
          className: "annotation-brush",
          transform: "translate(".concat(hoverX, ",").concat(hoverY, ")")
        }, React.createElement("rect", {
          style: {
            fillOpacity: 0
          },
          height: hoverHeight,
          width: hoverWidth,
          onMouseMove: hoverFunction,
          onClick: function onClick() {
            return _this2.props.annotationFunction({
              className: "dynamic-axis-annotation",
              type: annotationType,
              value: _this2.props.scale.invert(_this2.state.hoverAnnotation)
            });
          },
          onMouseOut: function onMouseOut() {
            return _this2.setState({
              hoverAnnotation: undefined
            });
          }
        }), annotationSymbol);
      }

      var summaryGraphic;

      if (marginalSummaryType && this.props.xyPoints) {
        var summaryWidth = Math.max(margin[orient] - 6, 5);
        var decoratedSummaryType = typeof marginalSummaryType === "string" ? {
          type: marginalSummaryType
        } : marginalSummaryType;

        if (decoratedSummaryType.flip === undefined && (orient === "bottom" || orient === "right")) {
          decoratedSummaryType.flip = true;
        }

        var summaryStyle = decoratedSummaryType.summaryStyle ? function () {
          return decoratedSummaryType.summaryStyle;
        } : function () {
          return {
            fill: "black",
            fillOpacity: 0.5,
            stroke: "black",
            strokeDasharray: "0"
          };
        };
        var summaryRenderMode = decoratedSummaryType.renderMode ? function () {
          return decoratedSummaryType.renderMode;
        } : function () {
          return undefined;
        };
        var summaryClass = decoratedSummaryType.summaryClass ? function () {
          return decoratedSummaryType.summaryClass;
        } : function () {
          return "";
        };
        var forSummaryData = this.props.xyPoints.filter(function (p) {
          return p.x && p.y;
        }).map(function (d) {
          return _objectSpread({}, d, {
            xy: {
              x: orient === "top" || orient === "bottom" ? scale(d.x) : 0,
              y: orient === "left" || orient === "right" ? scale(d.y) : 0
            },
            piece: {
              scaledVerticalValue: scale(d.y),
              scaledValue: scale(d.x)
            },
            value: orient === "top" || orient === "bottom" ? scale(d.y) : scale(d.x),
            scaledValue: scale(d.x),
            scaledVerticalValue: scale(d.y)
          });
        });
        var renderedSummary = (0, _summaryLayouts.drawSummaries)({
          data: {
            column: {
              middle: summaryWidth / 2,
              pieceData: forSummaryData,
              width: summaryWidth,
              xyData: forSummaryData
            }
          },
          type: decoratedSummaryType,
          renderMode: summaryRenderMode,
          eventListenersGenerator: decoratedSummaryType.eventListenersGenerator || function () {
            return {};
          },
          styleFn: summaryStyle,
          classFn: summaryClass,
          positionFn: function positionFn() {
            return [0, 0];
          },
          projection: orient === "top" || orient === "bottom" ? "horizontal" : "vertical",
          adjustedSize: size,
          margin: {
            top: 0,
            bottom: 0,
            left: 0,
            right: 0
          },
          baseMarkProps: {}
        });
        var points;

        if (decoratedSummaryType.showPoints === true) {
          var mappedPoints = marginalPointMapper(orient, summaryWidth, forSummaryData);
          points = mappedPoints.map(function (d, i) {
            return React.createElement("circle", {
              key: "axis-summary-point-".concat(i),
              cx: d[0],
              cy: d[1],
              r: decoratedSummaryType.r || 3,
              style: decoratedSummaryType.pointStyle || {
                fill: "black",
                fillOpacity: 0.1
              }
            });
          });
        }

        var translation = {
          left: [-margin.left + 2, 0],
          right: [size[0] + 2, 0],
          top: [0, -margin.top + 2],
          bottom: [0, size[1] + 2]
        };
        summaryGraphic = React.createElement("g", {
          transform: "translate(".concat(translation[orient], ")")
        }, React.createElement("g", {
          transform: "translate(".concat((decoratedSummaryType.type === "contour" || decoratedSummaryType.type === "boxplot") && (orient === "left" || orient === "right") ? summaryWidth / 2 : 0, ",").concat((decoratedSummaryType.type === "contour" || decoratedSummaryType.type === "boxplot") && (orient === "top" || orient === "bottom") ? summaryWidth / 2 : 0, ")")
        }, renderedSummary.marks), points);
      }

      var axisTitle;
      var axisTickLabels = (0, _axis.axisLabels)({
        tickFormat: tickFormat,
        axisParts: axisParts,
        orient: orient,
        rotate: rotate,
        center: center
      });

      if (label) {
        var labelName = label.name || label;
        var labelPosition = label.position || {};
        var locationMod = labelPosition.location || "outside";
        var anchorMod = labelPosition.anchor || "middle";
        var distance = label.locationDistance || this.state.calculatedLabelPosition;
        var rotateHash = {
          left: -90,
          right: 90,
          top: 0,
          bottom: 0
        };
        var rotation = labelPosition.rotation || rotateHash[orient];
        var positionHash = {
          left: {
            start: [0, size[1]],
            middle: [0, size[1] / 2],
            end: [0, 0],
            inside: [distance || 15, 0],
            outside: [-(distance || 45), 0]
          },
          right: {
            start: [size[0] + 0, size[1]],
            middle: [size[0] + 0, size[1] / 2],
            end: [size[0] + 0, 0],
            inside: [-(distance || 15), 0],
            outside: [distance || 45, 0]
          },
          top: {
            start: [0, 0],
            middle: [0 + size[0] / 2, 0],
            end: [0 + size[0], 0],
            inside: [0, distance || 15],
            outside: [0, -(distance || 40)]
          },
          bottom: {
            start: [0, size[1]],
            middle: [0 + size[0] / 2, size[1]],
            end: [0 + size[0], size[1]],
            inside: [0, -(distance || 5)],
            outside: [0, distance || 50]
          }
        };
        var _translation = positionHash[orient][anchorMod];
        var location = positionHash[orient][locationMod];
        _translation[0] = _translation[0] + location[0];
        _translation[1] = _translation[1] + location[1];

        if (anchorMod === "start" && orient === "right") {
          anchorMod = "end";
        } else if (anchorMod === "end" && orient === "right") {
          anchorMod = "start";
        }

        axisTitle = React.createElement("g", {
          className: "axis-title ".concat(className),
          transform: "translate(".concat([_translation[0] + position[0], _translation[1] + position[1]], ") rotate(").concat(rotation, ")")
        }, React.isValidElement(labelName) ? labelName : React.createElement("text", {
          textAnchor: anchorMod
        }, labelName));
      }

      var axisAriaLabel = "".concat(orient, " axis ").concat(axisParts && axisParts.length > 0 && "from ".concat(tickFormat(axisParts[0].value, 0), " to ").concat(tickFormat(axisParts[axisParts.length - 1].value, axisParts.length - 1)) || "without ticks");
      return React.createElement("g", {
        className: className,
        "aria-label": axisAriaLabel,
        ref: function ref(node) {
          return _this2.axisRef = node;
        }
      }, annotationBrush, axisTickLabels, axisTickLines, baseline ? React.createElement("line", {
        key: "baseline",
        className: "axis-baseline ".concat(className),
        stroke: "black",
        strokeLinecap: "square",
        x1: baselineX,
        x2: baselineX2,
        y1: baselineY,
        y2: baselineY2
      }) : null, axisTitle, summaryGraphic);
    }
  }]);

  return Axis;
}(React.Component);

var _default = Axis;
exports.default = _default;
module.exports = exports.default;
