// semiotic/lib/FacetController.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _memoizeOne = _interopRequireDefault(require("memoize-one"));

var _frame_props = require("./constants/frame_props");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var framePropHash = {
  NetworkFrame: _frame_props.networkframeproptypes,
  XYFrame: _frame_props.xyframeproptypes,
  OrdinalFrame: _frame_props.ordinalframeproptypes,
  ResponsiveNetworkFrame: _objectSpread({}, _frame_props.networkframeproptypes, _frame_props.responsiveprops),
  ResponsiveXYFrame: _objectSpread({}, _frame_props.xyframeproptypes, _frame_props.responsiveprops),
  ResponsiveOrdinalFrame: _objectSpread({}, _frame_props.ordinalframeproptypes, _frame_props.responsiveprops),
  SparkNetworkFrame: _objectSpread({}, _frame_props.networkframeproptypes),
  SparkXYFrame: _objectSpread({}, _frame_props.xyframeproptypes),
  SparkOrdinalFrame: _objectSpread({}, _frame_props.ordinalframeproptypes)
};

var buildNewState = function buildNewState(prevState, extentValue, extentType, extentPosition) {
  var oldExtents = prevState.rawExtents[extentType] || {};
  oldExtents[extentPosition] = extentValue;
  var extentMinMaxValues = Object.values(oldExtents).flat();
  return {
    extents: _objectSpread({}, prevState.extents, _defineProperty({}, extentType, extentMinMaxValues.length === 0 ? undefined : [Math.min.apply(Math, _toConsumableArray(extentMinMaxValues)), Math.max.apply(Math, _toConsumableArray(extentMinMaxValues))])),
    rawExtents: _objectSpread({}, prevState.rawExtents, _defineProperty({}, extentType, oldExtents))
  };
};

function validFrameProps(originalProps, frameType) {
  var newProps = {};
  var frameProps = framePropHash[frameType];
  Object.keys(originalProps).forEach(function (key) {
    if (frameProps[key]) {
      newProps[key] = originalProps[key];
    }
  });
  return newProps;
}

var FacetController =
/*#__PURE__*/
function (_React$Component) {
  _inherits(FacetController, _React$Component);

  function FacetController() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, FacetController);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(FacetController)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {
      extents: {},
      rawExtents: {},
      facetHover: undefined
      /**
       * Helper for creating extent if we have a  min/max value
       * use that else use the onChange version so we can in return
       * normalize all of the facets to have the same extents
       */

    });

    _defineProperty(_assertThisInitialized(_this), "createExtent", function (extentType, state, index) {
      return state.extents && state.extents[extentType] ? {
        onChange: _this.extentHandler(extentType, index),
        extent: state.extents[extentType]
      } : {
        onChange: _this.extentHandler(extentType, index)
      };
    });

    _defineProperty(_assertThisInitialized(_this), "extentHandler", function (extentType, extentPosition) {
      return function (extentValue) {
        _this.setState(function (prevState) {
          return buildNewState(prevState, extentValue, extentType, extentPosition);
        });

        return extentValue;
      };
    });

    _defineProperty(_assertThisInitialized(_this), "generateChildAnnotations", function (_ref) {
      var originalAnnotations = _ref.originalAnnotations,
          state = _ref.state;
      var annotationBase = state.facetHoverAnnotations;
      var annotationSettings = _this.props.hoverAnnotation || _this.props.pieceHoverAnnotation;

      if (!annotationSettings || !annotationBase) {
        return [];
      }

      if (state.facetHover) {
        var annotations = _toConsumableArray(originalAnnotations);

        if (annotationSettings === true) {
          annotationBase = [_objectSpread({}, state.facetHover)];
        } else {
          var annotationMap = annotationSettings;
          annotationBase = annotationMap.map(function (annotation) {
            var decoratedAnnotation = typeof annotation === "function" ? annotation(state.facetHover) : _objectSpread({}, state.facetHover, annotation);
            return decoratedAnnotation;
          });
        }

        if (Array.isArray(annotationBase)) {
          annotationBase.forEach(function (annotation) {
            if (typeof annotation !== "function") {
              if (annotation.type === "column-hover") {
                annotation.facetColumn = annotation.column.name;
                annotation.column = undefined;
              } else {
                if (!annotation.type) {
                  annotation.type = "frame-hover";
                }

                annotation.y = undefined;
                annotation.yBottom = undefined;
                annotation.yMiddle = undefined;
                annotation.yTop = undefined;
              }
            }
          });
          annotations.push.apply(annotations, _toConsumableArray(annotationBase));
        }

        return annotations;
      }

      return originalAnnotations;
    });

    _defineProperty(_assertThisInitialized(_this), "mapChildrenWithAppropriateProps", function (_ref2) {
      var child = _ref2.child,
          index = _ref2.index,
          originalAnnotations = _ref2.originalAnnotations,
          props = _ref2.props,
          state = _ref2.state;
      var childType = child.type;
      var frameType = childType.displayName;

      var annotations = _this.generateChildAnnotations({
        state: state,
        originalAnnotations: originalAnnotations
      });

      var customProps = _objectSpread({}, props, {
        annotations: annotations
      });

      if (!framePropHash[frameType]) {
        return React.cloneElement(child, {
          facetprops: customProps
        });
      } // pieceHoverAnnotation could be an object, so we need to be explicit in checking for true


      if (props.hoverAnnotation || props.pieceHoverAnnotation) {
        customProps.customHoverBehavior = function (d) {
          return _this.setState({
            facetHover: d,
            facetHoverAnnotations: props.hoverAnnotation || props.pieceHoverAnnotation
          });
        };
      }

      if ((frameType === "OrdinalFrame" || frameType === "ResponsiveOrdinalFrame" || frameType === "SparkOrdinalFrame") && props.sharedRExtent === true) {
        customProps.rExtent = _this.createExtent("rExtent", state, index);

        customProps.onUnmount = function () {
          _this.setState(function (prevState) {
            return buildNewState(prevState, [], "rExtent", index);
          });
        };
      }

      if ((frameType === "XYFrame" || frameType === "ResponsiveXYFrame" || frameType === "SparkXYFrame") && props.sharedXExtent === true) {
        customProps.xExtent = _this.createExtent("xExtent", state, index);

        customProps.onUnmount = function () {
          _this.setState(function (prevState) {
            return buildNewState(prevState, [], "xExtent", index);
          });
        };
      }

      if ((frameType === "XYFrame" || frameType === "ResponsiveXYFrame" || frameType === "SparkXYFrame") && props.sharedYExtent === true) {
        customProps.yExtent = _this.createExtent("yExtent", state, index);

        customProps.onUnmount = function () {
          _this.setState(function (prevState) {
            return buildNewState(prevState, [], "yExtent", index);
          });
        };
      }

      if (customProps.pieceHoverAnnotation) {
        customProps.pieceHoverAnnotation = [];
      } else if (customProps.hoverAnnotation) {
        customProps.hoverAnnotation = [];
      }

      return React.cloneElement(child, validFrameProps(customProps, frameType));
    });

    _defineProperty(_assertThisInitialized(_this), "processFacetController", (0, _memoizeOne.default)(function (props, state) {
      return React.Children.map(props.children, function (child, index) {
        if (!child) return null;
        return _this.mapChildrenWithAppropriateProps({
          child: child,
          index: index,
          originalAnnotations: child.props.annotations || [],
          props: props,
          state: state
        });
      });
    }));

    return _this;
  }

  _createClass(FacetController, [{
    key: "render",
    value: function render() {
      var Wrapper = this.props.react15Wrapper;

      if (Wrapper) {
        return React.cloneElement(Wrapper, undefined, this.processFacetController(this.props, this.state));
      }

      return React.createElement(React.Fragment, null, this.processFacetController(this.props, this.state));
    }
  }]);

  return FacetController;
}(React.Component);

var _default = FacetController;
exports.default = _default;
module.exports = exports.default;
