// semiotic/lib/InteractionLayer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _d3Brush = require("d3-brush");

var _d3Array = require("d3-array");

var _d3Selection = require("d3-selection");

var _d3Voronoi = require("d3-voronoi");

var _semioticMark = require("semiotic-mark");

var _Brush = _interopRequireDefault(require("./Brush"));

var _SpanOrDiv = _interopRequireDefault(require("./SpanOrDiv"));

var _coordinateNames = require("./constants/coordinateNames");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var generateOMappingFn = function generateOMappingFn(projectedColumns) {
  return function (d) {
    if (d) {
      var columnValues = Object.values(projectedColumns);
      var foundColumns = columnValues.filter(function (c) {
        return d[1] >= c.x && d[0] <= c.x + c.width;
      });
      return foundColumns;
    }

    return null;
  };
};

var generateOEndMappingFn = function generateOEndMappingFn(projectedColumns) {
  return function (d) {
    if (d && _d3Selection.event.sourceEvent && _d3Selection.event.sourceEvent.path && _d3Selection.event.sourceEvent.path[1] && _d3Selection.event.sourceEvent.path[1].classList.contains("xybrush") && _d3Selection.event.target.move) {
      var columnValues = Object.values(projectedColumns);
      var foundColumns = columnValues.filter(function (c) {
        return d[1] >= c.x && d[0] <= c.x + c.width;
      });
      var firstColumn = foundColumns[0] || {
        x: 0,
        width: 0
      };
      var lastColumn = foundColumns[foundColumns.length - 1] || {
        x: 0,
        width: 0
      };
      var columnPosition = [firstColumn.x + Math.min(5, firstColumn.width / 10), lastColumn.x + lastColumn.width - Math.min(5, lastColumn.width / 10)];
      (0, _d3Selection.select)(_d3Selection.event.sourceEvent.path[1]).transition(750).call(_d3Selection.event.target.move, columnPosition);
      return foundColumns;
    }

    return null;
  };
};

var InteractionLayer =
/*#__PURE__*/
function (_React$Component) {
  _inherits(InteractionLayer, _React$Component);

  function InteractionLayer(_props) {
    var _this;

    _classCallCheck(this, InteractionLayer);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(InteractionLayer).call(this, _props));

    _defineProperty(_assertThisInitialized(_this), "interactionContext", null);

    _defineProperty(_assertThisInitialized(_this), "canvasMap", new Map());

    _defineProperty(_assertThisInitialized(_this), "constructDataObject", function (d) {
      if (d === undefined) return d;
      var points = _this.props.points;
      return d && d.data ? _objectSpread({
        points: points
      }, d.data, d) : _objectSpread({
        points: points
      }, d);
    });

    _defineProperty(_assertThisInitialized(_this), "changeVoronoi", function (d, customHoverTypes) {
      var _this$props = _this.props,
          customHoverBehavior = _this$props.customHoverBehavior,
          voronoiHover = _this$props.voronoiHover; //Until semiotic 2

      var dataObject = _this.constructDataObject(d);

      if (customHoverBehavior) customHoverBehavior(dataObject);
      if (!d) voronoiHover(null);else if (customHoverTypes === true) {
        var vorD = _extends({}, dataObject);

        vorD.type = vorD.type === "column-hover" ? "column-hover" : "frame-hover";
        voronoiHover(vorD);
      } else if (customHoverTypes) {
        var arrayWrappedHoverTypes = Array.isArray(customHoverTypes) ? customHoverTypes : [customHoverTypes];
        var mappedHoverTypes = arrayWrappedHoverTypes.map(function (c) {
          var finalC = typeof c === "function" ? c(dataObject) : c;
          if (!finalC) return undefined;
          return _extends({}, dataObject, finalC);
        }).filter(function (d) {
          return d;
        });
        voronoiHover(mappedHoverTypes);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "clickVoronoi", function (d) {
      //Until semiotic 2
      var dataObject = _this.constructDataObject(d);

      if (_this.props.customClickBehavior) _this.props.customClickBehavior(dataObject);
    });

    _defineProperty(_assertThisInitialized(_this), "doubleclickVoronoi", function (d) {
      //Until semiotic 2
      var dataObject = _this.constructDataObject(d);

      if (_this.props.customDoubleClickBehavior) _this.props.customDoubleClickBehavior(dataObject);
    });

    _defineProperty(_assertThisInitialized(_this), "brushStart", function (e, column, data) {
      if (_this.props.interaction && _this.props.interaction.start) _this.props.interaction.start(e, column, data);
    });

    _defineProperty(_assertThisInitialized(_this), "brush", function (e, column, data) {
      if (_this.props.interaction && _this.props.interaction.during) _this.props.interaction.during(e, column, data);
    });

    _defineProperty(_assertThisInitialized(_this), "brushEnd", function (e, column, data) {
      if (_this.props.interaction && _this.props.interaction.end) _this.props.interaction.end(e, column, data);
    });

    _defineProperty(_assertThisInitialized(_this), "createBrush", function (interaction) {
      var semioticBrush, mappingFn, selectedExtent, endMappingFn;
      var _this$props2 = _this.props,
          xScale = _this$props2.xScale,
          yScale = _this$props2.yScale,
          size = _this$props2.size,
          renderPipeline = _this$props2.renderPipeline;
      var brushData = {};
      Object.entries(renderPipeline).forEach(function (_ref) {
        var _ref2 = _slicedToArray(_ref, 2),
            key = _ref2[0],
            value = _ref2[1];

        if (value.data && value.data.length > 0) {
          brushData[key] = value.data;
        }
      });
      var projection = interaction.projection,
          projectedColumns = interaction.projectedColumns;
      var actualBrush = interaction.brush === "oBrush" ? projection === "horizontal" ? "yBrush" : "xBrush" : interaction.brush;
      var _interaction$extent = interaction.extent,
          extent = _interaction$extent === void 0 ? actualBrush === "xyBrush" ? [[xScale.invert(0), yScale.invert(0)], [xScale.invert(size[0]), yScale.invert(size[1])]] : actualBrush === "xBrush" ? [xScale.invert(0), xScale.invert(size[0])] : [yScale.invert(0), yScale.invert(size[1])] : _interaction$extent;

      if (extent.indexOf && extent.indexOf(undefined) !== -1) {
        return React.createElement("g", null);
      }

      if (actualBrush === "xBrush") {
        var castExtent = extent;

        mappingFn = function mappingFn(d) {
          return !d ? null : [xScale.invert(d[0]), xScale.invert(d[1])];
        };

        semioticBrush = (0, _d3Brush.brushX)();
        selectedExtent = castExtent.map(function (d) {
          return xScale(d);
        });
        endMappingFn = mappingFn;
      } else if (actualBrush === "yBrush") {
        var _castExtent = extent;

        mappingFn = function mappingFn(d) {
          return !d ? null : [yScale.invert(d[0]), yScale.invert(d[1])].sort(function (a, b) {
            return a - b;
          });
        };

        semioticBrush = (0, _d3Brush.brushY)();
        selectedExtent = _castExtent.map(function (d) {
          return yScale(d);
        }).sort(function (a, b) {
          return a - b;
        });
        endMappingFn = mappingFn;
      } else {
        var _castExtent2 = extent;

        if (_castExtent2.indexOf(undefined) !== -1 || _castExtent2[0].indexOf(undefined) !== -1 || _castExtent2[1].indexOf(undefined) !== -1) {
          return React.createElement("g", null);
        }

        semioticBrush = (0, _d3Brush.brush)();

        mappingFn = function mappingFn(d) {
          if (!d) return null;
          var yValues = [yScale.invert(d[0][1]), yScale.invert(d[1][1])].sort(function (a, b) {
            return a - b;
          });
          return [[xScale.invert(d[0][0]), yValues[0]], [xScale.invert(d[1][0]), yValues[1]]];
        };

        var yValues = [yScale(extent[0][1]), yScale(extent[1][1])].sort(function (a, b) {
          return a - b;
        });
        selectedExtent = _castExtent2.map(function (d, i) {
          return [xScale(d[0]), yValues[i]];
        });
        endMappingFn = mappingFn;
      }

      if (interaction.brush === "oBrush") {
        selectedExtent = null;

        if (interaction.extent) {
          var _interaction$extent2 = _slicedToArray(interaction.extent, 2),
              leftExtent = _interaction$extent2[0],
              rightExtent = _interaction$extent2[1];

          if ((typeof leftExtent === "string" || typeof leftExtent === "number") && (typeof rightExtent === "string" || typeof rightExtent === "number")) {
            selectedExtent = [projectedColumns[leftExtent].x, projectedColumns[rightExtent].x + projectedColumns[rightExtent].width];
          }
        }

        mappingFn = generateOMappingFn(projectedColumns);
        endMappingFn = generateOEndMappingFn(projectedColumns);
      }

      semioticBrush.extent([[0, 0], [_this.props.size[0], _this.props.size[1]]]).on("start", function () {
        _this.brushStart(mappingFn(_d3Selection.event.selection), undefined, brushData);
      }).on("brush", function () {
        _this.brush(mappingFn(_d3Selection.event.selection), undefined, brushData);
      }).on("end", function () {
        _this.brushEnd(endMappingFn(_d3Selection.event.selection), undefined, brushData);
      });
      return React.createElement("g", {
        className: "brush"
      }, React.createElement(_Brush.default, {
        selectedExtent: selectedExtent,
        extent: extent,
        svgBrush: semioticBrush
      }));
    });

    _defineProperty(_assertThisInitialized(_this), "calculateOverlay", function (props) {
      var voronoiPaths = [];
      var xScale = props.xScale,
          yScale = props.yScale,
          points = props.points,
          projectedX = props.projectedX,
          showLinePoints = props.showLinePoints,
          size = props.size,
          overlay = props.overlay,
          _props$interactionOve = props.interactionOverflow,
          interactionOverflow = _props$interactionOve === void 0 ? {
        top: 0,
        bottom: 0,
        left: 0,
        right: 0
      } : _props$interactionOve,
          customClickBehavior = props.customClickBehavior,
          customDoubleClickBehavior = props.customDoubleClickBehavior,
          hoverAnnotation = props.hoverAnnotation;
      var whichPoints = {
        top: _coordinateNames.projectedYTop,
        bottom: _coordinateNames.projectedYBottom
      };
      var pointerStyle = customClickBehavior || customDoubleClickBehavior ? {
        cursor: "pointer"
      } : {};

      if (points && hoverAnnotation && !overlay) {
        var voronoiDataset = [];
        var voronoiUniqueHash = {};
        points.forEach(function (d) {
          var xValue = Math.floor(xScale(d[projectedX]));
          var yValue = Math.floor(yScale(showLinePoints && d[whichPoints[showLinePoints]] !== undefined ? d[whichPoints[showLinePoints]] : d[_coordinateNames.projectedYMiddle] !== undefined ? d[_coordinateNames.projectedYMiddle] : d[_coordinateNames.projectedY]));

          if (xValue >= 0 && xValue <= size[0] && yValue >= 0 && yValue <= size[1] && xValue !== undefined && yValue !== undefined && isNaN(xValue) === false && isNaN(yValue) === false) {
            var pointKey = "".concat(xValue, ",").concat(yValue);

            if (!voronoiUniqueHash[pointKey]) {
              var voronoiPoint = _objectSpread({}, d, {
                coincidentPoints: [d],
                voronoiX: xValue,
                voronoiY: yValue
              });

              voronoiDataset.push(voronoiPoint);
              voronoiUniqueHash[pointKey] = voronoiPoint;
            } else voronoiUniqueHash[pointKey].coincidentPoints.push(d);
          }
        });
        var voronoiXExtent = (0, _d3Array.extent)(voronoiDataset.map(function (d) {
          return d.voronoiX;
        }));
        var voronoiYExtent = (0, _d3Array.extent)(voronoiDataset.map(function (d) {
          return d.voronoiY;
        }));
        var voronoiExtent = [[Math.min(voronoiXExtent[0], -interactionOverflow.left), Math.min(voronoiYExtent[0], -interactionOverflow.top)], [Math.max(voronoiXExtent[1], size[0] + interactionOverflow.right), Math.max(voronoiYExtent[1], size[1] + interactionOverflow.bottom)]];
        var voronoiDiagram = (0, _d3Voronoi.voronoi)().extent(voronoiExtent).x(function (d) {
          return d.voronoiX;
        }).y(function (d) {
          return d.voronoiY;
        });
        var voronoiData = voronoiDiagram.polygons(voronoiDataset);
        voronoiPaths = voronoiData.map(function (d, i) {
          return React.createElement("path", {
            onClick: function onClick() {
              _this.clickVoronoi(voronoiDataset[i]);
            },
            onDoubleClick: function onDoubleClick() {
              _this.doubleclickVoronoi(voronoiDataset[i]);
            },
            onMouseEnter: function onMouseEnter() {
              _this.changeVoronoi(voronoiDataset[i], props.hoverAnnotation);
            },
            onMouseLeave: function onMouseLeave() {
              _this.changeVoronoi();
            },
            key: "interactionVoronoi".concat(i),
            d: "M".concat(d.join("L"), "Z"),
            style: _objectSpread({
              fillOpacity: 0
            }, pointerStyle)
          });
        }, _assertThisInitialized(_this));
        return voronoiPaths;
      } else if (overlay) {
        var renderedOverlay = overlay.map(function (overlayRegion, i) {
          var overlayData = overlayRegion.overlayData,
              rest = _objectWithoutProperties(overlayRegion, ["overlayData"]);

          if (React.isValidElement(overlayRegion.renderElement)) {
            var overlayProps = {
              key: "overlay-".concat(i),
              onMouseEnter: function onMouseEnter() {
                _this.changeVoronoi(overlayData, props.hoverAnnotation);
              },
              onMouseLeave: function onMouseLeave() {
                _this.changeVoronoi();
              },
              onClick: function onClick() {
                _this.clickVoronoi(overlayData);
              },
              onDoubleClick: function onDoubleClick() {
                _this.doubleclickVoronoi(overlayData);
              },
              style: _objectSpread({
                opacity: 0
              }, pointerStyle)
            };
            return React.cloneElement(overlayRegion.renderElement, overlayProps);
          } else {
            return React.createElement(_semioticMark.Mark, _extends({
              forceUpdate: true
            }, rest, {
              key: "overlay-".concat(i),
              onMouseEnter: function onMouseEnter() {
                _this.changeVoronoi(overlayData, props.hoverAnnotation);
              },
              onMouseLeave: function onMouseLeave() {
                _this.changeVoronoi();
              },
              onClick: function onClick() {
                _this.clickVoronoi(overlayData);
              },
              onDoubleClick: function onDoubleClick() {
                _this.doubleclickVoronoi(overlayData);
              },
              style: _objectSpread({
                opacity: 0
              }, pointerStyle)
            }));
          }
        });
        return renderedOverlay;
      }
    });

    _defineProperty(_assertThisInitialized(_this), "canvasRendering", function () {
      if (_this.interactionContext === null || !_this.state.overlayRegions) return;
      var _this$props3 = _this.props,
          svgSize = _this$props3.svgSize,
          margin = _this$props3.margin;
      var overlayRegions = _this.state.overlayRegions;

      _this.canvasMap.clear();

      var interactionContext = _this.interactionContext.getContext("2d");

      interactionContext.imageSmoothingEnabled = false;
      interactionContext.setTransform(1, 0, 0, 1, margin.left, margin.top);
      interactionContext.clearRect(-margin.left, -margin.top, svgSize[0], svgSize[1]);
      interactionContext.lineWidth = 1;
      overlayRegions.forEach(function (overlay, oi) {
        var interactionRGBA = "rgba(".concat(Math.floor(Math.random() * 255), ",").concat(Math.floor(Math.random() * 255), ",").concat(Math.floor(Math.random() * 255), ",255)");

        _this.canvasMap.set(interactionRGBA, oi);

        interactionContext.fillStyle = interactionRGBA;
        interactionContext.strokeStyle = interactionRGBA;
        var p = new Path2D(overlay.props.d);
        interactionContext.stroke(p);
        interactionContext.fill(p);
      });
    });

    _defineProperty(_assertThisInitialized(_this), "createColumnsBrush", function (interaction) {
      var _this$props4 = _this.props,
          projection = _this$props4.projection,
          rScale = _this$props4.rScale,
          size = _this$props4.size,
          oColumns = _this$props4.oColumns,
          renderPipeline = _this$props4.renderPipeline;
      if (!projection || !rScale || !oColumns) return;
      var brushData = {};
      Object.entries(renderPipeline).forEach(function (_ref3) {
        var _ref4 = _slicedToArray(_ref3, 2),
            key = _ref4[0],
            value = _ref4[1];

        if (value.data && value.data.length > 0) {
          brushData[key] = value.data;
        }
      });
      var semioticBrush, mappingFn;
      var max = rScale.domain()[1];

      if (projection && projection === "horizontal") {
        mappingFn = function mappingFn(d) {
          return !d ? null : [rScale.invert(d[0]), rScale.invert(d[1])];
        };
      } else mappingFn = function mappingFn(d) {
        return !d ? null : [Math.abs(rScale.invert(d[1]) - max), Math.abs(rScale.invert(d[0]) - max)];
      };

      var rRange = rScale.range();
      var columnHash = oColumns;
      var brushPosition, selectedExtent;
      var brushes = Object.keys(columnHash).map(function (c) {
        if (projection && projection === "horizontal") {
          selectedExtent = interaction.extent[c] ? interaction.extent[c].map(function (d) {
            return rScale(d);
          }) : rRange;
          brushPosition = [0, columnHash[c].x];
          semioticBrush = (0, _d3Brush.brushX)();
          semioticBrush.extent([[rRange[0], 0], [rRange[1], columnHash[c].width]]).on("start", function () {
            _this.brushStart(mappingFn(_d3Selection.event.selection), c, brushData);
          }).on("brush", function () {
            _this.brush(mappingFn(_d3Selection.event.selection), c, brushData);
          }).on("end", function () {
            _this.brushEnd(mappingFn(_d3Selection.event.selection), c, brushData);
          });
        } else {
          selectedExtent = interaction.extent[c] ? interaction.extent[c].map(function (d) {
            return rRange[1] - rScale(d);
          }).reverse() : rRange;
          brushPosition = [columnHash[c].x, 0];
          semioticBrush = (0, _d3Brush.brushY)();
          semioticBrush.extent([[0, rRange[0]], [columnHash[c].width, rRange[1]]]).on("start", function () {
            _this.brushStart(mappingFn(_d3Selection.event.selection), c, brushData);
          }).on("brush", function () {
            _this.brush(mappingFn(_d3Selection.event.selection), c, brushData);
          }).on("end", function () {
            _this.brushEnd(mappingFn(_d3Selection.event.selection), c, brushData);
          });
        }

        return React.createElement("g", {
          key: "column-brush-".concat(c),
          className: "brush"
        }, React.createElement(_Brush.default, {
          key: "orbrush".concat(c),
          selectedExtent: selectedExtent,
          svgBrush: semioticBrush,
          position: brushPosition
        }));
      });
      return brushes;
    });

    _this.state = {
      overlayRegions: _this.calculateOverlay(_props),
      interactionCanvas: React.createElement("canvas", {
        className: "frame-canvas-interaction",
        ref: function ref(canvasContext) {
          if (canvasContext) {
            canvasContext.onmousemove = function (e) {
              var interactionContext = canvasContext.getContext("2d");
              var hoverPoint = interactionContext.getImageData(e.offsetX, e.offsetY, 1, 1);
              var mostCommonRGB = "rgba(".concat(hoverPoint.data[0], ",").concat(hoverPoint.data[1], ",").concat(hoverPoint.data[2], ",255)");

              var overlay = _this.state.overlayRegions[_this.canvasMap.get(mostCommonRGB)];

              if (!overlay) {
                var hoverArea = interactionContext.getImageData(e.offsetX - 2, e.offsetY - 2, 5, 5);
                var _x = 0;

                while (!overlay && _x < 100) {
                  overlay = _this.state.overlayRegions[_this.canvasMap.get("rgba(".concat(hoverArea.data[_x], ",").concat(hoverArea.data[_x + 1], ",").concat(hoverArea.data[_x + 2], ",255)"))];
                  _x += 4;
                }
              }

              if (overlay && overlay.props) {
                overlay.props.onMouseEnter();
              } else {
                _this.changeVoronoi();
              }
            };
          }

          _this.interactionContext = canvasContext;
        },
        style: {
          position: "absolute",
          left: "0px",
          top: "0px",
          imageRendering: "pixelated",
          pointerEvents: "all",
          opacity: 0
        },
        width: _props.svgSize[0],
        height: _props.svgSize[1]
      })
    };
    return _this;
  }

  _createClass(InteractionLayer, [{
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      if (this.props.overlay !== nextProps.overlay || nextProps.points !== this.props.points || this.props.xScale !== nextProps.xScale || this.props.yScale !== nextProps.yScale || this.props.hoverAnnotation !== nextProps.hoverAnnotation) {
        this.setState({
          overlayRegions: this.calculateOverlay(nextProps)
        });
      }
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      this.canvasRendering();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState) {
      if (this.state.overlayRegions !== prevState.overlayRegions) {
        this.canvasRendering();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var semioticBrush = null;
      var _this$props5 = this.props,
          interaction = _this$props5.interaction,
          svgSize = _this$props5.svgSize,
          margin = _this$props5.margin,
          _this$props5$useSpans = _this$props5.useSpans,
          useSpans = _this$props5$useSpans === void 0 ? false : _this$props5$useSpans,
          canvasRendering = _this$props5.canvasRendering,
          disableCanvasInteraction = _this$props5.disableCanvasInteraction;
      var overlayRegions = this.state.overlayRegions;
      var enabled = this.props.enabled;

      if (interaction && interaction.brush) {
        enabled = true;
        semioticBrush = this.createBrush(interaction);
      }

      if (interaction && interaction.columnsBrush) {
        enabled = true;
        semioticBrush = this.createColumnsBrush(interaction);
      }

      if (!overlayRegions && !semioticBrush) {
        return null;
      }

      var interactionCanvas = !disableCanvasInteraction && canvasRendering && this.state.overlayRegions && this.state.interactionCanvas;
      return React.createElement(_SpanOrDiv.default, {
        span: useSpans,
        className: "interaction-layer",
        style: {
          position: "absolute",
          background: "none",
          pointerEvents: "none"
        }
      }, interactionCanvas || React.createElement("svg", {
        height: svgSize[1],
        width: svgSize[0],
        style: {
          background: "none",
          pointerEvents: "none"
        }
      }, React.createElement("g", {
        className: "interaction-overlay",
        transform: "translate(".concat(margin.left, ",").concat(margin.top, ")"),
        style: {
          pointerEvents: enabled ? "all" : "none"
        }
      }, React.createElement("g", {
        className: "interaction-regions"
      }, overlayRegions), semioticBrush)));
    }
  }]);

  return InteractionLayer;
}(React.Component);

_defineProperty(InteractionLayer, "defaultProps", {
  svgSize: [500, 500]
});

var _default = InteractionLayer;
exports.default = _default;
module.exports = exports.default;
