// semiotic/lib/NetworkFrame.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _d3Force = require("d3-force");

var _d3BboxCollide = require("d3-bboxCollide");

var _d3Scale = require("d3-scale");

var _d3Array = require("d3-array");

var _AnnotationLabel = _interopRequireDefault(require("react-annotation/lib/Types/AnnotationLabel"));

var _Frame = _interopRequireDefault(require("./Frame"));

var _DownloadButton = _interopRequireDefault(require("./DownloadButton"));

var _frameFunctions = require("./svg/frameFunctions");

var _pieceDrawing = require("./svg/pieceDrawing");

var _networkDrawing = require("./svg/networkDrawing");

var _dataFunctions = require("./data/dataFunctions");

var _downloadDataMapping = require("./downloadDataMapping");

var _d3SankeyCircular = require("d3-sankey-circular");

var _d3Chord = require("d3-chord");

var _d3Shape = require("d3-shape");

var _d3Hierarchy = require("d3-hierarchy");

var _frame_props = require("./constants/frame_props");

var _networkframeRules = require("./annotationRules/networkframeRules");

var _baseRules = require("./annotationRules/baseRules");

var _functions = require("./untyped_utilities/functions");

var _svgPathBoundingBox = _interopRequireDefault(require("svg-path-bounding-box"));

var _network = require("./processing/network");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

var emptyArray = [];
var matrixRenderOrder = ["nodes", "edges"];
var generalRenderOrder = ["edges", "nodes"];
var baseNodeProps = {
  id: undefined,
  degree: 0,
  inDegree: 0,
  outDegree: 0,
  x: 0,
  y: 0,
  x1: 0,
  x0: 0,
  y1: 0,
  y0: 0,
  height: 0,
  width: 0,
  radius: 0,
  r: 0,
  direction: undefined,
  textHeight: 0,
  textWidth: 0,
  fontSize: 0,
  scale: 1,
  nodeSize: 0,
  component: -99,
  shapeNode: false
};
var baseNetworkSettings = {
  iterations: 500,
  hierarchicalNetwork: false
};
var baseGraphSettings = {
  nodeHash: new Map(),
  edgeHash: new Map(),
  nodes: [],
  edges: [],
  hierarchicalNetwork: false,
  type: "force"
};

var basicMiddle = function basicMiddle(d) {
  return {
    edge: d,
    x: (d.source.x + d.target.x) / 2,
    y: (d.source.y + d.target.y) / 2
  };
};

var edgePointHash = {
  sankey: function sankey(d) {
    return {
      edge: d,
      x: (d.source.x1 + d.target.x0) / 2,
      y: d.circularPathData ? d.circularPathData.verticalFullExtent : ((d.y0 + d.y1) / 2 + (d.y0 + d.y1) / 2) / 2
    };
  },
  force: basicMiddle,
  tree: basicMiddle,
  cluster: basicMiddle
};
var hierarchicalTypeHash = {
  dendrogram: _d3Hierarchy.tree,
  tree: _d3Hierarchy.tree,
  circlepack: _d3Hierarchy.pack,
  cluster: _d3Hierarchy.cluster,
  treemap: _d3Hierarchy.treemap,
  partition: _d3Hierarchy.partition
};
var hierarchicalProjectable = {
  partition: true,
  cluster: true,
  tree: true,
  dendrogram: true
};
var radialProjectable = {
  partition: true,
  cluster: true,
  tree: true,
  dendrogram: true
  /*
  const customEdgeHashProject = {
    offset: glyphProject.offset,
    parallel: glyphProject.parallel
  }
  
  const customEdgeHashMutate = {
    particle: glyphMutate.particle
  }
  */

};

function determineNodeIcon(baseCustomNodeIcon, networkSettings, size, nodes) {
  if (baseCustomNodeIcon) return baseCustomNodeIcon;
  var center = [size[0] / 2, size[1] / 2];

  switch (networkSettings.type) {
    case "sankey":
      return _networkDrawing.sankeyNodeGenerator;

    case "partition":
      return networkSettings.projection === "radial" ? (0, _networkDrawing.radialRectNodeGenerator)(size, center, networkSettings) : _networkDrawing.hierarchicalRectNodeGenerator;

    case "treemap":
      return networkSettings.projection === "radial" ? (0, _networkDrawing.radialRectNodeGenerator)(size, center, networkSettings) : _networkDrawing.hierarchicalRectNodeGenerator;

    case "circlepack":
      return _networkDrawing.circleNodeGenerator;

    case "wordcloud":
      return _networkDrawing.wordcloudNodeGenerator;

    case "chord":
      return (0, _networkDrawing.chordNodeGenerator)(size);

    case "dagre":
      return _networkDrawing.hierarchicalRectNodeGenerator;

    case "matrix":
      return (0, _networkDrawing.matrixNodeGenerator)(size, nodes);
  }

  return _networkDrawing.circleNodeGenerator;
}

function determineEdgeIcon(_ref) {
  var baseCustomEdgeIcon = _ref.baseCustomEdgeIcon,
      networkSettings = _ref.networkSettings,
      size = _ref.size,
      graph = _ref.graph,
      nodes = _ref.nodes;
  if (baseCustomEdgeIcon) return baseCustomEdgeIcon;

  switch (networkSettings.type) {
    case "partition":
      return function () {
        return null;
      };

    case "treemap":
      return function () {
        return null;
      };

    case "circlepack":
      return function () {
        return null;
      };

    case "wordcloud":
      return function () {
        return null;
      };

    case "chord":
      return (0, _networkDrawing.chordEdgeGenerator)(size);

    case "matrix":
      return (0, _networkDrawing.matrixEdgeGenerator)(size, nodes);

    case "arc":
      return (0, _networkDrawing.arcEdgeGenerator)(size);

    case "dagre":
      if (graph) return (0, _networkDrawing.dagreEdgeGenerator)(graph.graph().rankdir);
  }

  return undefined;
}

function breadthFirstCompontents(baseNodes, hash) {
  var componentHash = {
    "0": {
      componentNodes: [],
      componentEdges: []
    }
  };
  var components = [componentHash["0"]];
  var componentID = 0;
  traverseNodesBF(baseNodes, true);

  function traverseNodesBF(nodes, top) {
    nodes.forEach(function (node) {
      var hashNode = hash.get(node);

      if (!hashNode) {
        componentHash["0"].componentNodes.push(node);
      } else if (hashNode.component === -99) {
        var _componentHash$compon;

        if (top === true) {
          componentID++;
          componentHash[componentID] = {
            componentNodes: [],
            componentEdges: []
          };
          components.push(componentHash[componentID]);
        }

        hashNode.component = componentID;
        componentHash[componentID].componentNodes.push(node);

        (_componentHash$compon = componentHash[componentID].componentEdges).push.apply(_componentHash$compon, _toConsumableArray(hashNode.edges));

        var traversibleNodes = _toConsumableArray(hashNode.connectedNodes);

        traverseNodesBF(traversibleNodes, hash);
      }
    });
  }

  return components.sort(function (a, b) {
    return b.componentNodes.length - a.componentNodes.length;
  });
}

var projectedCoordinateNames = {
  y: "y",
  x: "x"
};
var sankeyOrientHash = {
  left: _d3SankeyCircular.sankeyLeft,
  right: _d3SankeyCircular.sankeyRight,
  center: _d3SankeyCircular.sankeyCenter,
  justify: _d3SankeyCircular.sankeyJustify
};
var xScale = (0, _d3Scale.scaleLinear)();
var yScale = (0, _d3Scale.scaleLinear)();

var matrixify = function matrixify(_ref2) {
  var edgeHash = _ref2.edgeHash,
      nodes = _ref2.nodes,
      edgeWidthAccessor = _ref2.edgeWidthAccessor,
      nodeIDAccessor = _ref2.nodeIDAccessor;
  var matrix = [];
  nodes.forEach(function (nodeSource) {
    var nodeSourceID = nodeIDAccessor(nodeSource);
    var sourceRow = [];
    matrix.push(sourceRow);
    nodes.forEach(function (nodeTarget) {
      var nodeTargetID = nodeIDAccessor(nodeTarget);
      var theEdge = edgeHash.get("".concat(nodeSourceID, "|").concat(nodeTargetID));

      if (theEdge) {
        sourceRow.push(edgeWidthAccessor(theEdge));
      } else {
        sourceRow.push(0);
      }
    });
  });
  return matrix;
};

var NetworkFrame =
/*#__PURE__*/
function (_React$Component) {
  _inherits(NetworkFrame, _React$Component);

  function NetworkFrame(props) {
    var _this;

    _classCallCheck(this, NetworkFrame);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(NetworkFrame).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "defaultNetworkSVGRule", function (_ref3) {
      var baseD = _ref3.d,
          i = _ref3.i,
          annotationLayer = _ref3.annotationLayer;
      var _this$state = _this.state,
          projectedNodes = _this$state.projectedNodes,
          projectedEdges = _this$state.projectedEdges,
          nodeIDAccessor = _this$state.nodeIDAccessor,
          nodeSizeAccessor = _this$state.nodeSizeAccessor,
          networkFrameRender = _this$state.networkFrameRender,
          adjustedSize = _this$state.adjustedSize,
          adjustedPosition = _this$state.adjustedPosition; //TODO PASS FRAME STYLE FNs TO HIGHLIGHT

      var svgAnnotationRules = _this.props.svgAnnotationRules;
      var d = baseD.ids ? baseD : baseD.edge ? _objectSpread({}, projectedEdges.find(function (p) {
        return nodeIDAccessor(p.source) === nodeIDAccessor(baseD.source) && nodeIDAccessor(p.target) === nodeIDAccessor(baseD.target);
      }) || {}, baseD) : _objectSpread({}, projectedNodes.find(function (p) {
        return nodeIDAccessor(p) === baseD.id;
      }) || {}, baseD);
      var voronoiHover = annotationLayer.voronoiHover;

      if (svgAnnotationRules) {
        var customAnnotation = svgAnnotationRules({
          d: d,
          i: i,
          networkFrameProps: _this.props,
          networkFrameState: _this.state,
          nodes: projectedNodes,
          edges: projectedEdges,
          voronoiHover: voronoiHover,
          screenCoordinates: [d.x, d.y],
          adjustedPosition: adjustedPosition,
          adjustedSize: adjustedSize,
          annotationLayer: annotationLayer
        });

        if (customAnnotation !== null) {
          return customAnnotation;
        }
      }

      if (d.type === "node") {
        return (0, _networkframeRules.svgNodeRule)({
          d: d,
          i: i,
          nodeSizeAccessor: nodeSizeAccessor
        });
      } else if (d.type === "desaturation-layer") {
        return (0, _baseRules.desaturationLayer)({
          style: d.style instanceof Function ? d.style(d, i) : d.style,
          size: adjustedSize,
          i: i,
          key: d.key
        });
      } else if (d.type === "basic-node-label") {
        return React.createElement("g", {
          key: d.key || "basic-".concat(i),
          transform: "translate(".concat(d.x, ",").concat(d.y, ")")
        }, baseD.element || baseD.label);
      } else if (d.type === "react-annotation" || typeof d.type === "function") {
        return (0, _networkframeRules.svgReactAnnotationRule)({
          d: d,
          i: i,
          projectedNodes: projectedNodes,
          nodeIDAccessor: nodeIDAccessor
        });
      } else if (d.type === "enclose") {
        return (0, _networkframeRules.svgEncloseRule)({
          d: d,
          i: i,
          projectedNodes: projectedNodes,
          nodeIDAccessor: nodeIDAccessor,
          nodeSizeAccessor: nodeSizeAccessor
        });
      } else if (d.type === "enclose-rect") {
        return (0, _networkframeRules.svgRectEncloseRule)({
          d: d,
          i: i,
          projectedNodes: projectedNodes,
          nodeIDAccessor: nodeIDAccessor,
          nodeSizeAccessor: nodeSizeAccessor
        });
      } else if (d.type === "enclose-hull") {
        return (0, _networkframeRules.svgHullEncloseRule)({
          d: d,
          i: i,
          projectedNodes: projectedNodes,
          nodeIDAccessor: nodeIDAccessor,
          nodeSizeAccessor: nodeSizeAccessor
        });
      } else if (d.type === "highlight") {
        return (0, _networkframeRules.svgHighlightRule)({
          d: d,
          networkFrameRender: networkFrameRender
        });
      }

      return null;
    });

    _defineProperty(_assertThisInitialized(_this), "defaultNetworkHTMLRule", function (_ref4) {
      var baseD = _ref4.d,
          i = _ref4.i,
          annotationLayer = _ref4.annotationLayer;
      var _this$props = _this.props,
          tooltipContent = _this$props.tooltipContent,
          size = _this$props.size,
          useSpans = _this$props.useSpans;
      var _this$state2 = _this.state,
          projectedNodes = _this$state2.projectedNodes,
          projectedEdges = _this$state2.projectedEdges,
          nodeIDAccessor = _this$state2.nodeIDAccessor,
          adjustedSize = _this$state2.adjustedSize,
          adjustedPosition = _this$state2.adjustedPosition;
      var voronoiHover = annotationLayer.voronoiHover;
      var d = baseD.ids ? baseD : baseD.edge ? _objectSpread({}, projectedEdges.find(function (p) {
        return nodeIDAccessor(p.source) === nodeIDAccessor(baseD.source) && nodeIDAccessor(p.target) === nodeIDAccessor(baseD.target);
      }) || {}, baseD) : _objectSpread({}, projectedNodes.find(function (p) {
        return nodeIDAccessor(p) === baseD.id;
      }) || {}, baseD);

      if (_this.props.htmlAnnotationRules) {
        var customAnnotation = _this.props.htmlAnnotationRules({
          d: d,
          i: i,
          networkFrameProps: _this.props,
          networkFrameState: _this.state,
          nodes: projectedNodes,
          edges: projectedEdges,
          voronoiHover: voronoiHover,
          screenCoordinates: [d.x, d.y],
          adjustedPosition: adjustedPosition,
          adjustedSize: adjustedSize,
          annotationLayer: annotationLayer
        });

        if (customAnnotation !== null) {
          return customAnnotation;
        }
      }

      if (d.type === "frame-hover") {
        return (0, _networkframeRules.htmlFrameHoverRule)({
          d: d,
          i: i,
          tooltipContent: tooltipContent,
          useSpans: useSpans,
          nodes: projectedNodes,
          edges: projectedEdges,
          nodeIDAccessor: nodeIDAccessor
        });
      }

      return null;
    });

    _this.state = {
      dataVersion: undefined,
      nodeData: [],
      edgeData: [],
      adjustedPosition: [],
      adjustedSize: [],
      backgroundGraphics: null,
      foregroundGraphics: null,
      projectedNodes: [],
      projectedEdges: [],
      renderNumber: 0,
      nodeLabelAnnotations: [],
      graphSettings: {
        type: "empty-start",
        nodes: [],
        edges: [],
        nodeHash: new Map(),
        edgeHash: new Map(),
        hierarchicalNetwork: false
      },
      edgeWidthAccessor: (0, _dataFunctions.stringToFn)("weight"),
      legendSettings: {},
      margin: {
        top: 0,
        left: 0,
        right: 0,
        bottom: 0
      },
      networkFrameRender: {},
      nodeIDAccessor: (0, _dataFunctions.stringToFn)("id"),
      nodeSizeAccessor: (0, _functions.genericFunction)(5),
      overlay: [],
      projectedXYPoints: [],
      sourceAccessor: (0, _dataFunctions.stringToFn)("source"),
      targetAccessor: (0, _dataFunctions.stringToFn)("target"),
      title: {
        title: undefined
      }
    };
    return _this;
  }

  _createClass(NetworkFrame, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.props.onUnmount) {
        this.props.onUnmount(this.props, this.state);
      }
    }
  }, {
    key: "componentWillMount",
    value: function componentWillMount() {
      Object.keys(this.props).forEach(function (d) {
        if (!_frame_props.networkframeproptypes[d]) {
          if (_frame_props.xyframeproptypes[d]) {
            console.error("".concat(d, " is an XYFrame prop are you sure you're using the right frame?"));
          } else if (_frame_props.ordinalframeproptypes[d]) {
            console.error("".concat(d, " is an OrdinalFrame prop are you sure you're using the right frame?"));
          } else {
            console.error("".concat(d, " is not a valid NetworkFrame prop"));
          }
        }
      });
      this.calculateNetworkFrame(this.props);
    }
  }, {
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      var _this2 = this;

      if (this.state.dataVersion && this.state.dataVersion !== nextProps.dataVersion || !this.state.projectedNodes && !this.state.projectedEdges) {
        this.calculateNetworkFrame(nextProps);
      } else if (this.props.size[0] !== nextProps.size[0] || this.props.size[1] !== nextProps.size[1] || !this.state.dataVersion && _frame_props.networkFrameChangeProps.find(function (d) {
        return _this2.props[d] !== nextProps[d];
      })) {
        this.calculateNetworkFrame(nextProps);
      }
    }
  }, {
    key: "onNodeClick",
    value: function onNodeClick(d, i) {
      if (this.props.onNodeClick) {
        this.props.onNodeClick(d, i);
      }
    }
  }, {
    key: "onNodeEnter",
    value: function onNodeEnter(d, i) {
      if (this.props.onNodeEnter) {
        this.props.onNodeEnter(d, i);
      }
    }
  }, {
    key: "onNodeOut",
    value: function onNodeOut(d, i) {
      if (this.props.onNodeOut) {
        this.props.onNodeOut(d, i);
      }
    }
  }, {
    key: "calculateNetworkFrame",
    value: function calculateNetworkFrame(currentProps) {
      var _this3 = this;

      var graph = currentProps.graph,
          _currentProps$nodes = currentProps.nodes,
          nodes = _currentProps$nodes === void 0 ? Array.isArray(graph) || typeof graph === "function" ? emptyArray : graph && graph.nodes || emptyArray : _currentProps$nodes,
          _currentProps$edges = currentProps.edges,
          edges = _currentProps$edges === void 0 ? typeof graph === "function" ? emptyArray : Array.isArray(graph) ? graph : graph && graph.edges || emptyArray : _currentProps$edges,
          networkType = currentProps.networkType,
          size = currentProps.size,
          nodeStyle = currentProps.nodeStyle,
          nodeClass = currentProps.nodeClass,
          canvasNodes = currentProps.canvasNodes,
          edgeStyle = currentProps.edgeStyle,
          edgeClass = currentProps.edgeClass,
          canvasEdges = currentProps.canvasEdges,
          nodeRenderMode = currentProps.nodeRenderMode,
          edgeRenderMode = currentProps.edgeRenderMode,
          nodeLabels = currentProps.nodeLabels,
          baseTitle = currentProps.title,
          baseMargin = currentProps.margin,
          hoverAnnotation = currentProps.hoverAnnotation,
          baseCustomNodeIcon = currentProps.customNodeIcon,
          baseCustomEdgeIcon = currentProps.customEdgeIcon,
          filterRenderedNodes = currentProps.filterRenderedNodes;
      var _edgeType = currentProps.edgeType;
      var networkSettings;
      var nodeHierarchicalIDFill = {};
      var networkSettingsKeys = ["type"];

      if (typeof networkType === "string") {
        networkSettings = _objectSpread({
          type: networkType
        }, baseNetworkSettings, {
          graphSettings: baseGraphSettings
        });
      } else {
        if (networkType) networkSettingsKeys = Object.keys(networkType);
        networkSettings = _objectSpread({
          type: "force"
        }, baseNetworkSettings, networkType, {
          graphSettings: baseGraphSettings
        });
      }

      if (networkSettings.projection === "vertical" && networkSettings.type === "sankey") {
        networkSettings.direction = "down";
      }

      networkSettingsKeys.push("height", "width");
      var title = _typeof(baseTitle) === "object" && !React.isValidElement(baseTitle) && baseTitle !== null ? baseTitle : {
        title: baseTitle,
        orient: "top"
      };
      var margin = (0, _frameFunctions.calculateMargin)({
        margin: baseMargin,
        title: title
      });

      var _adjustedPositionSize = (0, _frameFunctions.adjustedPositionSize)({
        size: size,
        margin: margin
      }),
          adjustedPosition = _adjustedPositionSize.adjustedPosition,
          adjustedSize = _adjustedPositionSize.adjustedSize;

      networkSettings.graphSettings.nodes = nodes;
      networkSettings.graphSettings.edges = edges;
      var _networkSettings$grap = networkSettings.graphSettings,
          edgeHash = _networkSettings$grap.edgeHash,
          nodeHash = _networkSettings$grap.nodeHash;
      var createPointLayer = networkSettings.type === "treemap" || networkSettings.type === "partition" || networkSettings.type === "sankey";
      var nodeIDAccessor = (0, _dataFunctions.stringToFn)(currentProps.nodeIDAccessor, function (d) {
        return d.id;
      });
      var sourceAccessor = (0, _dataFunctions.stringToFn)(currentProps.sourceAccessor, function (d) {
        return d.source;
      });
      var targetAccessor = (0, _dataFunctions.stringToFn)(currentProps.targetAccessor, function (d) {
        return d.target;
      });
      var nodeSizeAccessor = typeof currentProps.nodeSizeAccessor === "number" ? (0, _functions.genericFunction)(currentProps.nodeSizeAccessor) : (0, _dataFunctions.stringToFn)(currentProps.nodeSizeAccessor, function (d) {
        return d.r || 5;
      });
      var edgeWidthAccessor = (0, _dataFunctions.stringToFn)(currentProps.edgeWidthAccessor, function (d) {
        return d.weight || 1;
      });
      var nodeStyleFn = (0, _dataFunctions.stringToFn)(nodeStyle, function () {
        return {};
      }, true);
      var nodeClassFn = (0, _dataFunctions.stringToFn)(nodeClass, function () {
        return "";
      }, true);
      var nodeRenderModeFn = (0, _dataFunctions.stringToFn)(nodeRenderMode, undefined, true);
      var nodeCanvasRenderFn = canvasNodes && (0, _dataFunctions.stringToFn)(canvasNodes, undefined, true);
      var _this$state3 = this.state,
          projectedNodes = _this$state3.projectedNodes,
          projectedEdges = _this$state3.projectedEdges;
      var isHierarchical = typeof networkSettings.type === "string" && hierarchicalTypeHash[networkSettings.type];
      var changedData = !this.state.projectedNodes || !this.state.projectedEdges || this.state.graphSettings.nodes !== nodes || this.state.graphSettings.edges !== edges || isHierarchical;

      if (networkSettings.type === "dagre") {
        var dagreGraph = graph;
        var dagreNodeHash = {};
        projectedNodes = dagreGraph.nodes().map(function (n) {
          var baseNode = dagreGraph.node(n);
          dagreNodeHash[n] = _objectSpread({}, baseNode, {
            x0: baseNode.x - baseNode.width / 2,
            x1: baseNode.x + baseNode.width / 2,
            y0: baseNode.y - baseNode.height / 2,
            y1: baseNode.y + baseNode.height / 2,
            id: n,
            shapeNode: true,
            sourceLinks: [],
            targetLinks: []
          });
          return dagreNodeHash[n];
        });
        projectedEdges = dagreGraph.edges().map(function (e) {
          var dagreEdge = dagreGraph.edge(e);

          var baseEdge = _objectSpread({}, dagreEdge, {
            points: dagreEdge.points.map(function (d) {
              return _objectSpread({}, d);
            })
          });

          baseEdge.source = projectedNodes.find(function (p) {
            return p.id === e.v;
          });
          baseEdge.target = projectedNodes.find(function (p) {
            return p.id === e.w;
          });
          baseEdge.points.unshift({
            x: baseEdge.source.x,
            y: baseEdge.source.y
          });
          baseEdge.points.push({
            x: baseEdge.target.x,
            y: baseEdge.target.y
          });
          dagreNodeHash[e.v].targetLinks.push(baseEdge);
          dagreNodeHash[e.w].sourceLinks.push(baseEdge);
          return baseEdge;
        });
      } else if (changedData) {
        edgeHash = new Map();
        nodeHash = new Map();
        networkSettings.graphSettings.edgeHash = edgeHash;
        networkSettings.graphSettings.nodeHash = nodeHash;
        projectedNodes = [];
        projectedEdges = [];
        nodes.forEach(function (node) {
          var projectedNode = _objectSpread({}, node);

          var id = nodeIDAccessor(projectedNode);
          nodeHash.set(id, projectedNode);
          nodeHash.set(node, projectedNode);
          projectedNodes.push(projectedNode);
          projectedNode.id = id;
          projectedNode.inDegree = 0;
          projectedNode.outDegree = 0;
          projectedNode.degree = 0;
        });
        var operationalEdges = edges;
        var baseEdges = edges;

        if (isHierarchical && Array.isArray(edges)) {
          var createdHierarchicalData = (0, _networkDrawing.softStack)(edges, projectedNodes, sourceAccessor, targetAccessor, nodeIDAccessor);

          if (createdHierarchicalData.isHierarchical) {
            baseEdges = createdHierarchicalData.hierarchy;
            projectedNodes = [];
          } else {
            console.error("You've sent an edge list that is not strictly hierarchical (there are nodes with multiple parents) defaulting to force-directed network layout");
            networkSettings.type = "force";
          }
        }

        if (!Array.isArray(baseEdges)) {
          networkSettings.hierarchicalNetwork = true;
          var rootNode = (0, _d3Hierarchy.hierarchy)(baseEdges, networkSettings.hierarchyChildren);
          rootNode.sum(networkSettings.hierarchySum || function (d) {
            return d.value;
          });

          if (isHierarchical) {
            var layout = networkSettings.layout || isHierarchical;
            var hierarchicalLayout = layout();
            var networkSettingKeys = Object.keys(networkSettings);

            if ((networkSettings.type === "dendrogram" || networkSettings.type === "tree" || networkSettings.type === "cluster") && hierarchicalLayout.separation) {
              hierarchicalLayout.separation(function (a, b) {
                return (nodeSizeAccessor(_objectSpread({}, a, a.data)) || 1) + (networkSettings.nodePadding || 0) + (nodeSizeAccessor(_objectSpread({}, b, b.data)) || 1);
              });
            }

            networkSettingKeys.forEach(function (key) {
              if (hierarchicalLayout[key]) {
                hierarchicalLayout[key](networkSettings[key]);
              }
            });
            var layoutSize = networkSettings.projection === "horizontal" && isHierarchical ? [adjustedSize[1], adjustedSize[0]] : adjustedSize;

            if (!networkSettings.nodeSize && hierarchicalLayout.size) {
              hierarchicalLayout.size(layoutSize);
            }

            hierarchicalLayout(rootNode);
          }

          operationalEdges = (0, _network.nodesEdgesFromHierarchy)(rootNode, nodeIDAccessor).edges;
        }

        baseNodeProps.shapeNode = createPointLayer;

        if (Array.isArray(operationalEdges)) {
          operationalEdges.forEach(function (edge) {
            var source = sourceAccessor(edge);
            var target = targetAccessor(edge);
            var sourceTarget = [source, target];
            sourceTarget.forEach(function (nodeDirection) {
              if (!nodeHash.get(nodeDirection)) {
                var nodeObject = _typeof(nodeDirection) === "object" ? _objectSpread({}, baseNodeProps, nodeDirection) : _objectSpread({}, baseNodeProps, {
                  id: nodeDirection,
                  createdByFrame: true
                });
                var nodeIDValue = nodeObject.id || nodeIDAccessor(nodeObject);
                nodeHierarchicalIDFill[nodeIDValue] ? nodeHierarchicalIDFill[nodeIDValue] += 1 : nodeHierarchicalIDFill[nodeIDValue] = 1;

                if (!nodeObject.id) {
                  var nodeSuffix = nodeHierarchicalIDFill[nodeIDValue] === 1 ? "" : "-".concat(nodeHierarchicalIDFill[nodeIDValue]);
                  nodeObject.id = "".concat(nodeIDValue).concat(nodeSuffix);
                }

                nodeHash.set(nodeDirection, nodeObject);
                projectedNodes.push(nodeObject);
              }
            });
            var edgeWeight = edge.weight || 1;
            var sourceNode = nodeHash.get(source);
            var targetNode = nodeHash.get(target);
            targetNode.inDegree += edgeWeight;
            sourceNode.outDegree += edgeWeight;
            targetNode.degree += edgeWeight;
            sourceNode.degree += edgeWeight;
            var edgeKey = "".concat(nodeIDAccessor(sourceNode) || source, "|").concat(nodeIDAccessor(targetNode) || target);

            var newEdge = _extends({}, edge, {
              source: nodeHash.get(source),
              target: nodeHash.get(target)
            });

            edgeHash.set(edgeKey, newEdge);
            projectedEdges.push(newEdge);
          });
        }
      } else {
        edgeHash = new Map();
        networkSettings.graphSettings.edgeHash = edgeHash;
        projectedEdges.forEach(function (edge) {
          var edgeSource = typeof edge.source === "string" ? edge.source : nodeIDAccessor(edge.source);
          var edgeTarget = typeof edge.target === "string" ? edge.target : nodeIDAccessor(edge.target);
          var edgeKey = "".concat(edgeSource, "|").concat(edgeTarget);
          edgeHash.set(edgeKey, edge);
        });
      }

      var customNodeIcon = determineNodeIcon(baseCustomNodeIcon, networkSettings, adjustedSize, projectedNodes);
      var customEdgeIcon = determineEdgeIcon({
        baseCustomEdgeIcon: baseCustomEdgeIcon,
        networkSettings: networkSettings,
        size: adjustedSize,
        nodes: projectedNodes,
        graph: graph
      });

      if ((networkSettings.type === "sankey" || networkSettings.type === "flowchart") && (0, _networkDrawing.topologicalSort)(projectedNodes, projectedEdges) === null) {
        networkSettings.customSankey = _d3SankeyCircular.sankeyCircular;
      }

      networkSettings.width = size[0];
      networkSettings.height = size[1];
      var networkSettingsChanged = false;
      networkSettingsKeys.forEach(function (key) {
        if (key !== "edgeType" && key !== "graphSettings" && networkSettings[key] !== _this3.state.graphSettings[key]) {
          networkSettingsChanged = true;
        }
      }); //Support bubble chart with circle pack and with force

      if (networkSettings.type === "sankey") {
        _edgeType = function edgeType(d) {
          return d.circular ? (0, _networkDrawing.circularAreaLink)(d) : _edgeType === "angled" ? (0, _networkDrawing.ribbonLink)(d) : (0, _networkDrawing.areaLink)(d);
        };
      } else if (isHierarchical) {
        projectedNodes.forEach(function (node) {
          if (createPointLayer) {
            node.x = (node.x0 + node.x1) / 2;
            node.y = (node.y0 + node.y1) / 2;
          }

          if (typeof networkSettings.type === "string" && hierarchicalProjectable[networkSettings.type] && networkSettings.projection === "horizontal") {
            var ox = node.x;
            node.x = node.y;
            node.y = ox;

            if (createPointLayer) {
              var ox0 = node.x0;
              var ox1 = node.x1;
              node.x0 = node.y0;
              node.x1 = node.y1;
              node.y0 = ox0;
              node.y1 = ox1;
            }
          } else if (typeof networkSettings.type === "string" && radialProjectable[networkSettings.type] && networkSettings.projection === "radial") {
            var radialPoint = node.depth === 0 ? [adjustedSize[0] / 2, adjustedSize[1] / 2] : (0, _pieceDrawing.pointOnArcAtAngle)([adjustedSize[0] / 2, adjustedSize[1] / 2], node.x / adjustedSize[0], node.y / 2);
            node.x = radialPoint[0];
            node.y = radialPoint[1];
          } else {
            node.x = node.x;
            node.y = node.y;

            if (createPointLayer) {
              node.x0 = node.x0;
              node.x1 = node.x1;
              node.y0 = node.y0;
              node.y1 = node.y1;
            }
          }
        });
      }

      if (networkSettings.type !== "static" && (changedData || networkSettingsChanged)) {
        var components = [{
          componentNodes: projectedNodes,
          componentEdges: projectedEdges
        }];

        if (networkSettings.type === "chord") {
          var radius = adjustedSize[1] / 2;
          var _networkSettings = networkSettings,
              _networkSettings$grou = _networkSettings.groupWidth,
              groupWidth = _networkSettings$grou === void 0 ? 20 : _networkSettings$grou,
              _networkSettings$padA = _networkSettings.padAngle,
              padAngle = _networkSettings$padA === void 0 ? 0.01 : _networkSettings$padA,
              sortGroups = _networkSettings.sortGroups;
          var arcGenerator = (0, _d3Shape.arc)().innerRadius(radius - groupWidth).outerRadius(radius);
          var ribbonGenerator = (0, _d3Chord.ribbon)().radius(radius - groupWidth);
          var matrixifiedNetwork = matrixify({
            edgeHash: edgeHash,
            nodes: projectedNodes,
            edgeWidthAccessor: edgeWidthAccessor,
            nodeIDAccessor: nodeIDAccessor
          });
          var chordLayout = (0, _d3Chord.chord)().padAngle(padAngle);

          if (sortGroups) {
            chordLayout.sortGroups(sortGroups);
          }

          var chords = chordLayout(matrixifiedNetwork);
          var groups = chords.groups;
          groups.forEach(function (group) {
            var groupCentroid = arcGenerator.centroid(group);
            var groupD = arcGenerator(group);
            var groupNode = projectedNodes[group.index];
            groupNode.d = groupD;
            groupNode.index = group.index;
            groupNode.x = groupCentroid[0] + adjustedSize[0] / 2;
            groupNode.y = groupCentroid[1] + adjustedSize[1] / 2;
          });
          chords.forEach(function (generatedChord) {
            var chordD = ribbonGenerator(generatedChord); //this is incorrect should use edgeHash

            var nodeSourceID = nodeIDAccessor(projectedNodes[generatedChord.source.index]);
            var nodeTargetID = nodeIDAccessor(projectedNodes[generatedChord.target.index]);
            var chordEdge = edgeHash.get("".concat(nodeSourceID, "|").concat(nodeTargetID));
            chordEdge.d = chordD;
            var chordBounds = (0, _svgPathBoundingBox.default)(chordD);
            chordEdge.x = adjustedSize[0] / 2 + (chordBounds.x1 + chordBounds.x2) / 2;
            chordEdge.y = adjustedSize[1] / 2 + (chordBounds.y1 + chordBounds.y2) / 2;
          });
        } else if (networkSettings.type === "sankey" || networkSettings.type === "flowchart") {
          var _networkSettings2 = networkSettings,
              _networkSettings2$ori = _networkSettings2.orient,
              orient = _networkSettings2$ori === void 0 ? "center" : _networkSettings2$ori,
              _networkSettings2$ite = _networkSettings2.iterations,
              iterations = _networkSettings2$ite === void 0 ? 100 : _networkSettings2$ite,
              nodePadding = _networkSettings2.nodePadding,
              _networkSettings2$nod = _networkSettings2.nodePaddingRatio,
              nodePaddingRatio = _networkSettings2$nod === void 0 ? nodePadding ? undefined : 0.5 : _networkSettings2$nod,
              _networkSettings2$nod2 = _networkSettings2.nodeWidth,
              nodeWidth = _networkSettings2$nod2 === void 0 ? networkSettings.type === "flowchart" ? 2 : 24 : _networkSettings2$nod2,
              customSankey = _networkSettings2.customSankey,
              _networkSettings2$dir = _networkSettings2.direction,
              direction = _networkSettings2$dir === void 0 ? "right" : _networkSettings2$dir;
          var sankeyOrient = sankeyOrientHash[orient];
          var actualSankey = customSankey || _d3SankeyCircular.sankeyCircular;
          var frameExtent = [[0, 0], adjustedSize];

          if (networkSettings.direction === "up" || networkSettings.direction === "down") {
            frameExtent = [[0, 0], [adjustedSize[1], adjustedSize[0]]];
          }

          var frameSankey = actualSankey().extent(frameExtent).links(projectedEdges).nodes(projectedNodes).nodeAlign(sankeyOrient).nodeId(nodeIDAccessor).nodeWidth(nodeWidth).iterations(iterations);

          if (frameSankey.nodePaddingRatio && nodePaddingRatio) {
            frameSankey.nodePaddingRatio(nodePaddingRatio);
          } else if (nodePadding) {
            frameSankey.nodePadding(nodePadding);
          }

          frameSankey();
          projectedNodes.forEach(function (d) {
            d.height = d.y1 - d.y0;
            d.width = d.x1 - d.x0;
            d.x = d.x0 + d.width / 2;
            d.y = d.y0 + d.height / 2;
            d.radius = d.height / 2;
            d.direction = direction;
          });
          projectedEdges.forEach(function (d) {
            d.sankeyWidth = d.width;
            d.direction = direction;
            d.width = undefined;
          });
        } else if (networkSettings.type === "wordcloud") {
          var _networkSettings3 = networkSettings,
              _networkSettings3$ite = _networkSettings3.iterations,
              _iterations = _networkSettings3$ite === void 0 ? 500 : _networkSettings3$ite,
              _networkSettings3$fon = _networkSettings3.fontSize,
              fontSize = _networkSettings3$fon === void 0 ? 18 : _networkSettings3$fon,
              rotate = _networkSettings3.rotate,
              _networkSettings3$fon2 = _networkSettings3.fontWeight,
              fontWeight = _networkSettings3$fon2 === void 0 ? 300 : _networkSettings3$fon2,
              _networkSettings3$tex = _networkSettings3.textAccessor,
              textAccessor = _networkSettings3$tex === void 0 ? function (d) {
            return d.text;
          } : _networkSettings3$tex;

          var fontWeightMod = (fontWeight / 300 - 1) / 5 + 1;
          var fontWidth = fontSize / 1.5 * fontWeightMod;
          projectedNodes.forEach(function (d, i) {
            var calcualatedNodeSize = nodeSizeAccessor(d);
            d._NWFText = textAccessor(d) || "";
            var textWidth = fontWidth * d._NWFText.length * calcualatedNodeSize * 1.4;
            var textHeight = fontSize * calcualatedNodeSize;
            d.textHeight = textHeight + 4;
            d.textWidth = textWidth + 4;
            d.rotate = rotate ? rotate(d, i) : 0;
            d.fontSize = fontSize * calcualatedNodeSize;
            d.fontWeight = fontWeight;
            d.radius = d.r = textWidth / 2;
          });
          projectedNodes.sort(function (a, b) {
            return b.textWidth - a.textWidth;
          }); //bubblepack for initial position

          (0, _d3Hierarchy.packSiblings)(projectedNodes); //        if (rotate) {

          var collide = (0, _d3BboxCollide.bboxCollide)(function (d) {
            if (d.rotate) {
              return [[-d.textHeight / 2, -d.textWidth / 2], [d.textHeight / 2, d.textWidth / 2]];
            }

            return [[-d.textWidth / 2, -d.textHeight / 2], [d.textWidth / 2, d.textHeight / 2]];
          }).iterations(1);
          var xCenter = size[0] / 2;
          var yCenter = size[1] / 2;
          var simulation = (0, _d3Force.forceSimulation)(projectedNodes).velocityDecay(0.6).force("x", (0, _d3Force.forceX)(xCenter).strength(1.2)).force("y", (0, _d3Force.forceY)(yCenter).strength(1.2)).force("collide", collide);
          simulation.stop();

          for (var i = 0; i < _iterations; ++i) {
            simulation.tick();
          } //      }


          var xMin = (0, _d3Array.min)(projectedNodes.map(function (p) {
            return p.x - (p.rotate ? p.textHeight / 2 : p.textWidth / 2);
          }));
          var xMax = (0, _d3Array.max)(projectedNodes.map(function (p) {
            return p.x + (p.rotate ? p.textHeight / 2 : p.textWidth / 2);
          }));
          var yMin = (0, _d3Array.min)(projectedNodes.map(function (p) {
            return p.y - (p.rotate ? p.textWidth / 2 : p.textHeight / 2);
          }));
          var yMax = (0, _d3Array.max)(projectedNodes.map(function (p) {
            return p.y + (p.rotate ? p.textWidth / 2 : p.textHeight / 2);
          }));
          var projectionScaleX = (0, _d3Scale.scaleLinear)().domain([xMin, xMax]).range([0, adjustedSize[0]]);
          var projectionScaleY = (0, _d3Scale.scaleLinear)().domain([yMin, yMax]).range([0, adjustedSize[1]]);
          var xMod = adjustedSize[0] / xMax;
          var yMod = adjustedSize[1] / yMax;
          var sizeMod = Math.min(xMod, yMod) * 1.2;
          projectedNodes.forEach(function (node) {
            node.x = projectionScaleX(node.x);
            node.y = projectionScaleY(node.y);
            node.fontSize = node.fontSize * sizeMod;
            node.scale = 1;
            node.radius = node.r = Math.max(node.textHeight / 4 * yMod, node.textWidth / 4 * xMod); //      node.textHeight = projectionScaleY(node.textHeight)
            //      node.textWidth = projectionScaleY(node.textWidth)
          });
        } else if (networkSettings.type === "force") {
          var _networkSettings4 = networkSettings,
              _networkSettings4$ite = _networkSettings4.iterations,
              _iterations2 = _networkSettings4$ite === void 0 ? 500 : _networkSettings4$ite,
              _networkSettings4$edg = _networkSettings4.edgeStrength,
              edgeStrength = _networkSettings4$edg === void 0 ? 0.1 : _networkSettings4$edg,
              _networkSettings4$dis = _networkSettings4.distanceMax,
              distanceMax = _networkSettings4$dis === void 0 ? Infinity : _networkSettings4$dis,
              edgeDistance = _networkSettings4.edgeDistance;

          var linkForce = (0, _d3Force.forceLink)().strength(function (d) {
            return Math.min(2.5, d.weight ? d.weight * edgeStrength : edgeStrength);
          });

          if (edgeDistance) {
            linkForce.distance(edgeDistance);
          }

          var _simulation = networkSettings.simulation || (0, _d3Force.forceSimulation)().force("charge", (0, _d3Force.forceManyBody)().distanceMax(distanceMax).strength(networkSettings.forceManyBody || function (d) {
            return -25 * nodeSizeAccessor(d);
          })); //        simulation.force("link", linkForce).nodes(projectedNodes)


          _simulation.nodes(projectedNodes);

          var forceMod = adjustedSize[1] / adjustedSize[0];

          if (!_simulation.force("x")) {
            _simulation.force("x", (0, _d3Force.forceX)(adjustedSize[0] / 2).strength(forceMod * 0.1));
          }

          if (!_simulation.force("y")) {
            _simulation.force("y", (0, _d3Force.forceY)(adjustedSize[1] / 2).strength(0.1));
          }

          if (projectedEdges.length !== 0 && !_simulation.force("link")) {
            _simulation.force("link", linkForce);

            _simulation.force("link").links(projectedEdges);
          } //reset alpha if it's too cold


          if (_simulation.alpha() < 0.1) {
            _simulation.alpha(1);
          }

          _simulation.stop();

          for (var _i = 0; _i < _iterations2; ++_i) {
            _simulation.tick();
          }
        } else if (networkSettings.type === "motifs") {
          var componentHash = new Map();
          projectedEdges.forEach(function (edge) {
            ;
            [edge.source, edge.target].forEach(function (node) {
              if (!componentHash.get(node)) {
                componentHash.set(node, {
                  node: node,
                  component: -99,
                  connectedNodes: [],
                  edges: []
                });
              }
            });
            componentHash.get(edge.source).connectedNodes.push(edge.target);
            componentHash.get(edge.target).connectedNodes.push(edge.source);
            componentHash.get(edge.source).edges.push(edge);
          });
          components = breadthFirstCompontents(projectedNodes, componentHash);
          var largestComponent = Math.max(projectedNodes.length / 3, components[0].componentNodes.length);

          var _layoutSize = size[0] > size[1] ? size[1] : size[0];

          var layoutDirection = size[0] > size[1] ? "horizontal" : "vertical";

          var _networkSettings5 = networkSettings,
              _networkSettings5$ite = _networkSettings5.iterations,
              _iterations3 = _networkSettings5$ite === void 0 ? 500 : _networkSettings5$ite,
              _networkSettings5$edg = _networkSettings5.edgeStrength,
              _edgeStrength = _networkSettings5$edg === void 0 ? 0.1 : _networkSettings5$edg,
              _edgeDistance = _networkSettings5.edgeDistance;

          var currentX = 0;
          var currentY = 0;
          components.forEach(function (_ref5) {
            var componentNodes = _ref5.componentNodes,
                componentEdges = _ref5.componentEdges;
            var linkForce = (0, _d3Force.forceLink)().strength(function (d) {
              return Math.min(2.5, d.weight ? d.weight * _edgeStrength : _edgeStrength);
            });

            if (_edgeDistance) {
              linkForce.distance(_edgeDistance);
            }

            var componentLayoutSize = Math.max(componentNodes.length / largestComponent, 0.2) * _layoutSize;

            var xBound = componentLayoutSize + currentX;
            var yBound = componentLayoutSize + currentY;

            if (layoutDirection === "horizontal") {
              if (yBound > size[1]) {
                currentX = componentLayoutSize + currentX;
                currentY = componentLayoutSize;
              } else {
                currentY = componentLayoutSize + currentY;
              }
            } else {
              if (xBound > size[0]) {
                currentY = componentLayoutSize + currentY;
                currentX = componentLayoutSize;
              } else {
                currentX = componentLayoutSize + currentX;
              }
            }

            var xCenter = currentX - componentLayoutSize / 2;
            var yCenter = currentY - componentLayoutSize / 2;
            var simulation = (0, _d3Force.forceSimulation)().force("charge", (0, _d3Force.forceManyBody)().strength(networkSettings.forceManyBody || function (d) {
              return -25 * nodeSizeAccessor(d);
            })).force("link", linkForce);
            simulation.force("x", (0, _d3Force.forceX)(xCenter)).force("y", (0, _d3Force.forceY)(yCenter)).nodes(componentNodes);
            simulation.force("link").links(componentEdges);
            simulation.stop();

            for (var _i2 = 0; _i2 < _iterations3; ++_i2) {
              simulation.tick();
            }

            var maxX = (0, _d3Array.max)(componentNodes.map(function (d) {
              return d.x;
            }));
            var maxY = (0, _d3Array.max)(componentNodes.map(function (d) {
              return d.y;
            }));
            var minX = (0, _d3Array.min)(componentNodes.map(function (d) {
              return d.x;
            }));
            var minY = (0, _d3Array.min)(componentNodes.map(function (d) {
              return d.y;
            }));
            var resetX = (0, _d3Scale.scaleLinear)().domain([minX, maxX]).range([currentX - componentLayoutSize, currentX - 20]);
            var resetY = (0, _d3Scale.scaleLinear)().domain([minY, maxY]).range([currentY - componentLayoutSize, currentY - 20]);
            componentNodes.forEach(function (node) {
              node.x = resetX(node.x);
              node.y = resetY(node.y);
            });
          });
        } else if (networkSettings.type === "matrix") {
          if (networkSettings.sort) {
            projectedNodes = projectedNodes.sort(networkSettings.sort);
          }

          var gridSize = Math.min.apply(Math, _toConsumableArray(adjustedSize));
          var stepSize = gridSize / (projectedNodes.length + 1);
          projectedNodes.forEach(function (node, index) {
            node.x = 0;
            node.y = (index + 1) * stepSize;
          });
        } else if (networkSettings.type === "arc") {
          if (networkSettings.sort) {
            projectedNodes = projectedNodes.sort(networkSettings.sort);
          }

          var _stepSize = adjustedSize[0] / (projectedNodes.length + 2);

          projectedNodes.forEach(function (node, index) {
            node.x = (index + 1) * _stepSize;
            node.y = adjustedSize[1] / 2;
          });
        } else if (typeof networkSettings.type === "function") {
          networkSettings.type({
            nodes: projectedNodes,
            edges: projectedEdges
          });
        } else {
          projectedNodes.forEach(function (node) {
            node.x = node.x === undefined ? (node.x0 + node.x1) / 2 : node.x;
            node.y = node.y === undefined ? node.y0 : node.y;
          });
        }

        this.state.graphSettings.nodes = currentProps.nodes;
        this.state.graphSettings.edges = currentProps.edges;
      } //filter out user-defined nodes


      projectedNodes = projectedNodes.filter(filterRenderedNodes);
      projectedEdges = projectedEdges.filter(function (d) {
        return projectedNodes.indexOf(d.target) !== -1 && projectedNodes.indexOf(d.source) !== -1;
      });

      if (networkSettings.direction === "flip") {
        projectedNodes.forEach(function (node) {
          // const ox = node.x
          // const oy = node.y
          node.x = adjustedSize[0] - node.x;
          node.y = adjustedSize[1] - node.y;
        });
      } else if (networkSettings.direction === "up" || networkSettings.direction === "down") {
        var mod = networkSettings.direction === "up" ? function (value) {
          return adjustedSize[1] - value;
        } : function (value) {
          return value;
        };
        projectedNodes.forEach(function (node) {
          var ox = node.x;
          var ox0 = node.x0;
          var ox1 = node.x1;
          node.x = mod(node.y);
          node.x0 = mod(node.y0);
          node.x1 = mod(node.y1);
          node.y = ox;
          node.y0 = ox0;
          node.y1 = ox1;
        });
      } else if (networkSettings.direction === "left") {
        projectedNodes.forEach(function (node) {
          node.x = adjustedSize[0] - node.x;
          node.x0 = adjustedSize[0] - node.x0;
          node.x1 = adjustedSize[0] - node.x1;
        });
      }

      if (typeof networkSettings.zoom === "function") {
        networkSettings.zoom(projectedNodes, adjustedSize);
      } else if (networkSettings.zoom !== false && networkSettings.type !== "matrix" && networkSettings.type !== "wordcloud" && networkSettings.type !== "chord" && networkSettings.type !== "sankey" && networkSettings.type !== "partition" && networkSettings.type !== "treemap" && networkSettings.type !== "circlepack" && networkSettings.type !== "dagre") {
        // ZOOM SHOULD MAINTAIN ASPECT RATIO, ADD "stretch" to fill whole area
        var _xMin = (0, _d3Array.min)(projectedNodes.map(function (p) {
          return p.x - nodeSizeAccessor(p);
        }));

        var _xMax = (0, _d3Array.max)(projectedNodes.map(function (p) {
          return p.x + nodeSizeAccessor(p);
        }));

        var _yMin = (0, _d3Array.min)(projectedNodes.map(function (p) {
          return p.y - nodeSizeAccessor(p);
        }));

        var _yMax = (0, _d3Array.max)(projectedNodes.map(function (p) {
          return p.y + nodeSizeAccessor(p);
        }));

        var xSize = Math.abs(_xMax - _xMin);
        var ySize = Math.abs(_yMax - _yMin);
        var networkAspectRatio = xSize / ySize;
        var baseAspectRatio = adjustedSize[0] / adjustedSize[1];

        var _yMod, _xMod;

        if (networkSettings.zoom === "stretch") {
          _yMod = 0;
          _xMod = 0;
        } else if (xSize > ySize) {
          if (networkAspectRatio > baseAspectRatio) {
            _xMod = 0;
            _yMod = (adjustedSize[1] - adjustedSize[0] / xSize * ySize) / 2;
          } else {
            _yMod = 0;
            _xMod = (adjustedSize[0] - adjustedSize[1] / ySize * xSize) / 2;
          }
        } else {
          if (networkAspectRatio > baseAspectRatio) {
            _xMod = 0;
            _yMod = (adjustedSize[1] - adjustedSize[0] / xSize * ySize) / 2;
          } else {
            _yMod = 0;
            _xMod = (adjustedSize[0] - adjustedSize[1] / ySize * xSize) / 2;
          }
        }

        var _projectionScaleX = (0, _d3Scale.scaleLinear)().domain([_xMin, _xMax]).range([_xMod, adjustedSize[0] - _xMod]);

        var _projectionScaleY = (0, _d3Scale.scaleLinear)().domain([_yMin, _yMax]).range([_yMod, adjustedSize[1] - _yMod]);

        projectedNodes.forEach(function (node) {
          node.x = _projectionScaleX(node.x);
          node.y = _projectionScaleY(node.y);
        });
      } else if (networkSettings.zoom !== false && networkSettings.projection !== "radial" && (networkSettings.type === "partition" || networkSettings.type === "treemap" || networkSettings.type === "dagre")) {
        var _xMin2 = (0, _d3Array.min)(projectedNodes.map(function (p) {
          return p.x0;
        }));

        var _xMax2 = (0, _d3Array.max)(projectedNodes.map(function (p) {
          return p.x1;
        }));

        var _yMin2 = (0, _d3Array.min)(projectedNodes.map(function (p) {
          return p.y0;
        }));

        var _yMax2 = (0, _d3Array.max)(projectedNodes.map(function (p) {
          return p.y1;
        }));

        var _projectionScaleX2 = (0, _d3Scale.scaleLinear)().domain([_xMin2, _xMax2]).range([margin.left, adjustedSize[0] - margin.right]);

        var _projectionScaleY2 = (0, _d3Scale.scaleLinear)().domain([_yMin2, _yMax2]).range([margin.top, adjustedSize[1] - margin.bottom]);

        projectedNodes.forEach(function (node) {
          node.x = _projectionScaleX2(node.x);
          node.y = _projectionScaleY2(node.y);
          node.x0 = _projectionScaleX2(node.x0);
          node.y0 = _projectionScaleY2(node.y0);
          node.x1 = _projectionScaleX2(node.x1);
          node.y1 = _projectionScaleY2(node.y1);
          node.zoomedHeight = node.y1 - node.y0;
          node.zoomedWidth = node.x1 - node.x0;
        });
        projectedEdges.forEach(function (edge) {
          if (edge.points) {
            edge.points.forEach(function (p) {
              p.x = _projectionScaleX2(p.x);
              p.y = _projectionScaleY2(p.y);
            });
          }
        });
      }

      projectedNodes.forEach(function (node) {
        node.nodeSize = nodeSizeAccessor(node);
      });
      projectedEdges.forEach(function (edge) {
        edge.width = edgeWidthAccessor(edge);
      });
      var legendSettings;

      if (currentProps.legend) {
        legendSettings = currentProps.legend;

        if (!legendSettings.legendGroups) {
          ///Something auto for networks
          var legendGroups = [{
            styleFn: currentProps.nodeStyle,
            type: "fill",
            items: []
          }];
          legendSettings.legendGroups = legendGroups;
        }
      }

      var networkFrameRender = {
        edges: {
          accessibleTransform: function accessibleTransform(data, i) {
            var edgeX = (data[i].source.x + data[i].target.x) / 2;
            var edgeY = (data[i].source.y + data[i].target.y) / 2;
            return _objectSpread({
              type: "frame-hover"
            }, data[i], {
              x: edgeX,
              y: edgeY
            });
          },
          data: projectedEdges,
          styleFn: (0, _dataFunctions.stringToFn)(edgeStyle, function () {
            return {};
          }, true),
          classFn: (0, _dataFunctions.stringToFn)(edgeClass, function () {
            return "";
          }, true),
          renderMode: (0, _dataFunctions.stringToFn)(edgeRenderMode, undefined, true),
          canvasRenderFn: canvasEdges && (0, _dataFunctions.stringToFn)(canvasEdges, undefined, true),
          renderKeyFn: currentProps.edgeRenderKey ? currentProps.edgeRenderKey : function (d) {
            return d._NWFEdgeKey || "".concat(d.source.id, "-").concat(d.target.id);
          },
          behavior: _networkDrawing.drawEdges,
          projection: networkSettings.projection,
          type: _edgeType,
          customMark: customEdgeIcon,
          networkType: networkSettings.type,
          direction: networkSettings.direction
        },
        nodes: {
          accessibleTransform: function accessibleTransform(data, i) {
            return _objectSpread({
              type: "frame-hover"
            }, data[i], data[i].data || {});
          },
          data: projectedNodes,
          styleFn: nodeStyleFn,
          classFn: nodeClassFn,
          renderMode: nodeRenderModeFn,
          canvasRenderFn: nodeCanvasRenderFn,
          customMark: customNodeIcon,
          behavior: _networkDrawing.drawNodes,
          renderKeyFn: currentProps.nodeRenderKey
        }
      };
      var nodeLabelAnnotations = [];

      if (currentProps.nodeLabels && projectedNodes) {
        projectedNodes.forEach(function (node, nodei) {
          var feasibleLabel = nodeLabels && nodeLabels !== true && nodeLabels(node);

          if (nodeLabels === true || feasibleLabel) {
            var actualLabel = networkSettings.projection === "radial" && node.depth !== 0 ? (0, _networkDrawing.radialLabelGenerator)(node, nodei, nodeLabels === true ? nodeIDAccessor : nodeLabels, adjustedSize) : nodeLabels === true ? nodeIDAccessor(node, nodei) : feasibleLabel;
            var nodeLabel;

            if (React.isValidElement(actualLabel)) {
              nodeLabel = {
                key: "node-label-".concat(nodei),
                type: "basic-node-label",
                x: node.x,
                y: node.y,
                element: actualLabel
              };
            } else {
              nodeLabel = {
                key: "node-label-".concat(nodei),
                className: "node-label",
                dx: 0,
                dy: 0,
                x: node.x,
                y: node.y,
                note: {
                  label: actualLabel
                },
                connector: {
                  end: "none"
                },
                type: _AnnotationLabel.default,
                subject: {
                  radius: nodeSizeAccessor(node) + 2
                }
              };
            }

            nodeLabelAnnotations.push(nodeLabel);
          }
        });
      }

      var projectedXYPoints;
      var overlay = [];
      var areaBasedTypes = ["circlepack", "treemap", "partition", "chord"];

      if (hoverAnnotation && areaBasedTypes.find(function (d) {
        return d === networkSettings.type;
      }) || hoverAnnotation === "area") {
        if (hoverAnnotation !== "edge") {
          var renderedNodeOverlays = projectedNodes.map(function (d, i) {
            return _objectSpread({
              overlayData: d
            }, customNodeIcon({
              d: d,
              i: i,
              transform: "translate(".concat(d.x, ",").concat(d.y, ")"),
              styleFn: function styleFn() {
                return {
                  fill: "pink",
                  opacity: 0
                };
              }
            }).props);
          });
          overlay.push.apply(overlay, _toConsumableArray(renderedNodeOverlays));
        }

        if (hoverAnnotation !== "node") {
          projectedEdges.forEach(function (d, i) {
            var generatedIcon = customEdgeIcon({
              d: d,
              i: i,
              transform: "translate(".concat(d.x, ",").concat(d.y, ")"),
              styleFn: function styleFn() {
                return {
                  fill: "pink",
                  opacity: 0
                };
              }
            });

            if (generatedIcon) {
              overlay.push(_objectSpread({
                overlayData: _objectSpread({}, d, {
                  x: d.x || (d.source.x + d.target.x) / 2,
                  y: d.y || (d.source.y + d.target.y) / 2,
                  edge: true
                })
              }, generatedIcon.props));
            }
          });
        }
      } else if (hoverAnnotation === "edge" && typeof networkSettings.type === "string" && edgePointHash[networkSettings.type]) {
        projectedXYPoints = projectedEdges.map(edgePointHash[networkSettings.type]);
      } else if (Array.isArray(hoverAnnotation) || hoverAnnotation === true || hoverAnnotation === "node") {
        projectedXYPoints = projectedNodes;
        if (changedData || networkSettingsChanged) projectedXYPoints = _toConsumableArray(projectedNodes);
      } else if (hoverAnnotation === "all" && typeof networkSettings.type === "string") {
        projectedXYPoints = [].concat(_toConsumableArray(projectedEdges.map(edgePointHash[networkSettings.type])), _toConsumableArray(projectedNodes));
      }

      this.setState({
        adjustedPosition: adjustedPosition,
        adjustedSize: adjustedSize,
        backgroundGraphics: currentProps.backgroundGraphics,
        foregroundGraphics: currentProps.foregroundGraphics,
        title: title,
        renderNumber: this.state.renderNumber + 1,
        projectedNodes: projectedNodes,
        projectedEdges: projectedEdges,
        projectedXYPoints: projectedXYPoints,
        overlay: overlay,
        nodeIDAccessor: nodeIDAccessor,
        sourceAccessor: sourceAccessor,
        targetAccessor: targetAccessor,
        nodeSizeAccessor: nodeSizeAccessor,
        edgeWidthAccessor: edgeWidthAccessor,
        margin: margin,
        legendSettings: legendSettings,
        networkFrameRender: networkFrameRender,
        nodeLabelAnnotations: nodeLabelAnnotations,
        graphSettings: _objectSpread({}, networkSettings.graphSettings, networkSettings)
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          annotations = _this$props2.annotations,
          annotationSettings = _this$props2.annotationSettings,
          className = _this$props2.className,
          customClickBehavior = _this$props2.customClickBehavior,
          customDoubleClickBehavior = _this$props2.customDoubleClickBehavior,
          customHoverBehavior = _this$props2.customHoverBehavior,
          size = _this$props2.size,
          matte = _this$props2.matte,
          hoverAnnotation = _this$props2.hoverAnnotation,
          beforeElements = _this$props2.beforeElements,
          afterElements = _this$props2.afterElements,
          interaction = _this$props2.interaction,
          disableContext = _this$props2.disableContext,
          canvasPostProcess = _this$props2.canvasPostProcess,
          baseMarkProps = _this$props2.baseMarkProps,
          useSpans = _this$props2.useSpans,
          canvasNodes = _this$props2.canvasNodes,
          canvasEdges = _this$props2.canvasEdges,
          name = _this$props2.name,
          downloadFields = _this$props2.downloadFields,
          download = _this$props2.download,
          additionalDefs = _this$props2.additionalDefs,
          _this$props2$renderOr = _this$props2.renderOrder,
          renderOrder = _this$props2$renderOr === void 0 ? this.state.graphSettings && this.state.graphSettings.type === "matrix" ? matrixRenderOrder : generalRenderOrder : _this$props2$renderOr;
      var _this$state4 = this.state,
          backgroundGraphics = _this$state4.backgroundGraphics,
          foregroundGraphics = _this$state4.foregroundGraphics,
          projectedXYPoints = _this$state4.projectedXYPoints,
          margin = _this$state4.margin,
          legendSettings = _this$state4.legendSettings,
          adjustedPosition = _this$state4.adjustedPosition,
          adjustedSize = _this$state4.adjustedSize,
          networkFrameRender = _this$state4.networkFrameRender,
          nodeLabelAnnotations = _this$state4.nodeLabelAnnotations,
          overlay = _this$state4.overlay,
          projectedNodes = _this$state4.projectedNodes,
          projectedEdges = _this$state4.projectedEdges,
          title = _this$state4.title;
      var downloadButton = [];

      if (download && projectedNodes.length > 0) {
        downloadButton.push(React.createElement(_DownloadButton.default, {
          key: "network-download-nodes",
          csvName: "".concat(name, "-").concat(new Date().toJSON()),
          width: size[0],
          label: "Download Node List",
          data: (0, _downloadDataMapping.networkNodeDownloadMapping)({
            data: projectedNodes,
            fields: downloadFields
          })
        }));
      }

      if (download && projectedEdges.length > 0) {
        downloadButton.push(React.createElement(_DownloadButton.default, {
          key: "network-download-edges",
          csvName: "".concat(name, "-").concat(new Date().toJSON()),
          width: size[0],
          label: "Download Edge List",
          data: (0, _downloadDataMapping.networkEdgeDownloadMapping)({
            data: projectedEdges,
            fields: downloadFields
          })
        }));
      }

      var formattedOverlay;

      if (overlay && overlay.length > 0) {
        formattedOverlay = overlay;
      }

      return React.createElement(_Frame.default, {
        name: "networkframe",
        renderPipeline: networkFrameRender,
        adjustedPosition: adjustedPosition,
        adjustedSize: adjustedSize,
        size: size,
        xScale: xScale,
        yScale: yScale,
        title: title,
        matte: matte,
        className: className,
        additionalDefs: additionalDefs,
        frameKey: "none",
        projectedCoordinateNames: projectedCoordinateNames,
        defaultSVGRule: this.defaultNetworkSVGRule,
        defaultHTMLRule: this.defaultNetworkHTMLRule,
        hoverAnnotation: Array.isArray(hoverAnnotation) ? hoverAnnotation : !!hoverAnnotation,
        annotations: [].concat(_toConsumableArray(annotations), _toConsumableArray(nodeLabelAnnotations)),
        annotationSettings: annotationSettings,
        legendSettings: legendSettings,
        interaction: interaction,
        customClickBehavior: customClickBehavior,
        customHoverBehavior: customHoverBehavior,
        customDoubleClickBehavior: customDoubleClickBehavior,
        points: projectedXYPoints,
        margin: margin,
        overlay: formattedOverlay,
        backgroundGraphics: backgroundGraphics,
        foregroundGraphics: foregroundGraphics,
        beforeElements: beforeElements,
        afterElements: afterElements,
        downloadButton: downloadButton,
        disableContext: disableContext,
        canvasPostProcess: canvasPostProcess,
        baseMarkProps: baseMarkProps,
        useSpans: !!useSpans,
        canvasRendering: !!(canvasNodes || canvasEdges),
        renderOrder: renderOrder,
        disableCanvasInteraction: true
      });
    }
  }]);

  return NetworkFrame;
}(React.Component);

_defineProperty(NetworkFrame, "defaultProps", {
  annotations: [],
  foregroundGraphics: [],
  annotationSettings: {},
  size: [500, 500],
  className: "",
  name: "networkframe",
  networkType: {
    type: "force",
    iterations: 500
  },
  filterRenderedNodes: function filterRenderedNodes(d) {
    return d.id !== "root-generated";
  }
});

_defineProperty(NetworkFrame, "displayName", "NetworkFrame");

var _default = NetworkFrame;
exports.default = _default;
module.exports = exports.default;
