// semiotic/lib/XYFrame.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _d3Scale = require("d3-scale");

var _axis = require("./visualizationLayerBehavior/axis");

var _Axis = _interopRequireDefault(require("./Axis"));

var _DownloadButton = _interopRequireDefault(require("./DownloadButton"));

var _Frame = _interopRequireDefault(require("./Frame"));

var _xyframeRules = require("./annotationRules/xyframeRules");

var _baseRules = require("./annotationRules/baseRules");

var _general = require("./visualizationLayerBehavior/general");

var _lineDrawing = require("./svg/lineDrawing");

var _AnnotationCallout = _interopRequireDefault(require("react-annotation/lib/Types/AnnotationCallout"));

var _frameFunctions = require("./svg/frameFunctions");

var _downloadDataMapping = require("./downloadDataMapping");

var _coordinateNames = require("./constants/coordinateNames");

var _dataFunctions = require("./data/dataFunctions");

var _unflowedFunctions = require("./data/unflowedFunctions");

var _multiAccessorUtils = require("./data/multiAccessorUtils");

var _frame_props = require("./constants/frame_props");

var _SpanOrDiv = _interopRequireDefault(require("./SpanOrDiv"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var naturalLanguageLineType = {
  line: {
    items: "line",
    chart: "line chart"
  },
  area: {
    items: "summary",
    chart: "summary chart"
  },
  summary: {
    items: "summary",
    chart: "summary chart"
  },
  cumulative: {
    items: "line",
    chart: "cumulative chart"
  },
  "cumulative-reverse": {
    items: "line",
    chart: "cumulative chart"
  },
  linepercent: {
    items: "line",
    chart: "line chart"
  },
  stackedarea: {
    items: "stacked area",
    chart: "stacked area chart"
  },
  "stackedarea-invert": {
    items: "stacked area",
    chart: "stacked area chart"
  },
  stackedpercent: {
    items: "stacked area",
    chart: "stacked area chart"
  },
  "stackedpercent-invert": {
    items: "stacked area",
    chart: "stacked area chart"
  },
  bumparea: {
    items: "ranked area",
    chart: "ranked area chart"
  },
  "bumparea-invert": {
    items: "ranked area",
    chart: "ranked area chart"
  },
  bumpline: {
    items: "ranked line",
    chart: "ranked line chart"
  },
  difference: {
    items: "line",
    chart: "difference chart"
  }
};

var emptyObjectReturnFunction = function emptyObjectReturnFunction() {
  return {};
};

var emptyStringReturnFunction = function emptyStringReturnFunction() {
  return "";
};

var xyframeKey = "";
var chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

for (var i = 32; i > 0; --i) {
  xyframeKey += chars[Math.floor(Math.random() * chars.length)];
}

var projectedCoordinateNames = {
  y: _coordinateNames.projectedY,
  x: _coordinateNames.projectedX,
  yMiddle: _coordinateNames.projectedYMiddle,
  yTop: _coordinateNames.projectedYTop,
  yBottom: _coordinateNames.projectedYBottom,
  xMiddle: _coordinateNames.projectedXMiddle,
  xTop: _coordinateNames.projectedXTop,
  xBottom: _coordinateNames.projectedXBottom
};

function mapParentsToPoints(fullDataset) {
  return fullDataset.map(function (d) {
    if (d.parentLine) {
      return _extends({}, d.parentLine, d);
    }

    if (d.parentSummary) {
      return _extends({}, d.parentSummary, d);
    }

    return d;
  });
}

var XYFrame =
/*#__PURE__*/
function (_React$Component) {
  _inherits(XYFrame, _React$Component);

  function XYFrame() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, XYFrame);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(XYFrame)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {
      size: [500, 500],
      dataVersion: undefined,
      lineData: undefined,
      pointData: undefined,
      summaryData: undefined,
      projectedLines: undefined,
      projectedPoints: undefined,
      projectedSummaries: undefined,
      fullDataset: [],
      adjustedPosition: [0, 0],
      adjustedSize: [500, 500],
      backgroundGraphics: null,
      foregroundGraphics: null,
      axesData: undefined,
      axes: undefined,
      axesTickLines: undefined,
      renderNumber: 0,
      margin: {
        top: 0,
        bottom: 0,
        left: 0,
        right: 0
      },
      calculatedXExtent: [0, 0],
      calculatedYExtent: [0, 0],
      xAccessor: [function (d) {
        return d.x;
      }],
      yAccessor: [function (d) {
        return d.y;
      }],
      xExtent: [0, 0],
      yExtent: [0, 0],
      areaAnnotations: [],
      xScale: (0, _d3Scale.scaleLinear)(),
      yScale: (0, _d3Scale.scaleLinear)(),
      title: null,
      legendSettings: undefined,
      xyFrameRender: {},
      canvasDrawing: [],
      annotatedSettings: {
        xAccessor: undefined,
        yAccessor: undefined,
        summaryDataAccessor: undefined,
        lineDataAccessor: undefined,
        renderKeyFn: undefined,
        lineType: undefined,
        summaryType: undefined,
        lineIDAccessor: undefined,
        summaries: undefined,
        lines: undefined,
        title: undefined,
        xExtent: undefined,
        yExtent: undefined
      },
      overlay: undefined
    });

    _defineProperty(_assertThisInitialized(_this), "calculateXYFrame", function (currentProps, updateData) {
      var legend = currentProps.legend,
          lines = currentProps.lines,
          lineClass = currentProps.lineClass,
          pointStyle = currentProps.pointStyle,
          pointRenderMode = currentProps.pointRenderMode,
          pointClass = currentProps.pointClass,
          areaClass = currentProps.areaClass,
          _currentProps$summary = currentProps.summaryClass,
          summaryClass = _currentProps$summary === void 0 ? areaClass : _currentProps$summary,
          canvasLines = currentProps.canvasLines,
          canvasPoints = currentProps.canvasPoints,
          canvasAreas = currentProps.canvasAreas,
          _currentProps$canvasS = currentProps.canvasSummaries,
          canvasSummaries = _currentProps$canvasS === void 0 ? canvasAreas : _currentProps$canvasS,
          defined = currentProps.defined,
          size = currentProps.size,
          renderKey = currentProps.renderKey,
          lineType = currentProps.lineType,
          areaType = currentProps.areaType,
          _currentProps$summary2 = currentProps.summaryType,
          summaryType = _currentProps$summary2 === void 0 ? areaType : _currentProps$summary2,
          customLineMark = currentProps.customLineMark,
          customPointMark = currentProps.customPointMark,
          customAreaMark = currentProps.customAreaMark,
          _currentProps$customS = currentProps.customSummaryMark,
          customSummaryMark = _currentProps$customS === void 0 ? customAreaMark : _currentProps$customS,
          areaStyle = currentProps.areaStyle,
          _currentProps$summary3 = currentProps.summaryStyle,
          summaryStyle = _currentProps$summary3 === void 0 ? areaStyle : _currentProps$summary3,
          areaRenderMode = currentProps.areaRenderMode,
          _currentProps$summary4 = currentProps.summaryRenderMode,
          summaryRenderMode = _currentProps$summary4 === void 0 ? areaRenderMode : _currentProps$summary4,
          lineStyle = currentProps.lineStyle,
          lineRenderMode = currentProps.lineRenderMode,
          baseXExtent = currentProps.xExtent,
          baseYExtent = currentProps.yExtent,
          title = currentProps.title,
          _currentProps$xScaleT = currentProps.xScaleType,
          xScaleType = _currentProps$xScaleT === void 0 ? (0, _d3Scale.scaleLinear)() : _currentProps$xScaleT,
          _currentProps$yScaleT = currentProps.yScaleType,
          yScaleType = _currentProps$yScaleT === void 0 ? (0, _d3Scale.scaleLinear)() : _currentProps$yScaleT,
          lineIDAccessor = currentProps.lineIDAccessor,
          invertX = currentProps.invertX,
          invertY = currentProps.invertY,
          showLinePoints = currentProps.showLinePoints,
          showSummaryPoints = currentProps.showSummaryPoints,
          points = currentProps.points,
          areas = currentProps.areas,
          lineDataAccessor = currentProps.lineDataAccessor,
          areaDataAccessor = currentProps.areaDataAccessor,
          _currentProps$summary5 = currentProps.summaryDataAccessor,
          summaryDataAccessor = _currentProps$summary5 === void 0 ? areaDataAccessor : _currentProps$summary5,
          yAccessor = currentProps.yAccessor,
          xAccessor = currentProps.xAccessor,
          useSummariesAsInteractionLayer = currentProps.useSummariesAsInteractionLayer,
          _currentProps$useArea = currentProps.useAreasAsInteractionLayer,
          useAreasAsInteractionLayer = _currentProps$useArea === void 0 ? useSummariesAsInteractionLayer : _currentProps$useArea,
          baseMarkProps = currentProps.baseMarkProps;
      var projectedLines = currentProps.projectedLines,
          projectedPoints = currentProps.projectedPoints,
          _currentProps$project = currentProps.projectedSummaries,
          projectedSummaries = _currentProps$project === void 0 ? currentProps.projectedAreas : _currentProps$project,
          _currentProps$summari = currentProps.summaries,
          summaries = _currentProps$summari === void 0 ? areas : _currentProps$summari,
          fullDataset = currentProps.fullDataset;

      if (summaryType && points && !summaries) {
        summaries = [{
          coordinates: points
        }];
      }

      var annotatedSettings = {
        xAccessor: (0, _dataFunctions.stringToArrayFn)(xAccessor, function (d) {
          return d[0];
        }),
        yAccessor: (0, _dataFunctions.stringToArrayFn)(yAccessor, function (d) {
          return d[1];
        }),
        summaryDataAccessor: (0, _dataFunctions.stringToArrayFn)(summaryDataAccessor, function (d) {
          return Array.isArray(d) ? d : d.coordinates;
        }),
        lineDataAccessor: (0, _dataFunctions.stringToArrayFn)(lineDataAccessor, function (d) {
          return Array.isArray(d) ? d : d.coordinates;
        }),
        renderKeyFn: (0, _dataFunctions.stringToFn)(renderKey, function (d, i) {
          return "line-".concat(i);
        }, true),
        lineType: (0, _frameFunctions.objectifyType)(lineType),
        summaryType: (0, _frameFunctions.objectifyType)(summaryType),
        lineIDAccessor: (0, _dataFunctions.stringToFn)(lineIDAccessor, function (l) {
          return l.semioticLineID;
        }),
        summaries: !summaries || Array.isArray(summaries) && summaries.length === 0 ? undefined : !Array.isArray(summaries) ? [summaries] : !summaryDataAccessor && !summaries[0].coordinates ? [{
          coordinates: summaries
        }] : summaries,
        lines: !lines || Array.isArray(lines) && lines.length === 0 ? undefined : !Array.isArray(lines) ? [lines] : !lineDataAccessor && !lines[0].coordinates ? [{
          coordinates: lines
        }] : lines,
        title: _typeof(title) === "object" && !React.isValidElement(title) && title !== null ? title : {
          title: title,
          orient: "top"
        },
        xExtent: Array.isArray(baseXExtent) ? baseXExtent : !baseXExtent ? undefined : baseXExtent.extent,
        yExtent: Array.isArray(baseYExtent) ? baseYExtent : !baseYExtent ? undefined : baseYExtent.extent
      };
      annotatedSettings.lineType.simpleLine = annotatedSettings.lineType.type === "line" && !annotatedSettings.lineType.y1 && annotatedSettings.lineType.simpleLine !== false;

      if (annotatedSettings.lineType.type === "area") {
        annotatedSettings.lineType.y1 = function () {
          return 0;
        };
      }

      var summaryStyleFn = (0, _dataFunctions.stringToFn)(summaryStyle, emptyObjectReturnFunction, true);
      var summaryClassFn = (0, _dataFunctions.stringToFn)(summaryClass, emptyStringReturnFunction, true);
      var summaryRenderModeFn = (0, _dataFunctions.stringToFn)(summaryRenderMode, undefined, true);
      var margin = (0, _frameFunctions.calculateMargin)({
        margin: currentProps.margin,
        axes: currentProps.axes,
        title: annotatedSettings.title
      });

      var _adjustedPositionSize = (0, _frameFunctions.adjustedPositionSize)({
        size: currentProps.size,
        margin: margin
      }),
          adjustedPosition = _adjustedPositionSize.adjustedPosition,
          adjustedSize = _adjustedPositionSize.adjustedSize;

      var calculatedXExtent = [],
          calculatedYExtent = [],
          yExtent,
          xExtent,
          xExtentSettings,
          yExtentSettings;

      if (_typeof(baseXExtent) === "object") {
        xExtentSettings = baseXExtent;
      } else {
        xExtentSettings = {
          extent: baseXExtent
        };
      }

      if (_typeof(baseYExtent) === "object") {
        yExtentSettings = baseYExtent;
      } else {
        yExtentSettings = {
          extent: baseYExtent
        };
      }

      var xScale, yScale;

      if (updateData || currentProps.dataVersion && currentProps.dataVersion !== _this.state.dataVersion) {
        //This will always fire at this point because xExtent/yExtent are just defined up there so revisit this logic
        if (!xExtent || !yExtent || !fullDataset || !projectedLines && !projectedPoints && !projectedSummaries) {
          ;

          var _calculateDataExtent = (0, _dataFunctions.calculateDataExtent)({
            lineDataAccessor: annotatedSettings.lineDataAccessor,
            summaryDataAccessor: annotatedSettings.summaryDataAccessor,
            xAccessor: annotatedSettings.xAccessor,
            yAccessor: annotatedSettings.yAccessor,
            lineType: annotatedSettings.lineType,
            summaryType: annotatedSettings.summaryType,
            summaries: annotatedSettings.summaries,
            points: points,
            lines: annotatedSettings.lines,
            showLinePoints: showLinePoints,
            showSummaryPoints: showSummaryPoints,
            xExtent: baseXExtent,
            yExtent: baseYExtent,
            invertX: invertX,
            invertY: invertY,
            adjustedSize: adjustedSize,
            margin: margin,
            baseMarkProps: baseMarkProps,
            summaryStyleFn: summaryStyleFn,
            summaryClassFn: summaryClassFn,
            summaryRenderModeFn: summaryRenderModeFn,
            chartSize: size,
            xScaleType: xScaleType,
            yScaleType: yScaleType,
            defined: defined
          });

          xExtent = _calculateDataExtent.xExtent;
          yExtent = _calculateDataExtent.yExtent;
          projectedLines = _calculateDataExtent.projectedLines;
          projectedPoints = _calculateDataExtent.projectedPoints;
          projectedSummaries = _calculateDataExtent.projectedSummaries;
          fullDataset = _calculateDataExtent.fullDataset;
          calculatedXExtent = _calculateDataExtent.calculatedXExtent;
          calculatedYExtent = _calculateDataExtent.calculatedYExtent;
        }

        ;

        var _this$screenScales = _this.screenScales({
          xExtent: xExtent,
          yExtent: yExtent,
          adjustedSize: adjustedSize,
          xScaleType: xScaleType,
          yScaleType: yScaleType
        });

        xScale = _this$screenScales.xScale;
        yScale = _this$screenScales.yScale;
      } else {
        ;
        var _this$state = _this.state;
        xExtent = _this$state.xExtent;
        yExtent = _this$state.yExtent;
        projectedLines = _this$state.projectedLines;
        projectedPoints = _this$state.projectedPoints;
        projectedSummaries = _this$state.projectedSummaries;
        fullDataset = _this$state.fullDataset;
        calculatedXExtent = _this$state.calculatedXExtent;
        calculatedYExtent = _this$state.calculatedYExtent;

        if (adjustedSize[0] === _this.state.adjustedSize[0] && adjustedSize[1] === _this.state.adjustedSize[1]) {
          xScale = _this.state.xScale;
          yScale = _this.state.yScale;
        } else {
          ;

          var _this$screenScales2 = _this.screenScales({
            xExtent: xExtent,
            yExtent: yExtent,
            adjustedSize: adjustedSize,
            xScaleType: xScaleType,
            yScaleType: yScaleType
          });

          xScale = _this$screenScales2.xScale;
          yScale = _this$screenScales2.yScale;
        }
      }

      xExtent = Array.isArray(xExtentSettings.extent) && xExtentSettings.extent.length === 2 ? xExtentSettings.extent : xExtent;
      yExtent = Array.isArray(yExtentSettings.extent) && yExtentSettings.extent.length === 2 ? yExtentSettings.extent : yExtent;
      var canvasDrawing = [];
      var axes;
      var axesTickLines;
      var existingBaselines = {};

      if (currentProps.axes) {
        axesTickLines = [];
        axes = currentProps.axes.map(function (d, i) {
          var axisClassname = d.className || "";
          axisClassname += " axis";
          var axisScale = yScale;

          if (existingBaselines[d.orient]) {
            d.baseline = d.baseline || false;
          }

          existingBaselines[d.orient] = true;

          if (d.orient === "top" || d.orient === "bottom") {
            axisClassname += " x";
            axisScale = xScale;
          } else {
            axisClassname += " y";
          }

          axisClassname += " ".concat(d.orient);
          var tickValues;

          if (d.tickValues && Array.isArray(d.tickValues)) {
            tickValues = d.tickValues;
          } else if (d.tickValues instanceof Function) {
            //otherwise assume a function
            tickValues = d.tickValues(fullDataset, currentProps.size, axisScale);
          }

          var axisSize = [adjustedSize[0], adjustedSize[1]];
          var axisParts = (0, _axis.axisPieces)({
            padding: d.padding,
            tickValues: tickValues,
            scale: axisScale,
            ticks: d.ticks,
            orient: d.orient,
            size: axisSize,
            footer: d.footer,
            tickSize: d.tickSize
          });
          var axisTickLines = React.createElement("g", {
            key: "axes-tick-lines-".concat(i),
            className: "axis ".concat(axisClassname)
          }, (0, _axis.axisLines)({
            axisParts: axisParts,
            orient: d.orient,
            tickLineGenerator: d.tickLineGenerator,
            baseMarkProps: baseMarkProps,
            className: axisClassname
          }));
          axesTickLines.push(axisTickLines);
          return React.createElement(_Axis.default, {
            label: d.label,
            axisParts: axisParts,
            key: d.key || "axis-".concat(i),
            orient: d.orient,
            size: axisSize,
            margin: margin,
            ticks: d.ticks,
            tickSize: d.tickSize,
            tickFormat: d.tickFormat,
            tickValues: tickValues,
            scale: axisScale,
            className: axisClassname,
            padding: d.padding,
            rotate: d.rotate,
            annotationFunction: d.axisAnnotationFunction,
            glyphFunction: d.glyphFunction,
            baseline: d.baseline,
            dynamicLabelPosition: d.dynamicLabelPosition,
            center: d.center,
            xyPoints: fullDataset,
            marginalSummaryType: d.marginalSummaryType
          });
        });
      }

      var legendSettings;

      if (legend) {
        legendSettings = legend === true ? {} : legend;

        if (projectedLines && !legendSettings.legendGroups) {
          var typeString = annotatedSettings.lineType.type;
          var type = typeof typeString === "string" && ["stackedarea", "stackedpercent", "bumparea"].indexOf(typeString) === -1 ? "line" : "fill";
          var legendGroups = [{
            styleFn: currentProps.lineStyle,
            type: type,
            items: projectedLines.map(function (d) {
              return _extends({
                label: annotatedSettings.lineIDAccessor(d)
              }, d);
            })
          }];
          legendSettings.legendGroups = legendGroups;
        }
      }

      var areaAnnotations = [];

      if (annotatedSettings.summaryType.label && projectedSummaries) {
        projectedSummaries.forEach(function (d, i) {
          if (d.bounds) {
            var bounds = Array.isArray(d.bounds) ? d.bounds : [d.bounds];
            bounds.forEach(function (labelBounds) {
              var label = typeof annotatedSettings.summaryType.label === "function" ? annotatedSettings.summaryType.label(d) : annotatedSettings.summaryType.label;

              if (label && label !== null) {
                var labelPosition = label.position || "center";
                var labelCenter = [xScale(labelBounds[labelPosition][0]), yScale(labelBounds[labelPosition][1])] || [xScale(d._xyfCoordinates[0]), yScale(d._xyfCoordinates[1])];

                var labelContent = label.content || function (p) {
                  return p.value || p.id || i;
                };

                areaAnnotations.push({
                  x: labelCenter[0],
                  y: labelCenter[1],
                  dx: label.dx,
                  dy: label.dy,
                  className: label.className,
                  type: label.type || _AnnotationCallout.default,
                  note: label.note || {
                    title: labelContent(d)
                  },
                  subject: label.subject || {
                    text: labelContent(d)
                  },
                  connector: label.connector
                });
              }
            });
          }
        });
      }

      var lineAriaLabel = annotatedSettings.lineType.type !== undefined && typeof annotatedSettings.lineType.type === "string" && naturalLanguageLineType[annotatedSettings.lineType.type];
      var xyFrameRender = {
        lines: {
          accessibleTransform: function accessibleTransform(data, i) {
            return _objectSpread({}, data[i].data[data[i].data.length - 1], {
              type: "frame-hover"
            });
          },
          data: projectedLines,
          styleFn: (0, _dataFunctions.stringToFn)(lineStyle, emptyObjectReturnFunction, true),
          classFn: (0, _dataFunctions.stringToFn)(lineClass, emptyStringReturnFunction, true),
          renderMode: (0, _dataFunctions.stringToFn)(lineRenderMode, undefined, true),
          canvasRender: (0, _dataFunctions.stringToFn)(canvasLines, undefined, true),
          customMark: customLineMark,
          type: annotatedSettings.lineType,
          defined: defined,
          renderKeyFn: annotatedSettings.renderKeyFn,
          ariaLabel: lineAriaLabel,
          axesData: currentProps.axes,
          behavior: _general.createLines
        },
        summaries: {
          accessibleTransform: function accessibleTransform(data, i) {
            return _objectSpread({}, data[i], {
              type: "frame-hover"
            });
          },
          data: projectedSummaries,
          styleFn: summaryStyleFn,
          classFn: summaryClassFn,
          renderMode: summaryRenderModeFn,
          canvasRender: (0, _dataFunctions.stringToFn)(canvasSummaries, undefined, true),
          customMark: customSummaryMark,
          type: annotatedSettings.summaryType,
          renderKeyFn: annotatedSettings.renderKeyFn,
          behavior: _general.createSummaries
        },
        points: {
          accessibleTransform: function accessibleTransform(data, i) {
            return _objectSpread({
              type: "frame-hover"
            }, data[i].data || data[i]);
          },
          data: projectedPoints,
          styleFn: (0, _dataFunctions.stringToFn)(pointStyle, emptyObjectReturnFunction, true),
          classFn: (0, _dataFunctions.stringToFn)(pointClass, emptyStringReturnFunction, true),
          renderMode: (0, _dataFunctions.stringToFn)(pointRenderMode, undefined, true),
          canvasRender: (0, _dataFunctions.stringToFn)(canvasPoints, undefined, true),
          customMark: customPointMark,
          renderKeyFn: annotatedSettings.renderKeyFn,
          showLinePoints: showLinePoints,
          behavior: _general.createPoints
        }
      };

      if (xExtentSettings.onChange && _this.state.calculatedXExtent.join(",") !== calculatedXExtent.join(",")) {
        xExtentSettings.onChange(calculatedXExtent);
      }

      if (yExtentSettings.onChange && _this.state.calculatedYExtent.join(",") !== calculatedYExtent.join(",")) {
        yExtentSettings.onChange(calculatedYExtent);
      }

      var overlay = undefined;

      if (useAreasAsInteractionLayer && projectedSummaries) {
        overlay = (0, _general.createSummaries)({
          xScale: xScale,
          yScale: yScale,
          data: projectedSummaries
        }).map(function (m, i) {
          return _objectSpread({}, m.props, {
            style: {
              fillOpacity: 0
            },
            overlayData: projectedSummaries && projectedSummaries[i] // luckily createSummaries is a map fn

          });
        });
      }

      _this.setState({
        lineData: currentProps.lines,
        pointData: currentProps.points,
        summaryData: currentProps.summaries || currentProps.areas,
        dataVersion: currentProps.dataVersion,
        projectedLines: projectedLines,
        projectedPoints: projectedPoints,
        projectedSummaries: projectedSummaries,
        canvasDrawing: canvasDrawing,
        fullDataset: fullDataset,
        adjustedPosition: adjustedPosition,
        adjustedSize: adjustedSize,
        backgroundGraphics: currentProps.backgroundGraphics,
        foregroundGraphics: currentProps.foregroundGraphics,
        axesData: currentProps.axes,
        axes: axes,
        axesTickLines: axesTickLines,
        renderNumber: _this.state.renderNumber + 1,
        xScale: xScale,
        yScale: yScale,
        xAccessor: annotatedSettings.xAccessor,
        yAccessor: annotatedSettings.yAccessor,
        xExtent: [xExtent[0] === undefined ? calculatedXExtent[0] : xExtent[0], xExtent[1] === undefined ? calculatedXExtent[1] : xExtent[1]],
        yExtent: [yExtent[0] === undefined ? calculatedYExtent[0] : yExtent[0], yExtent[1] === undefined ? calculatedYExtent[1] : yExtent[1]],
        calculatedXExtent: calculatedXExtent,
        calculatedYExtent: calculatedYExtent,
        margin: margin,
        legendSettings: legendSettings,
        areaAnnotations: areaAnnotations,
        xyFrameRender: xyFrameRender,
        size: size,
        annotatedSettings: annotatedSettings,
        overlay: overlay
      });
    });

    _defineProperty(_assertThisInitialized(_this), "defaultXYSVGRule", function (_ref) {
      var baseD = _ref.d,
          i = _ref.i,
          annotationLayer = _ref.annotationLayer,
          lines = _ref.lines,
          summaries = _ref.summaries,
          points = _ref.points;
      var showLinePoints = _this.props.showLinePoints;
      var _this$state2 = _this.state,
          xyFrameRender = _this$state2.xyFrameRender,
          xScale = _this$state2.xScale,
          yScale = _this$state2.yScale,
          xAccessor = _this$state2.xAccessor,
          yAccessor = _this$state2.yAccessor;
      var screenCoordinates = [];
      var idAccessor = _this.state.annotatedSettings.lineIDAccessor;

      if (baseD.type === "highlight") {
        return (0, _xyframeRules.svgHighlight)({
          d: baseD,
          i: i,
          idAccessor: idAccessor,
          lines: lines,
          summaries: summaries,
          points: points,
          xScale: xScale,
          yScale: yScale,
          xyFrameRender: xyFrameRender
        });
      }

      var d = baseD.coordinates ? baseD : (0, _lineDrawing.findPointByID)({
        point: baseD,
        idAccessor: idAccessor,
        lines: lines,
        xScale: xScale,
        projectedX: _coordinateNames.projectedX,
        xAccessor: xAccessor
      });
      if (!d) return null;
      var margin = (0, _frameFunctions.calculateMargin)({
        margin: _this.props.margin,
        axes: _this.props.axes,
        title: _this.state.annotatedSettings.title
      });

      var _adjustedPositionSize2 = (0, _frameFunctions.adjustedPositionSize)({
        size: _this.props.size,
        margin: margin
      }),
          adjustedPosition = _adjustedPositionSize2.adjustedPosition,
          adjustedSize = _adjustedPositionSize2.adjustedSize;

      if (!d.coordinates && !d.bounds) {
        screenCoordinates = [(0, _lineDrawing.relativeX)({
          point: d,
          projectedXMiddle: _coordinateNames.projectedXMiddle,
          projectedX: _coordinateNames.projectedX,
          xAccessor: xAccessor,
          xScale: xScale
        }) || 0, (0, _lineDrawing.relativeY)({
          point: d,
          projectedYMiddle: _coordinateNames.projectedYMiddle,
          projectedY: _coordinateNames.projectedY,
          yAccessor: yAccessor,
          yScale: yScale,
          showLinePoints: showLinePoints
        }) || 0];
      } else if (!d.bounds) {
        screenCoordinates = d.coordinates.reduce(function (coords, p) {
          var xCoordinate = (0, _lineDrawing.relativeX)({
            point: p,
            projectedXMiddle: _coordinateNames.projectedXMiddle,
            projectedX: _coordinateNames.projectedX,
            xAccessor: xAccessor,
            xScale: xScale
          });
          var yCoordinate = (0, _lineDrawing.relativeY)({
            point: p,
            projectedYMiddle: _coordinateNames.projectedYMiddle,
            projectedY: _coordinateNames.projectedY,
            yAccessor: yAccessor,
            yScale: yScale
          });

          if (Array.isArray(yCoordinate)) {
            return [].concat(_toConsumableArray(coords), [[xCoordinate, Math.min.apply(Math, _toConsumableArray(yCoordinate))], [xCoordinate, Math.max.apply(Math, _toConsumableArray(yCoordinate))]]);
          } else if (Array.isArray(xCoordinate)) {
            return [].concat(_toConsumableArray(coords), [[Math.min.apply(Math, _toConsumableArray(xCoordinate)), yCoordinate], [Math.max.apply(Math, _toConsumableArray(xCoordinate)), yCoordinate]]);
          } else {
            return [].concat(_toConsumableArray(coords), [[xCoordinate, yCoordinate]]);
          }
        }, []);
      }

      var voronoiHover = annotationLayer.voronoiHover;

      var customSVG = _this.props.svgAnnotationRules && _this.props.svgAnnotationRules({
        d: d,
        i: i,
        screenCoordinates: screenCoordinates,
        xScale: xScale,
        yScale: yScale,
        xAccessor: xAccessor,
        yAccessor: yAccessor,
        xyFrameProps: _this.props,
        xyFrameState: _this.state,
        summaries: summaries,
        points: points,
        lines: lines,
        voronoiHover: voronoiHover,
        adjustedPosition: adjustedPosition,
        adjustedSize: adjustedSize,
        annotationLayer: annotationLayer
      });

      if (_this.props.svgAnnotationRules !== undefined && customSVG !== null) {
        return customSVG;
      } else if (d.type === "desaturation-layer") {
        return (0, _baseRules.desaturationLayer)({
          style: d.style instanceof Function ? d.style(d, i) : d.style,
          size: adjustedSize,
          i: i,
          key: d.key
        });
      } else if (d.type === "xy" || d.type === "frame-hover") {
        return (0, _xyframeRules.svgXYAnnotation)({
          d: d,
          i: i,
          screenCoordinates: screenCoordinates
        });
      } else if (d.type === "react-annotation" || typeof d.type === "function") {
        return (0, _xyframeRules.basicReactAnnotation)({
          d: d,
          screenCoordinates: screenCoordinates,
          i: i
        });
      } else if (d.type === "enclose") {
        return (0, _xyframeRules.svgEncloseAnnotation)({
          d: d,
          screenCoordinates: screenCoordinates,
          i: i
        });
      } else if (d.type === "enclose-rect") {
        return (0, _xyframeRules.svgRectEncloseAnnotation)({
          d: d,
          screenCoordinates: screenCoordinates,
          i: i
        });
      } else if (d.type === "enclose-hull") {
        return (0, _xyframeRules.svgHullEncloseAnnotation)({
          d: d,
          screenCoordinates: screenCoordinates,
          i: i
        });
      } else if (d.type === "x") {
        return (0, _xyframeRules.svgXAnnotation)({
          d: d,
          screenCoordinates: screenCoordinates,
          i: i,
          adjustedSize: adjustedSize
        });
      } else if (d.type === "y") {
        return (0, _xyframeRules.svgYAnnotation)({
          d: d,
          screenCoordinates: screenCoordinates,
          i: i,
          adjustedSize: adjustedSize,
          adjustedPosition: adjustedPosition
        });
      } else if (d.type === "bounds") {
        return (0, _xyframeRules.svgBoundsAnnotation)({
          d: d,
          i: i,
          adjustedSize: adjustedSize,
          xAccessor: xAccessor,
          yAccessor: yAccessor,
          xScale: xScale,
          yScale: yScale
        });
      } else if (d.type === "line") {
        return (0, _xyframeRules.svgLineAnnotation)({
          d: d,
          i: i,
          screenCoordinates: screenCoordinates
        });
      } else if (d.type === "area") {
        return (0, _xyframeRules.svgAreaAnnotation)({
          d: d,
          i: i,
          xScale: xScale,
          xAccessor: xAccessor,
          yScale: yScale,
          yAccessor: yAccessor,
          annotationLayer: annotationLayer
        });
      } else if (d.type === "horizontal-points") {
        return (0, _xyframeRules.svgHorizontalPointsAnnotation)({
          d: d,
          lines: lines.data,
          points: points.data,
          xScale: xScale,
          yScale: yScale,
          pointStyle: points.styleFn
        });
      } else if (d.type === "vertical-points") {
        return (0, _xyframeRules.svgVerticalPointsAnnotation)({
          d: d,
          lines: lines.data,
          points: points.data,
          xScale: xScale,
          yScale: yScale,
          pointStyle: points.styleFn
        });
      }

      return null;
    });

    _defineProperty(_assertThisInitialized(_this), "defaultXYHTMLRule", function (_ref2) {
      var baseD = _ref2.d,
          i = _ref2.i,
          lines = _ref2.lines,
          summaries = _ref2.summaries,
          points = _ref2.points,
          annotationLayer = _ref2.annotationLayer;
      var xAccessor = _this.state.xAccessor;
      var yAccessor = _this.state.yAccessor;
      var showLinePoints = _this.props.showLinePoints;
      var xScale = _this.state.xScale;
      var yScale = _this.state.yScale;
      var voronoiHover = annotationLayer.voronoiHover;
      var screenCoordinates = [];
      var useSpans = _this.props.useSpans;
      var idAccessor = _this.state.annotatedSettings.lineIDAccessor;
      var d = (0, _lineDrawing.findPointByID)({
        point: baseD,
        idAccessor: idAccessor,
        lines: lines,
        xScale: xScale,
        projectedX: _coordinateNames.projectedX,
        xAccessor: xAccessor
      });

      if (!d) {
        return null;
      }

      var xCoord = d[_coordinateNames.projectedXMiddle] || d[_coordinateNames.projectedX] || (0, _multiAccessorUtils.findFirstAccessorValue)(xAccessor, d);
      var yCoord = d[_coordinateNames.projectedYMiddle] || d[_coordinateNames.projectedY] || (0, _multiAccessorUtils.findFirstAccessorValue)(yAccessor, d);
      var xString = xCoord && xCoord.toString ? xCoord.toString() : xCoord;
      var yString = yCoord && yCoord.toString ? yCoord.toString() : yCoord;
      var margin = (0, _frameFunctions.calculateMargin)({
        margin: _this.props.margin,
        axes: _this.props.axes,
        title: _this.state.annotatedSettings.title
      });

      var _adjustedPositionSize3 = (0, _frameFunctions.adjustedPositionSize)({
        size: _this.props.size,
        margin: margin
      }),
          adjustedPosition = _adjustedPositionSize3.adjustedPosition,
          adjustedSize = _adjustedPositionSize3.adjustedSize;

      if (!d.coordinates) {
        screenCoordinates = [xScale(xCoord) || 0, (0, _lineDrawing.relativeY)({
          point: d,
          projectedYMiddle: _coordinateNames.projectedYMiddle,
          projectedY: _coordinateNames.projectedY,
          showLinePoints: showLinePoints,
          yAccessor: yAccessor,
          yScale: yScale
        }) || 0];
      } else {
        screenCoordinates = d.coordinates.map(function (p) {
          var foundP = (0, _lineDrawing.findPointByID)({
            point: _objectSpread({
              x: 0,
              y: 0
            }, p),
            idAccessor: idAccessor,
            lines: lines,
            xScale: xScale,
            projectedX: _coordinateNames.projectedX,
            xAccessor: xAccessor
          });
          return [(xScale((0, _multiAccessorUtils.findFirstAccessorValue)(xAccessor, d)) || 0) + adjustedPosition[0], ((0, _lineDrawing.relativeY)({
            point: foundP,
            projectedYMiddle: _coordinateNames.projectedYMiddle,
            projectedY: _coordinateNames.projectedY,
            yAccessor: yAccessor,
            yScale: yScale
          }) || 0) + adjustedPosition[1]];
        });
      }

      var customAnnotation = _this.props.htmlAnnotationRules && _this.props.htmlAnnotationRules({
        d: d,
        i: i,
        screenCoordinates: screenCoordinates,
        xScale: xScale,
        yScale: yScale,
        xAccessor: xAccessor,
        yAccessor: yAccessor,
        xyFrameProps: _this.props,
        xyFrameState: _this.state,
        summaries: summaries,
        points: points,
        lines: lines,
        voronoiHover: voronoiHover,
        adjustedPosition: adjustedPosition,
        adjustedSize: adjustedSize,
        annotationLayer: annotationLayer
      });

      if (_this.props.htmlAnnotationRules && customAnnotation !== null) {
        return customAnnotation;
      }

      if (d.type === "frame-hover") {
        var content = React.createElement(_SpanOrDiv.default, {
          span: useSpans,
          className: "tooltip-content"
        }, React.createElement("p", {
          key: "html-annotation-content-1"
        }, xString), React.createElement("p", {
          key: "html-annotation-content-2"
        }, yString), d.percent ? React.createElement("p", {
          key: "html-annotation-content-3"
        }, Math.floor(d.percent * 1000) / 10, "%") : null);

        if (d.type === "frame-hover" && _this.props.tooltipContent) {
          content = _this.props.tooltipContent(d);
        }

        return (0, _xyframeRules.htmlTooltipAnnotation)({
          content: content,
          screenCoordinates: screenCoordinates,
          i: i,
          d: d,
          useSpans: useSpans
        });
      }

      return null;
    });

    return _this;
  }

  _createClass(XYFrame, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.props.onUnmount) {
        this.props.onUnmount(this.props, this.state);
      }
    }
  }, {
    key: "componentWillMount",
    value: function componentWillMount() {
      Object.keys(this.props).forEach(function (d) {
        if (!_frame_props.xyframeproptypes[d]) {
          if (_frame_props.ordinalframeproptypes[d]) {
            console.error("".concat(d, " is an OrdinalFrame prop are you sure you're using the right frame?"));
          } else if (_frame_props.networkframeproptypes[d]) {
            console.error("".concat(d, " is a NetworkFrame prop are you sure you're using the right frame?"));
          } else {
            console.error("".concat(d, " is not a valid XYFrame prop"));
          }
        }
      });
      this.calculateXYFrame(this.props, true);
    }
  }, {
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      var _this2 = this;

      var _this$state3 = this.state,
          _this$state3$xExtent = _this$state3.xExtent,
          oldXExtent = _this$state3$xExtent === void 0 ? [] : _this$state3$xExtent,
          _this$state3$yExtent = _this$state3.yExtent,
          oldYExtent = _this$state3$yExtent === void 0 ? [] : _this$state3$yExtent,
          oldSize = _this$state3.size,
          oldDataVersion = _this$state3.dataVersion,
          lineData = _this$state3.lineData,
          summaryData = _this$state3.summaryData,
          pointData = _this$state3.pointData;
      var baseNewXExtent = nextProps.xExtent,
          baseNewYExtent = nextProps.yExtent,
          newSize = nextProps.size,
          newDataVersion = nextProps.dataVersion,
          newLines = nextProps.lines,
          areas = nextProps.areas,
          _nextProps$summaries = nextProps.summaries,
          newSummaries = _nextProps$summaries === void 0 ? areas : _nextProps$summaries,
          newPoints = nextProps.points;
      var newXExtent = (0, _unflowedFunctions.extentValue)(baseNewXExtent);
      var newYExtent = (0, _unflowedFunctions.extentValue)(baseNewYExtent);
      var extentChange = oldXExtent[0] !== newXExtent[0] && newXExtent[0] !== undefined || oldYExtent[0] !== newYExtent[0] && newYExtent[0] !== undefined || oldXExtent[1] !== newXExtent[1] && newXExtent[1] !== undefined || oldYExtent[1] !== newYExtent[1] && newYExtent[1] !== undefined;
      var lineChange = lineData !== newLines || Array.isArray(lineData) && Array.isArray(newLines) && !!lineData.find(function (p) {
        return newLines.indexOf(p) === -1;
      });
      var summaryChange = summaryData !== newSummaries || Array.isArray(summaryData) && Array.isArray(newSummaries) && !!summaryData.find(function (p) {
        return newSummaries.indexOf(p) === -1;
      });
      var pointChange = pointData !== newPoints || Array.isArray(pointData) && Array.isArray(newPoints) && !!pointData.find(function (p) {
        return newPoints.indexOf(p) === -1;
      });

      if (oldDataVersion && oldDataVersion !== newDataVersion || !this.state.fullDataset) {
        this.calculateXYFrame(nextProps, true);
      } else if (lineChange || summaryChange || pointChange || oldSize[0] !== newSize[0] || oldSize[1] !== newSize[1] || extentChange || !oldDataVersion && _frame_props.xyFrameChangeProps.find(function (d) {
        return _this2.props[d] !== nextProps[d];
      })) {
        var dataChanged = lineChange || summaryChange || pointChange || extentChange || !!_frame_props.xyFrameDataProps.find(function (d) {
          return _this2.props[d] !== nextProps[d];
        });
        this.calculateXYFrame(nextProps, dataChanged);
      }
    }
  }, {
    key: "screenScales",
    value: function screenScales(_ref3) {
      var xExtent = _ref3.xExtent,
          yExtent = _ref3.yExtent,
          adjustedSize = _ref3.adjustedSize,
          xScaleType = _ref3.xScaleType,
          yScaleType = _ref3.yScaleType;
      var xDomain = [0, adjustedSize[0]];
      var yDomain = [adjustedSize[1], 0];
      var xScale = xScaleType;
      var yScale = yScaleType;

      if (xScaleType.domain) {
        xScaleType.domain(xExtent);
      }

      if (yScaleType.domain) {
        yScaleType.domain(yExtent);
      }

      xScaleType.range(xDomain);
      yScaleType.range(yDomain);
      return {
        xScale: xScale,
        yScale: yScale
      };
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          downloadFields = _this$props.downloadFields,
          xAccessor = _this$props.xAccessor,
          yAccessor = _this$props.yAccessor,
          lines = _this$props.lines,
          points = _this$props.points,
          areas = _this$props.areas,
          _this$props$summaries = _this$props.summaries,
          summaries = _this$props$summaries === void 0 ? areas : _this$props$summaries,
          name = _this$props.name,
          download = _this$props.download,
          size = _this$props.size,
          className = _this$props.className,
          annotationSettings = _this$props.annotationSettings,
          annotations = _this$props.annotations,
          additionalDefs = _this$props.additionalDefs,
          hoverAnnotation = _this$props.hoverAnnotation,
          interaction = _this$props.interaction,
          customClickBehavior = _this$props.customClickBehavior,
          customHoverBehavior = _this$props.customHoverBehavior,
          customDoubleClickBehavior = _this$props.customDoubleClickBehavior,
          canvasPostProcess = _this$props.canvasPostProcess,
          baseMarkProps = _this$props.baseMarkProps,
          useSpans = _this$props.useSpans,
          canvasAreas = _this$props.canvasAreas,
          _this$props$canvasSum = _this$props.canvasSummaries,
          canvasSummaries = _this$props$canvasSum === void 0 ? canvasAreas : _this$props$canvasSum,
          canvasPoints = _this$props.canvasPoints,
          canvasLines = _this$props.canvasLines,
          afterElements = _this$props.afterElements,
          beforeElements = _this$props.beforeElements,
          renderOrder = _this$props.renderOrder,
          matte = _this$props.matte,
          frameKey = _this$props.frameKey,
          showLinePoints = _this$props.showLinePoints;
      var _this$state4 = this.state,
          backgroundGraphics = _this$state4.backgroundGraphics,
          foregroundGraphics = _this$state4.foregroundGraphics,
          adjustedPosition = _this$state4.adjustedPosition,
          adjustedSize = _this$state4.adjustedSize,
          margin = _this$state4.margin,
          axes = _this$state4.axes,
          axesTickLines = _this$state4.axesTickLines,
          xScale = _this$state4.xScale,
          yScale = _this$state4.yScale,
          dataVersion = _this$state4.dataVersion,
          fullDataset = _this$state4.fullDataset,
          areaAnnotations = _this$state4.areaAnnotations,
          legendSettings = _this$state4.legendSettings,
          xyFrameRender = _this$state4.xyFrameRender,
          annotatedSettings = _this$state4.annotatedSettings,
          overlay = _this$state4.overlay;
      var downloadButton;

      if (download && (points || lines)) {
        var downloadData = download === "points" ? mapParentsToPoints(fullDataset) : points || lines || summaries || areas;
        downloadButton = React.createElement(_DownloadButton.default, {
          csvName: "".concat(name, "-").concat(new Date().toJSON()),
          width: Math.floor(size[0]),
          data: (0, _downloadDataMapping.xyDownloadMapping)({
            data: downloadData,
            xAccessor: download === "points" || points ? (0, _dataFunctions.stringToArrayFn)(xAccessor) : undefined,
            yAccessor: download === "points" || points ? (0, _dataFunctions.stringToArrayFn)(yAccessor) : undefined,
            fields: downloadFields
          })
        });
      }

      return React.createElement(_Frame.default, {
        name: "xyframe",
        renderPipeline: xyFrameRender,
        adjustedPosition: adjustedPosition,
        size: size,
        projectedCoordinateNames: projectedCoordinateNames,
        xScale: xScale,
        yScale: yScale,
        axes: axes,
        axesTickLines: axesTickLines,
        title: annotatedSettings.title,
        dataVersion: dataVersion,
        matte: matte,
        className: className,
        adjustedSize: adjustedSize,
        frameKey: frameKey || xyframeKey,
        additionalDefs: additionalDefs,
        hoverAnnotation: hoverAnnotation,
        defaultSVGRule: this.defaultXYSVGRule,
        defaultHTMLRule: this.defaultXYHTMLRule,
        annotations: areaAnnotations.length > 0 ? [].concat(_toConsumableArray(annotations), _toConsumableArray(areaAnnotations)) : annotations,
        annotationSettings: annotationSettings,
        legendSettings: legendSettings,
        projectedYMiddle: _coordinateNames.projectedYMiddle,
        interaction: interaction,
        customClickBehavior: customClickBehavior,
        customHoverBehavior: customHoverBehavior,
        customDoubleClickBehavior: customDoubleClickBehavior,
        points: fullDataset,
        showLinePoints: typeof showLinePoints === "string" ? showLinePoints : undefined,
        margin: margin,
        backgroundGraphics: backgroundGraphics,
        foregroundGraphics: foregroundGraphics,
        beforeElements: beforeElements,
        afterElements: afterElements,
        downloadButton: downloadButton,
        disableContext: this.props.disableContext,
        canvasPostProcess: canvasPostProcess,
        baseMarkProps: baseMarkProps,
        useSpans: useSpans,
        canvasRendering: !!(canvasSummaries || canvasPoints || canvasLines),
        renderOrder: renderOrder,
        overlay: overlay
      });
    }
  }]);

  return XYFrame;
}(React.Component);

_defineProperty(XYFrame, "defaultProps", {
  annotations: [],
  foregroundGraphics: undefined,
  size: [500, 500],
  className: "",
  lineType: "line",
  name: "xyframe",
  dataVersion: undefined
});

_defineProperty(XYFrame, "displayName", "XYFrame");

var _default = XYFrame;
exports.default = _default;
module.exports = exports.default;
